{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CLLeQw9Mxab4"
      },
      "source": [
        "## Ссылка на диск с картированными чтениями"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iX7NcoC_CLY0"
      },
      "source": [
        "\n",
        "| | SRA accession  | Stage    |\n",
        "|---:|:-------------|:-----------|\n",
        "| 1 | SRR5836473  | 8 cell       |\n",
        "| 2 | SRR3824222 | Epiblast    |\n",
        "| 3 | SRR5836475  | ICM       |\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RQAK_lggxsXi"
      },
      "source": [
        "[Bam-файлы и отчеты](https://drive.google.com/drive/folders/1hGzGjuUZNLIeUsCdSxbEzgthcV_BaqMw?usp=sharing)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5l0a7C6JHe40"
      },
      "source": [
        "# Установка FastQC"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "esYQuHKO8nAg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "059a6a15-3fb6-497d-a541-a2286be6b662"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-02-10 17:32:53--  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
            "Resolving www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)... 149.155.133.4\n",
            "Connecting to www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)|149.155.133.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 10249221 (9.8M) [application/zip]\n",
            "Saving to: ‘fastqc_v0.11.9.zip.1’\n",
            "\n",
            "fastqc_v0.11.9.zip. 100%[===================>]   9.77M  11.1MB/s    in 0.9s    \n",
            "\n",
            "2024-02-10 17:32:54 (11.1 MB/s) - ‘fastqc_v0.11.9.zip.1’ saved [10249221/10249221]\n",
            "\n",
            "Archive:  fastqc_v0.11.9.zip\n",
            "  inflating: FastQC/cisd-jhdf5.jar   \n",
            "   creating: FastQC/Configuration/\n",
            "  inflating: FastQC/Configuration/adapter_list.txt  \n",
            "  inflating: FastQC/Configuration/contaminant_list.txt  \n",
            "  inflating: FastQC/Configuration/limits.txt  \n",
            "  inflating: FastQC/fastqc           \n",
            "  inflating: FastQC/fastqc_icon.ico  \n",
            "   creating: FastQC/Help/\n",
            "   creating: FastQC/Help/1 Introduction/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/entries  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/props/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/text-base/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  \n",
            "   creating: FastQC/Help/2 Basic Operations/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/entries  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/props/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/text-base/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.1 Opening a sequence file.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.2 Evaluating Results.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.3 Saving a Report.html.svn-base  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/2 Basic Operations/2.1 Opening a sequence file.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.2 Evaluating Results.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.3 Saving a Report.html  \n",
            "   creating: FastQC/Help/3 Analysis Modules/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/entries  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/prop-base/\n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/prop-base/kmer_profiles.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_n_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_sequence_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_tile_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/props/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/text-base/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/1 Basic Statistics.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/10 Adapter Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/11 Kmer Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/12 Per Tile Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/2 Per Base Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/3 Per Sequence Quality Scores.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/4 Per Base Sequence Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/5 Per Sequence GC Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/6 Per Base N Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/7 Sequence Length Distribution.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/8 Duplicate Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/9 Overrepresented Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/kmer_profiles.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_n_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_sequence_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_tile_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/1 Basic Statistics.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/10 Adapter Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/11 Kmer Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/12 Per Tile Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/2 Per Base Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/3 Per Sequence Quality Scores.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/4 Per Base Sequence Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/5 Per Sequence GC Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/6 Per Base N Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/7 Sequence Length Distribution.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/8 Duplicate Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/9 Overrepresented Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/duplication_levels.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/kmer_profiles.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_n_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_sequence_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_tile_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/sequence_length_distribution.png  \n",
            "  inflating: FastQC/INSTALL.txt      \n",
            "  inflating: FastQC/jbzip2-0.9.jar   \n",
            "  inflating: FastQC/LICENSE          \n",
            "  inflating: FastQC/LICENSE.txt      \n",
            "  inflating: FastQC/LICENSE_JHDF5.txt  \n",
            "   creating: FastQC/net/\n",
            "   creating: FastQC/net/sourceforge/\n",
            "   creating: FastQC/net/sourceforge/iharder/\n",
            "   creating: FastQC/net/sourceforge/iharder/base64/\n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$1.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$InputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$OutputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64.class  \n",
            "   creating: FastQC/org/\n",
            "   creating: FastQC/org/apache/\n",
            "   creating: FastQC/org/apache/commons/\n",
            "   creating: FastQC/org/apache/commons/math3/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/solvers/\n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AllowedSolution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseAbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BracketedUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BrentSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolverUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/UnivariateFunction.class  \n",
            "   creating: FastQC/org/apache/commons/math3/distribution/\n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractIntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BetaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BinomialDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/CauchyDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ChiSquaredDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/FDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/GammaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/HypergeometricDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/IntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/NormalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PascalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PoissonDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/RealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/SaddlePointExpansion.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/TDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/WeibullDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ZipfDistribution.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/ConvergenceException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/DimensionMismatchException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathArithmeticException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalStateException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathInternalError.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MaxCountExceededException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NoBracketingException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotFiniteNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotStrictlyPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NullArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooLargeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooSmallException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/OutOfRangeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/TooManyEvaluationsException.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ArgUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContext.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContextProvider.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/Localizable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/LocalizedFormats.class  \n",
            "   creating: FastQC/org/apache/commons/math3/random/\n",
            "  inflating: FastQC/org/apache/commons/math3/random/AbstractWell.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/BitsStreamGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomData.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomDataImpl.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/Well19937c.class  \n",
            "   creating: FastQC/org/apache/commons/math3/special/\n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Erf.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma.class  \n",
            "   creating: FastQC/org/apache/commons/math3/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/util/ArithmeticUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ContinuedFraction.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/DoubleArray.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpFracTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpIntTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$lnMant.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathCalc.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathLiteralArrays.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$MaxCountExceededCallback.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/MathUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Precision.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ResizableDoubleArray.class  \n",
            "  inflating: FastQC/README.md        \n",
            "  inflating: FastQC/README.txt       \n",
            "  inflating: FastQC/RELEASE_NOTES.txt  \n",
            "  inflating: FastQC/run_fastqc.bat   \n",
            "  inflating: FastQC/sam-1.103.jar    \n",
            "   creating: FastQC/Templates/\n",
            "  inflating: FastQC/Templates/fastqc2fo.xsl  \n",
            "  inflating: FastQC/Templates/header_template.html  \n",
            "   creating: FastQC/Templates/Icons/\n",
            " extracting: FastQC/Templates/Icons/error.png  \n",
            " extracting: FastQC/Templates/Icons/fastqc_icon.png  \n",
            " extracting: FastQC/Templates/Icons/tick.png  \n",
            " extracting: FastQC/Templates/Icons/warning.png  \n",
            "   creating: FastQC/uk/\n",
            "   creating: FastQC/uk/ac/\n",
            "   creating: FastQC/uk/ac/babraham/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Analysis/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisListener.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisQueue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisRunner.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/OfflineRunner.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Dialogs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel$SmoothJLabel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/WelcomePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCMenuBar.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/FileFilters/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/BAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/CasavaFastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/FastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/GobyFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/MappedBAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/SequenceFileFilter.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Graphs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/BaseGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/LineGraph.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/QualityBoxPlot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/TileGraph.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Help/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot$FileSorter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPage.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay$HelpEditor.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpSearchPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AbstractQCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$Adapter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/DuplicationLevel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModelValue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$Kmer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/NContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$OverrepresentedSeq.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseSequenceContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceGCContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerTileQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/QCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/SequenceLengthDistribution.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Report/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/HTMLReportArchive.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/stylesheet.txt  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Resources/\n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/error.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.png  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.svg  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon_100.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/tick.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/warning.png  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Results/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel$ModuleRenderer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/BAMFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/Contaminant.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminantHit.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminentFinder.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Fast5File.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/FastQFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/PhredEncoding.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Sequence.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFile.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFileGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFormatException.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Statistics/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/NormalDistribution.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/PearsonCorrelation.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Utilities/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/CasavaBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/HotColdColourGradient.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageToBase64.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/MultiMemberGZIPInputStream.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NameFormatException.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NanoporeBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/QualityCount.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/RGB.class  \n"
          ]
        }
      ],
      "source": [
        "!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
        "!unzip fastqc_v0.11.9.zip\n",
        "!chmod a+x FastQC/fastqc"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_zEx3USYPKoR"
      },
      "source": [
        "Можно получать данные секвенировия с ENA. Например, https://www.ebi.ac.uk/ena/browser/view/SRR5836473"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Скачиваю данные с ENA, распаковываю их и делаю fastqc отчет. Затем  удаляю сначала образец секвенирования и папку FastQC для экономии места на диске."
      ],
      "metadata": {
        "id": "yDo3kVejC5ds"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "c5QMGKJYxrza",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e683687a-c15f-4c63-f57b-a1d3a3015206"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-02-10 16:51:44--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/003/SRR5836473/SRR5836473_1.fastq.gz\n",
            "           => ‘SRR5836473_1.fastq.gz’\n",
            "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.165\n",
            "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.165|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR583/003/SRR5836473 ... done.\n",
            "==> SIZE SRR5836473_1.fastq.gz ... 6856386252\n",
            "==> PASV ... done.    ==> RETR SRR5836473_1.fastq.gz ... done.\n",
            "Length: 6856386252 (6.4G) (unauthoritative)\n",
            "\n",
            "SRR5836473_1.fastq. 100%[===================>]   6.38G  33.5MB/s    in 3m 23s  \n",
            "\n",
            "2024-02-10 16:55:08 (32.2 MB/s) - ‘SRR5836473_1.fastq.gz’ saved [6856386252]\n",
            "\n",
            "./FastQC/fastqc SRR5836473_1.fastq\n",
            "Started analysis of SRR5836473_1.fastq\n",
            "Approx 5% complete for SRR5836473_1.fastq\n",
            "Approx 10% complete for SRR5836473_1.fastq\n",
            "Approx 15% complete for SRR5836473_1.fastq\n",
            "Approx 20% complete for SRR5836473_1.fastq\n",
            "Approx 25% complete for SRR5836473_1.fastq\n",
            "Approx 30% complete for SRR5836473_1.fastq\n",
            "Approx 35% complete for SRR5836473_1.fastq\n",
            "Approx 40% complete for SRR5836473_1.fastq\n",
            "Approx 45% complete for SRR5836473_1.fastq\n",
            "Approx 50% complete for SRR5836473_1.fastq\n",
            "Approx 55% complete for SRR5836473_1.fastq\n",
            "Approx 60% complete for SRR5836473_1.fastq\n",
            "Approx 65% complete for SRR5836473_1.fastq\n",
            "Approx 70% complete for SRR5836473_1.fastq\n",
            "Approx 75% complete for SRR5836473_1.fastq\n",
            "Approx 80% complete for SRR5836473_1.fastq\n",
            "Approx 85% complete for SRR5836473_1.fastq\n",
            "Approx 90% complete for SRR5836473_1.fastq\n",
            "Approx 95% complete for SRR5836473_1.fastq\n",
            "Analysis complete for SRR5836473_1.fastq\n"
          ]
        }
      ],
      "source": [
        "! wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/003/SRR5836473/SRR5836473_1.fastq.gz\n",
        "! gunzip SRR5836473_1.fastq.gz\n",
        "! ls SRR* | xargs -P 2 -tI{} ./FastQC/fastqc {}"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! rm SRR5836473_1.fastq"
      ],
      "metadata": {
        "id": "vCS0uajcDamO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!rm -rf FastQC"
      ],
      "metadata": {
        "id": "o5mZMl9EDj7m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Ugk3bV0LI5El"
      },
      "source": [
        "# Установка Bismark с зависимостями (+bowtie2, samtools)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uSOZLyn5n5Nl",
        "outputId": "2b8be1ae-5ae2-4bc4-9b32-9730a76c7850"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
            "\u001b[0m✨🍰✨ Everything looks OK!\n"
          ]
        }
      ],
      "source": [
        "! pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "npb--DNtn52o",
        "outputId": "6800d6f1-ccf1-4f78-fa8c-144a4fc991fa"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "✨🍰✨ Everything looks OK!\n"
          ]
        }
      ],
      "source": [
        "import condacolab\n",
        "condacolab.check()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "MfBSN4YuoCL6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "89922c83-8de9-4cbc-d3d0-1be453953d4b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Solving environment: / \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - bismark\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    bismark-0.24.2             |       hdfd78af_0         7.7 MB  bioconda\n",
            "    bowtie2-2.5.3              |  py310ha0a81b8_0        14.1 MB  bioconda\n",
            "    ca-certificates-2024.2.2   |       hbcca054_0         152 KB  conda-forge\n",
            "    certifi-2024.2.2           |     pyhd8ed1ab_0         157 KB  conda-forge\n",
            "    coreutils-8.25             |                1         9.1 MB  bioconda\n",
            "    hisat2-2.2.1               |       hdbdd923_6        16.2 MB  bioconda\n",
            "    htslib-1.19.1              |       h81da01d_1         2.9 MB  bioconda\n",
            "    k8-0.2.5                   |       hdcf5f25_4         1.7 MB  bioconda\n",
            "    libdeflate-1.18            |       h0b41bf4_0          64 KB  conda-forge\n",
            "    libgcc-7.2.0               |       h69d50b8_2         304 KB  conda-forge\n",
            "    libhwloc-2.9.3             |default_h554bfaf_1009         2.5 MB  conda-forge\n",
            "    libxcrypt-4.4.36           |       hd590300_1          98 KB  conda-forge\n",
            "    minimap2-2.26              |       he4a0461_2         1.1 MB  bioconda\n",
            "    openssl-3.2.1              |       hd590300_0         2.7 MB  conda-forge\n",
            "    perl-5.32.1                | 7_hd590300_perl5        12.7 MB  conda-forge\n",
            "    samtools-1.19.2            |       h50ea8bc_0         452 KB  bioconda\n",
            "    tbb-2021.11.0              |       h00ab1b0_1         191 KB  conda-forge\n",
            "    zlib-1.2.13                |       hd590300_5          91 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        72.1 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  bismark            bioconda/noarch::bismark-0.24.2-hdfd78af_0 \n",
            "  bowtie2            bioconda/linux-64::bowtie2-2.5.3-py310ha0a81b8_0 \n",
            "  coreutils          bioconda/linux-64::coreutils-8.25-1 \n",
            "  hisat2             bioconda/linux-64::hisat2-2.2.1-hdbdd923_6 \n",
            "  htslib             bioconda/linux-64::htslib-1.19.1-h81da01d_1 \n",
            "  k8                 bioconda/linux-64::k8-0.2.5-hdcf5f25_4 \n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.18-h0b41bf4_0 \n",
            "  libgcc             conda-forge/linux-64::libgcc-7.2.0-h69d50b8_2 \n",
            "  libhwloc           conda-forge/linux-64::libhwloc-2.9.3-default_h554bfaf_1009 \n",
            "  libxcrypt          conda-forge/linux-64::libxcrypt-4.4.36-hd590300_1 \n",
            "  minimap2           bioconda/linux-64::minimap2-2.26-he4a0461_2 \n",
            "  perl               conda-forge/linux-64::perl-5.32.1-7_hd590300_perl5 \n",
            "  samtools           bioconda/linux-64::samtools-1.19.2-h50ea8bc_0 \n",
            "  tbb                conda-forge/linux-64::tbb-2021.11.0-h00ab1b0_1 \n",
            "  zlib               conda-forge/linux-64::zlib-1.2.13-hd590300_5 \n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  ca-certificates                     2023.11.17-hbcca054_0 --> 2024.2.2-hbcca054_0 \n",
            "  certifi                           2023.11.17-pyhd8ed1ab_0 --> 2024.2.2-pyhd8ed1ab_0 \n",
            "  openssl                                  3.2.0-hd590300_1 --> 3.2.1-hd590300_0 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "hisat2-2.2.1         | 16.2 MB   | :   0% 0/1 [00:00<?, ?it/s]\n",
            "bowtie2-2.5.3        | 14.1 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.19.1        | 2.9 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.2.1        | 2.7 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libhwloc-2.9.3       | 2.5 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "k8-0.2.5             | 1.7 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.26        | 1.1 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.19.2      | 452 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libgcc-7.2.0         | 304 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "tbb-2021.11.0        | 191 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2024.2.2     | 157 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2024 | 152 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libxcrypt-4.4.36     | 98 KB     | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.2.13          | 91 KB     | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.18      | 64 KB     | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "hisat2-2.2.1         | 16.2 MB   | :   0% 0.0009635402010155163/1 [00:00<02:16, 136.97s/it]\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :   0% 0.0012277751453917628/1 [00:00<01:44, 104.66s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | :   0% 0.0020357622695866633/1 [00:00<01:04, 64.25s/it]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :   0% 0.0017100384706469664/1 [00:00<01:52, 112.74s/it]\u001b[A\u001b[A\u001b[A\n",
            "hisat2-2.2.1         | 16.2 MB   | :  18% 0.18018201758990154/1 [00:00<00:00,  1.07s/it]   \n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :  23% 0.23204950247904318/1 [00:00<00:00,  1.17it/s]   \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | :  40% 0.3969736425693994/1 [00:00<00:00,  1.90it/s]   \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :  35% 0.3454277710706872/1 [00:00<00:00,  1.46it/s]    \u001b[A\u001b[A\u001b[A\n",
            "hisat2-2.2.1         | 16.2 MB   | :  40% 0.3950514824163617/1 [00:00<00:00,  1.45it/s] \n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :  53% 0.5316266379546333/1 [00:00<00:00,  1.94it/s] \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | :  96% 0.9629155535144919/1 [00:00<00:00,  3.43it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :  83% 0.8259485813224847/1 [00:00<00:00,  2.72it/s]\u001b[A\u001b[A\u001b[A\n",
            "hisat2-2.2.1         | 16.2 MB   | :  66% 0.6590614974946132/1 [00:00<00:00,  1.90it/s]\n",
            "\n",
            "hisat2-2.2.1         | 16.2 MB   | :  87% 0.8739309623210733/1 [00:00<00:00,  1.92it/s]\n",
            "bowtie2-2.5.3        | 14.1 MB   | :  98% 0.9771901912318356/1 [00:00<00:00,  1.79it/s]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.19.1        | 2.9 MB    | :   1% 0.005448681494902171/1 [00:00<01:44, 105.16s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.2.1        | 2.7 MB    | :   1% 0.005722530613128779/1 [00:00<01:49, 110.17s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libhwloc-2.9.3       | 2.5 MB    | :   1% 0.006263799072128885/1 [00:00<02:06, 127.05s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "k8-0.2.5             | 1.7 MB    | :   1% 0.009349703684470327/1 [00:00<01:26, 87.23s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.26        | 1.1 MB    | :   1% 0.01474465929073725/1 [00:00<00:55, 55.98s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.19.2      | 452 KB    | :   4% 0.03536926819896077/1 [00:00<00:24, 25.72s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libgcc-7.2.0         | 304 KB    | :   5% 0.05255779785906529/1 [00:00<00:16, 17.72s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "tbb-2021.11.0        | 191 KB    | :   8% 0.08378848317479799/1 [00:00<00:10, 11.53s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2024 | 152 KB    | :  11% 0.10540943949765814/1 [00:00<00:08,  9.19s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2024.2.2     | 157 KB    | :  10% 0.10204348557228184/1 [00:00<00:08,  9.62s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libxcrypt-4.4.36     | 98 KB     | :  16% 0.163198629386511/1 [00:00<00:05,  6.06s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.2.13          | 91 KB     | :  18% 0.1765041745219499/1 [00:00<00:04,  5.66s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.18      | 64 KB     | :  25% 0.25137701950074415/1 [00:01<00:02,  3.98s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | : 100% 1.0/1 [00:04<00:00,  3.43it/s]               \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | : 100% 1.0/1 [00:05<00:00,  2.72it/s]               \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.19.1        | 2.9 MB    | : 100% 1.0/1 [00:05<00:00,  5.83s/it]                  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.19.1        | 2.9 MB    | : 100% 1.0/1 [00:05<00:00,  5.83s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.2.1        | 2.7 MB    | : 100% 1.0/1 [00:06<00:00,  5.99s/it]                  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.2.1        | 2.7 MB    | : 100% 1.0/1 [00:06<00:00,  5.99s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | : 100% 1.0/1 [00:08<00:00,  2.26it/s]              \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.26        | 1.1 MB    | : 100% 1.0/1 [00:08<00:00,  7.98s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.26        | 1.1 MB    | : 100% 1.0/1 [00:08<00:00,  7.98s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "k8-0.2.5             | 1.7 MB    | : 100% 1.0/1 [00:08<00:00,  8.39s/it]                 \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "k8-0.2.5             | 1.7 MB    | : 100% 1.0/1 [00:08<00:00,  8.39s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.19.2      | 452 KB    | : 100% 1.0/1 [00:08<00:00,  8.60s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.19.2      | 452 KB    | : 100% 1.0/1 [00:08<00:00,  8.60s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "bowtie2-2.5.3        | 14.1 MB   | : 100% 1.0/1 [00:10<00:00,  1.79it/s]               \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libgcc-7.2.0         | 304 KB    | : 100% 1.0/1 [00:10<00:00, 10.01s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libgcc-7.2.0         | 304 KB    | : 100% 1.0/1 [00:10<00:00, 10.01s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libhwloc-2.9.3       | 2.5 MB    | : 100% 1.0/1 [00:10<00:00, 10.21s/it]                  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libhwloc-2.9.3       | 2.5 MB    | : 100% 1.0/1 [00:10<00:00, 10.21s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2024 | 152 KB    | : 100% 1.0/1 [00:10<00:00, 10.48s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2024 | 152 KB    | : 100% 1.0/1 [00:10<00:00, 10.48s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "tbb-2021.11.0        | 191 KB    | : 100% 1.0/1 [00:10<00:00, 10.43s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "tbb-2021.11.0        | 191 KB    | : 100% 1.0/1 [00:10<00:00, 10.43s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2024.2.2     | 157 KB    | : 100% 1.0/1 [00:10<00:00, 10.51s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2024.2.2     | 157 KB    | : 100% 1.0/1 [00:10<00:00, 10.51s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libxcrypt-4.4.36     | 98 KB     | : 100% 1.0/1 [00:10<00:00, 10.73s/it]              \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libxcrypt-4.4.36     | 98 KB     | : 100% 1.0/1 [00:10<00:00, 10.73s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.18      | 64 KB     | : 100% 1.0/1 [00:10<00:00, 11.05s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.18      | 64 KB     | : 100% 1.0/1 [00:10<00:00, 11.05s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.2.13          | 91 KB     | : 100% 1.0/1 [00:10<00:00, 10.81s/it]               \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \n",
            "                                                                        \u001b[A\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "Preparing transaction: - \b\b\\ \b\bdone\n",
            "Verifying transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Executing transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n"
          ]
        }
      ],
      "source": [
        "! conda install -c bioconda bismark"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "13VM_exnoHEl",
        "outputId": "52df91fc-f82e-4eea-9276-ec1bf4ee96fc"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Usage: samtools view [options] <in.bam>|<in.sam>|<in.cram> [region ...]\n",
            "\n",
            "Output options:\n",
            "  -b, --bam                  Output BAM\n",
            "  -C, --cram                 Output CRAM (requires -T)\n",
            "  -1, --fast                 Use fast BAM compression (and default to --bam)\n",
            "  -u, --uncompressed         Uncompressed BAM output (and default to --bam)\n",
            "  -h, --with-header          Include header in SAM output\n",
            "  -H, --header-only          Print SAM header only (no alignments)\n",
            "      --no-header            Print SAM alignment records only [default]\n",
            "  -c, --count                Print only the count of matching records\n",
            "  -o, --output FILE          Write output to FILE [standard output]\n",
            "  -U, --unoutput FILE, --output-unselected FILE\n",
            "                             Output reads not selected by filters to FILE\n",
            "  -p, --unmap                Set flag to UNMAP on reads not selected\n",
            "                             then write to output file.\n",
            "  -P, --fetch-pairs          Retrieve complete pairs even when outside of region\n",
            "Input options:\n",
            "  -t, --fai-reference FILE   FILE listing reference names and lengths\n",
            "  -M, --use-index            Use index and multi-region iterator for regions\n",
            "      --region[s]-file FILE  Use index to include only reads overlapping FILE\n",
            "  -X, --customized-index     Expect extra index file argument after <in.bam>\n",
            "\n",
            "Filtering options (Only include in output reads that...):\n",
            "  -L, --target[s]-file FILE  ...overlap (BED) regions in FILE\n",
            "  -N, --qname-file [^]FILE   ...whose read name is listed in FILE (\"^\" negates)\n",
            "  -r, --read-group STR       ...are in read group STR\n",
            "  -R, --read-group-file [^]FILE\n",
            "                             ...are in a read group listed in FILE\n",
            "  -d, --tag STR1[:STR2]      ...have a tag STR1 (with associated value STR2)\n",
            "  -D, --tag-file STR:FILE    ...have a tag STR whose value is listed in FILE\n",
            "  -q, --min-MQ INT           ...have mapping quality >= INT\n",
            "  -l, --library STR          ...are in library STR\n",
            "  -m, --min-qlen INT         ...cover >= INT query bases (as measured via CIGAR)\n",
            "  -e, --expr STR             ...match the filter expression STR\n",
            "  -f, --require-flags FLAG   ...have all of the FLAGs present\n",
            "  -F, --excl[ude]-flags FLAG ...have none of the FLAGs present\n",
            "      --rf, --incl-flags, --include-flags FLAG\n",
            "                             ...have some of the FLAGs present\n",
            "  -G FLAG                    EXCLUDE reads with all of the FLAGs present\n",
            "      --subsample FLOAT      Keep only FLOAT fraction of templates/read pairs\n",
            "      --subsample-seed INT   Influence WHICH reads are kept in subsampling [0]\n",
            "  -s INT.FRAC                Same as --subsample 0.FRAC --subsample-seed INT\n",
            "\n",
            "Processing options:\n",
            "      --add-flags FLAG       Add FLAGs to reads\n",
            "      --remove-flags FLAG    Remove FLAGs from reads\n",
            "  -x, --remove-tag STR\n",
            "               Comma-separated read tags to strip (repeatable) [null]\n",
            "      --keep-tag STR\n",
            "               Comma-separated read tags to preserve (repeatable) [null].\n",
            "               Equivalent to \"-x ^STR\"\n",
            "  -B, --remove-B             Collapse the backward CIGAR operation\n",
            "  -z, --sanitize FLAGS       Perform sanitity checking and fixing on records.\n",
            "                             FLAGS is comma separated (see manual). [off]\n",
            "\n",
            "General options:\n",
            "  -?, --help   Print long help, including note about region specification\n",
            "  -S           Ignored (input format is auto-detected)\n",
            "      --no-PG  Do not add a PG line\n",
            "      --input-fmt-option OPT[=VAL]\n",
            "               Specify a single input file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -O, --output-fmt FORMAT[,OPT[=VAL]]...\n",
            "               Specify output format (SAM, BAM, CRAM)\n",
            "      --output-fmt-option OPT[=VAL]\n",
            "               Specify a single output file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -T, --reference FILE\n",
            "               Reference sequence FASTA FILE [null]\n",
            "  -@, --threads INT\n",
            "               Number of additional threads to use [0]\n",
            "      --write-index\n",
            "               Automatically index the output files [off]\n",
            "      --verbosity INT\n",
            "               Set level of verbosity\n",
            "\n"
          ]
        }
      ],
      "source": [
        "! samtools view"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "1. Скачиваем чтения из SRA с помощью SRAtoolkit или любым другим способом\n",
        "\n",
        "2. Индексируем геном\n",
        "\n",
        "3. Строим выравнивание\n",
        "\n",
        "(Описание шагов 2-3 см https://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_User_Guide.pdf)\n",
        "\n",
        "#Внимание!\n",
        "\n",
        "Т.к. вычислительные мощности нужны большие, то этот шаг был сделан за вас.\n",
        " Достаточно скачать выравнивания с диска https://drive.google.com/drive/folders/1hGzGjuUZNLIeUsCdSxbEzgthcV_BaqMw?usp=sharing или с сервера майнора)"
      ],
      "metadata": {
        "id": "XVeS3m8fWWNR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Скаченные с приложенного диска файлы добавляю на свой диск и работаю с ними через обращение к конкретной папке на диске"
      ],
      "metadata": {
        "id": "z_ZjabvhLAOf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bW6SUXbcGazD",
        "outputId": "80b5a601-61d9-4889-b967-94d8f0a71f96"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Проверка на корректность скаченных файлов"
      ],
      "metadata": {
        "id": "45KDAdofnF8G"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! md5sum -c checksum.md5"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "e7T0myqZlimj",
        "outputId": "11aca832-5bf7-49ba-d00f-3cdd8062f3e7"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam: OK\n",
            "/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam: OK\n",
            "/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam: OK\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AxMpAQFVQFR8",
        "outputId": "93309251-8009-4e94-a0d2-98b2658277d7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "chr11.fa\t\t\t\tSRR5836473_1_bismark_bt2_PE_report.txt\n",
            "SRR3824222_1_bismark_bt2_pe.bam\t\tSRR5836475_1_bismark_bt2_pe.bam\n",
            "SRR3824222_1_bismark_bt2_PE_report.txt\tSRR5836475_1_bismark_bt2_PE_report.txt\n",
            "SRR5836473_1_bismark_bt2_pe.bam\n"
          ]
        }
      ],
      "source": [
        "! ls /content/drive/MyDrive/bioinf"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yMQy_LQ8xSOP",
        "outputId": "dec67492-0e7a-4183-9f91-e86187790ff7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "SRR5836473.41_41_length=140\t99\tchr11\t40076664\t0\t140M\t=\t40076703\t178\tAATTTTTTTGGGGATTGTATTGAATTTGTAGATTGTTTTTGGTAAGAGAGTTATTTTTATTATATTGATTTTGTTAATTTATGAATATGGGAGATTTTTTTATTTTTTGAGATTTTTTTTAATTTTTTTTTTTAGAGATT\tKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKFKKKKKKKKKFKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKF7<77FF\tNM:i:28\tMD:Z:6G0A9C7C9C6C4T2C0C7C9C0C2C0C3C0C5C9C3C0C2C2C6C2C8C3C2C5C1\tXM:Z:.................h.......x.........h......h.......hh.......h.........hx..hh...hh.....h.........h...hh..h..x......h..h........h...h..x.....h.\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.41_41_length=140\t147\tchr11\t40076703\t0\t139M\t=\t40076664\t-178\tTGGTAAGAGAGTTATTTTTATTATATTGATTTTGTTAATTTATGAATATGGGAGATTTTTTTATTTTTTGAGATTTTTTTTAATTTTTTTTTTTAGAGATTTGAAGTTTTTATTATATAGATTTTTTATTTTTTTAGTG\tKKKKKAAFFFKKKKKKKKKKKKKFFKKKKKKKKKKFAFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK\tNM:i:31\tMD:Z:3C4T2C0C7C9C0C2C0C3C0C5C9C3C0C2C2C6C2C8C3C2C5C8C4C3A4C3C1C2C6T0\tXM:Z:...h.......hh.......h.........hx..hh...hh.....h.........h...hh..h..x......h..h........h...h..x.....h........h....h........h...h.h..h.......\tXR:Z:GA\tXG:Z:CT\n",
            "SRR5836473.231_231_length=140\t83\tchr11\t48401999\t0\t140M\t=\t48401992\t-147\tACAAATTACTCCTTCTAATTCATTAAAAAATTAAATTAAAATTATAATAAAAATTACATTAAATCTATAAATTACTTTTAACAAAATAACCATTTTTACAATATTAATCCTACCAATCCATAAACACAAAAAAACTTTCC\tKKKKKKKKF<KKKKKKKKKFKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKK\tNM:i:34\tMD:Z:0G2G3G13G2G2G4G1G2G0G6G2G0G0G0G3G4G5G2G3G5G0G3G3G13G2G5G9G1G2T1G0G1G1T6\tXM:Z:h..x...h.............z..h..h....h.h..hh......h..hhhh...h....h.....x..h...h.....hh...h...h.............h..h.....x.........h.h....hh.h........\tXR:Z:CT\tXG:Z:GA\n",
            "SRR5836473.231_231_length=140\t163\tchr11\t48401992\t0\t139M\t=\t48401999\t147\tTAAATTTACAAATTACTCCTTCTAATTCATTAAAAAATTAAATTAAAATTATAATAAAAATTACATTAAATCTATAAATTACTTTTAACAAAATAACCATTTTTACAATATTAATCCTACCAATCCATAAACACAAAAA\tFFKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKAKFFKKKKKKKKKKKKKKKKFKAF<FF<FKKKKKKKKKKFKKKK<FFKAFF<FKF<FK7AF7A,7<AAF<KF\tNM:i:35\tMD:Z:1G3A1G2G3G13G2G2G4G1G2G0G6G2G0G0G0G3G4G5G2G3G5G0G3G3G13G2G5G9G1G2T1G0G1G0\tXM:Z:.h.....h..x...h.............z..h..h....h.h..hh......h..hhhh...h....h.....x..h...h.....hh...h...h.............h..h.....x.........h.h....hh.h\tXR:Z:GA\tXG:Z:GA\n",
            "SRR5836473.1421_1421_length=140\t99\tchr11\t30560802\t0\t140M\t=\t30560952\t289\tTGTTGAGGGGTATTTGGGTTTTTTTTAGTTTTCGGTTATTATAAATAAGGTTGTTATGAATATAGTGTAGTATGTGTTTTTTTTATTGGTTGGAATATTTTTTGGATATATGTTTAGATGAGGTATTGTGGGATTTTTTG\tKKKKKKKKFKFKKKKKKKFKKKKKFKKKKKAK,FAKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKAKKKKKKKKKKKKKKKKKKKKKKKKKKKKKAKKKKAFFK<AFKKKKK\tNM:i:30\tMD:Z:10C2C6C3C0C2C2C0T2C9C4C2C6C6G2A6C3C3C0C8C2C2C10C0C0C3A9C5C0C1C2\tXM:Z:..........h..x......h...hx..h..x...h.........h....x..h......h................h...h...xz........h..h..x..........hhx.............z.....hh.x..\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.1421_1421_length=140\t147\tchr11\t30560952\t0\t139M\t=\t30560802\t-289\tGTTTAATTTTTTGAAGAATTGTTAGATTGATTTTTAGAGTGGTTGTAAAAGTTTGTAATTTTATTAATAATGGAGGAGTGTTTTTTTTTTTTTATATTTTTTTTAGTATTTGTTGTTATTTGAATTTTTGATTTTAGTT\t<<<<<<KF<KKKKKFFFKKAA<KAFKFKKKKKKKKKFFFFFFKKKFFAFAAKKKKFKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKFKKKKKKFF\tNM:i:41\tMD:Z:2C0C6C3G3C2C0C3C6C0C12C3C3C3C0C0C1C0C2C14C0C1C3C1C0C1C1G0C0C1C0G0C0C2C2C2C3C1C0C12C5C0\tXM:Z:..hh......x.......x..hx...x......hx................h...h...hhh.hh..h..............hh.h...h.hh.x..hh.z.hx..h..x..x...h.hx............h.....h\tXR:Z:GA\tXG:Z:CT\n",
            "SRR5836473.2061_2061_length=140\t99\tchr11\t73270947\t0\t139M\t=\t73270962\t154\tATAAGAGATAGTTATATTAGGGTTTTGTTAGTAAAATTTTGTTGGTATTTGTAGTAGTGTTTGGGTTTGGTGGTTGTTTATGGGATGGATTTATGGGTGGGGTAGTTTTTGGATGGTTTTTTTTTTTATTTTAGTTTTG\tKKKKKKKKKKKFKKKKKKKKKKFKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKFKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFFKKKKKKKFKKKK7AFAFKKKKKF7FKKKKKKFKFAKKKKKKKKF7FFKF\tNM:i:31\tMD:Z:1C6C2C5C5C0C3C2C5C3C3C2A2C1A4A1C1A27C0C10C3C1C9C2C0C2C0C2C1C2C1C2\tXM:Z:.h......x..h.....x.....hx...x..h.....h...x...h.....h........h.............................hh..........x...h.x.........h..hh..hh..h.x..h.x..\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.2061_2061_length=140\t147\tchr11\t73270962\t0\t139M\t=\t73270947\t-154\tATTAGGGTTTTGTTAGTAAAATTTTGTTGGTATTTGTAGTAGTGTTTGGGTTTGGTGGTTGTTTATGGGATGGATTTATGGGTGGGGTAGTTTTTGGATGGTTTTTTTTTTTATTTTAGTTTTGAATTTTGTTTTTGTA\tA7<AAAAA77AKKKFKKKFKKKKAKF<KKKFKKFKKFKAAKKKKFKAKKKKKKKFKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK\tNM:i:31\tMD:Z:2C5C0C3C2C5C3C3C2A2C1A4A1C1A27C0C10C3C1C9C2C0C2C0C2C1C2C1C4C5C1C4\tXM:Z:..x.....hx...x..h.....h...x...h.....h........h.............................hh..........x...h.x.........h..hh..hh..h.x..h.x....h.....h.x....\tXR:Z:GA\tXG:Z:CT\n",
            "SRR5836473.3331_3331_length=140\t99\tchr11\t43350802\t40\t140M\t=\t43350876\t214\tTTTTAATGTAAGTTGTATAAGGAGTATAATTAGGTTTAGGTGATGTTGAAAATAATATAATTTTATTTATAAATATTATTATTATTATGATAAAGGGAAAGGTGATTATTTTTAAAAGATTATGATGGTTTGTTGGTAAT\tKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKFFKKKKKKKKKF\tNM:i:29\tMD:Z:0C5C1C6C0T0C7G0C3C3C1C23C2C1C0C6C2C1C0C1C23C0C3C0C0C7C7C0C6C3\tXM:Z:h.....z.h......h.h........h...x...h.x.......................h..h.hh......h..h.hh.h.......................hh...hhh.......h.......hx......h...\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.3331_3331_length=140\t147\tchr11\t43350876\t40\t140M\t=\t43350802\t-214\tATTATTATTATTATGATAAAGGGAAAGGTGATTATTTTTAAAAGATTATGATGGTTTGTTGGTAATGTTATTGTTATACGTAGTATGTTTTGGTGTTATTTGGATTTTGTATTTTTTATAGTTTTTTTATTTTGTTAATT\tKFKKKKK<KKKKKFKAKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK\tNM:i:34\tMD:Z:2C1C0C1C23C0C3C0C0C7C7C0C6C4C2C2C6C2C3C0C7C2C0C3C0C1C1C4C0C11C1C0C4C3C0\tXM:Z:..h.hh.h.......................hh...hhh.......h.......hx......h....h..x..h....Z.x..h...hh.......h..xz...hh.z.h....hh...........h.hh....h...h\tXR:Z:GA\tXG:Z:CT\n"
          ]
        }
      ],
      "source": [
        "! samtools view /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam | head"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SrGGCxZQ2Ao3",
        "outputId": "23a00289-2080-4e1d-cc82-7fbc108c7cbd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
            "[bam_sort_core] merging from 9 files and 1 in-memory blocks...\n",
            "[bam_sort_core] merging from 5 files and 1 in-memory blocks...\n"
          ]
        }
      ],
      "source": [
        "! samtools sort /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam -o /content/drive/MyDrive/bioinf/sorted_8cell.bam\n",
        "! samtools sort /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam -o /content/drive/MyDrive/bioinf/sorted_epiblast.bam\n",
        "! samtools sort /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam -o /content/drive/MyDrive/bioinf/sorted_icm.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "GErvoOM-3Axj"
      },
      "outputs": [],
      "source": [
        "! samtools index  /content/drive/MyDrive/bioinf/sorted_8cell.bam\n",
        "! samtools index  /content/drive/MyDrive/bioinf/sorted_epiblast.bam\n",
        "! samtools index  /content/drive/MyDrive/bioinf/sorted_icm.bam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Lw--6KavBoXp"
      },
      "source": [
        "Сколько ридов закартировалось на регион 11347700-11367700; 40185800-40195800 в образцах?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b-ZhJMlax75S",
        "outputId": "ae0599e5-e36a-4e5f-b3e7-3450781e832e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "551\n",
            "194\n"
          ]
        }
      ],
      "source": [
        "! samtools view /content/drive/MyDrive/bioinf/sorted_8cell.bam chr11:11347700-11367700 | cut -f1 | sort -u | wc -l\n",
        "! samtools view /content/drive/MyDrive/bioinf/sorted_8cell.bam chr11:40185800-40195800 | cut -f1 | sort -u | wc -l"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools view /content/drive/MyDrive/bioinf/sorted_epiblast.bam chr11:11347700-11367700 | cut -f1 | sort -u | wc -l\n",
        "! samtools view /content/drive/MyDrive/bioinf/sorted_epiblast.bam chr11:40185800-40195800 | cut -f1 | sort -u | wc -l"
      ],
      "metadata": {
        "id": "iN_h3agaZw4j",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "12e82397-6676-43e7-db39-fa51173252e0"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1344\n",
            "565\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools view /content/drive/MyDrive/bioinf/sorted_icm.bam chr11:11347700-11367700 | cut -f1 | sort -u | wc -l\n",
        "! samtools view /content/drive/MyDrive/bioinf/sorted_icm.bam chr11:40185800-40195800 | cut -f1 | sort -u | wc -l"
      ],
      "metadata": {
        "id": "Nvg6mehhZwtR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "409393c6-ab9e-4695-80b5-9b231bba4d1f"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "797\n",
            "274\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CliRCSUQTmV-"
      },
      "source": [
        "#### Дедупликация"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! find /content/drive/MyDrive/bioinf -name \"*pe.bam\" | xargs -P 4 -tI{} deduplicate_bismark  --bam  --paired  -o /content/drive/MyDrive/bioinf/s_{} {}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1jtdMKegQ-Db",
        "outputId": "9c4b4d7e-b843-41b3-cbc3-cfa214e5a5be"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "deduplicate_bismark --bam --paired -o /content/drive/MyDrive/bioinf/s_/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam\n",
            "deduplicate_bismark --bam --paired -o /content/drive/MyDrive/bioinf/s_/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam\n",
            "deduplicate_bismark --bam --paired -o /content/drive/MyDrive/bioinf/s_/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam\n",
            "Output filename was given as: /content/drive/MyDrive/bioinf/s_/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam\n",
            "Processing paired-end Bismark output file(s) (SAM format):\n",
            "/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam\n",
            "\n",
            "Output filename was given as: /content/drive/MyDrive/bioinf/s_/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam\n",
            "Processing paired-end Bismark output file(s) (SAM format):\n",
            "/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam\n",
            "\n",
            "Output filename was given as: /content/drive/MyDrive/bioinf/s_/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam\n",
            "Processing paired-end Bismark output file(s) (SAM format):\n",
            "/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam\n",
            "\n",
            "\n",
            "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
            "\n",
            "Checking file >>/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
            "\n",
            "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
            "\n",
            "Checking file >>/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
            "\n",
            "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
            "\n",
            "Checking file >>/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "\n",
            "Now testing Bismark result file /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\t\n",
            "\n",
            "\n",
            "Now testing Bismark result file /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\t\n",
            "\n",
            "\n",
            "Now testing Bismark result file /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\t...passed!\n",
            "Output file is: SRR5836473_1_bismark_bt2_pe.deduplicated.bam\n",
            "\n",
            "skipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h --threads 1 /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam\n",
            "...passed!\n",
            "Output file is: SRR5836475_1_bismark_bt2_pe.deduplicated.bam\n",
            "\n",
            "...passed!\n",
            "Output file is: SRR3824222_1_bismark_bt2_pe.deduplicated.bam\n",
            "\n",
            "skipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR5836475_1.fq -2 SRR5836475_2.fq\"\n",
            "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h --threads 1 /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam\n",
            "skipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h --threads 1 /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam\n",
            "\n",
            "Total number of alignments analysed in /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam:\t2858170\n",
            "Total number duplicated alignments removed:\t522512 (18.28%)\n",
            "Duplicated alignments were found at:\t230677 different position(s)\n",
            "\n",
            "Total count of deduplicated leftover sequences: 2335658 (81.72% of total)\n",
            "\n",
            "\n",
            "Total number of alignments analysed in /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam:\t4170098\n",
            "Total number duplicated alignments removed:\t378125 (9.07%)\n",
            "Duplicated alignments were found at:\t229940 different position(s)\n",
            "\n",
            "Total count of deduplicated leftover sequences: 3791973 (90.93% of total)\n",
            "\n",
            "\n",
            "Total number of alignments analysed in /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam:\t7051731\n",
            "Total number duplicated alignments removed:\t205287 (2.91%)\n",
            "Duplicated alignments were found at:\t164998 different position(s)\n",
            "\n",
            "Total count of deduplicated leftover sequences: 6846444 (97.09% of total)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Сохраняю файлы на диск"
      ],
      "metadata": {
        "id": "DN4ri7GuzDnN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! mv SRR3824222_1_bismark_bt2_pe.deduplicated.bam SRR3824222_1_bismark_bt2_pe.deduplication_report.txt SRR5836473_1_bismark_bt2_pe.deduplicated.bam SRR5836473_1_bismark_bt2_pe.deduplication_report.txt SRR5836475_1_bismark_bt2_pe.deduplicated.bam SRR5836475_1_bismark_bt2_pe.deduplication_report.txt /content/drive/MyDrive/bioinf/"
      ],
      "metadata": {
        "id": "IulkeonJxqPQ"
      },
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### 11-ю хромосому мыши можно скачать с UCSC (сборка mm39). Скачиваю с ucsc и подготавливаю нужные данные с помощью bismark"
      ],
      "metadata": {
        "id": "969PlQMGwBO3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%cd /content/drive/MyDrive/bioinf\n",
        "! gunzip chr11.fa.gz"
      ],
      "metadata": {
        "id": "6GeHkoVJqxAO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%cd /content/drive/MyDrive/bioinf\n",
        "! mkdir mouse_chr\n",
        "! mv chr11.fa ./mouse_chr/"
      ],
      "metadata": {
        "id": "o9-_RzQ4r19p"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark_genome_preparation /content/drive/MyDrive/bioinf/mouse_chr/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "quIieGp4rpmE",
        "outputId": "a0ba772c-94c4-4033-abbc-72340906598d"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Writing bisulfite genomes out into a single MFA (multi FastA) file\n",
            "\n",
            "Bisulfite Genome Indexer version v0.24.2 (last modified: 19 May 2022)\n",
            "\n",
            "Step I - Prepare genome folders - completed\n",
            "\n",
            "\n",
            "\n",
            "Total number of conversions performed:\n",
            "C->T:\t26029159\n",
            "G->A:\t26024038\n",
            "\n",
            "Step II - Genome bisulfite conversions - completed\n",
            "\n",
            "\n",
            "Bismark Genome Preparation - Step III: Launching the Bowtie 2 indexer\n",
            "Please be aware that this process can - depending on genome size - take several hours!\n",
            "Settings:\n",
            "  Output files: \"BS_CT.*.bt2\"\n",
            "  Line rate: 6 (line is 64 bytes)\n",
            "  Lines per side: 1 (side is 64 bytes)\n",
            "  Offset rate: 4 (one in 16)\n",
            "  FTable chars: 10\n",
            "  Strings: unpacked\n",
            "  Max bucket size: default\n",
            "  Max bucket size, sqrt multiplier: default\n",
            "  Max bucket size, len divisor: 4\n",
            "  Difference-cover sample period: 1024\n",
            "  Endianness: little\n",
            "  Actual local endianness: little\n",
            "  Sanity checking: disabled\n",
            "  Assertions: disabled\n",
            "  Random seed: 0\n",
            "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
            "Input files DNA, FASTA:\n",
            "  genome_mfa.CT_conversion.fa\n",
            "Building a SMALL index\n",
            "Reading reference sizes\n",
            "Settings:\n",
            "  Output files: \"BS_GA.*.bt2\"\n",
            "  Line rate: 6 (line is 64 bytes)\n",
            "  Lines per side: 1 (side is 64 bytes)\n",
            "  Offset rate: 4 (one in 16)\n",
            "  FTable chars: 10\n",
            "  Strings: unpacked\n",
            "  Max bucket size: default\n",
            "  Max bucket size, sqrt multiplier: default\n",
            "  Max bucket size, len divisor: 4\n",
            "  Difference-cover sample period: 1024\n",
            "  Endianness: little\n",
            "  Actual local endianness: little\n",
            "  Sanity checking: disabled\n",
            "  Assertions: disabled\n",
            "  Random seed: 0\n",
            "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
            "Input files DNA, FASTA:\n",
            "  genome_mfa.GA_conversion.fa\n",
            "Building a SMALL index\n",
            "Reading reference sizes\n",
            "  Time reading reference sizes: 00:00:02\n",
            "Calculating joined length\n",
            "Writing header\n",
            "Reserving space for joined string\n",
            "Joining reference sequences\n",
            "  Time to join reference sequences: 00:00:01\n",
            "bmax according to bmaxDivN setting: 29705592\n",
            "Using parameters --bmax 22279194 --dcv 1024\n",
            "  Doing ahead-of-time memory usage test\n",
            "  Time reading reference sizes: 00:00:03\n",
            "Calculating joined length\n",
            "Writing header\n",
            "Reserving space for joined string\n",
            "Joining reference sequences\n",
            "  Passed!  Constructing with these parameters: --bmax 22279194 --dcv 1024\n",
            "Constructing suffix-array element generator\n",
            "Building DifferenceCoverSample\n",
            "  Building sPrime\n",
            "  Building sPrimeOrder\n",
            "  V-Sorting samples\n",
            "  Time to join reference sequences: 00:00:01\n",
            "bmax according to bmaxDivN setting: 29705592\n",
            "Using parameters --bmax 22279194 --dcv 1024\n",
            "  Doing ahead-of-time memory usage test\n",
            "  Passed!  Constructing with these parameters: --bmax 22279194 --dcv 1024\n",
            "Constructing suffix-array element generator\n",
            "Building DifferenceCoverSample\n",
            "  Building sPrime\n",
            "  Building sPrimeOrder\n",
            "  V-Sorting samples\n",
            "  V-Sorting samples time: 00:00:06\n",
            "  Allocating rank array\n",
            "  Ranking v-sort output\n",
            "  Ranking v-sort output time: 00:00:03\n",
            "  Invoking Larsson-Sadakane on ranks\n",
            "  V-Sorting samples time: 00:00:07\n",
            "  Allocating rank array\n",
            "  Ranking v-sort output\n",
            "  Ranking v-sort output time: 00:00:01\n",
            "  Invoking Larsson-Sadakane on ranks\n",
            "  Invoking Larsson-Sadakane on ranks time: 00:00:02\n",
            "  Sanity-checking and returning\n",
            "Building samples\n",
            "Reserving space for 12 sample suffixes\n",
            "Generating random suffixes\n",
            "QSorting 12 sample offsets, eliminating duplicates\n",
            "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
            "Multikey QSorting 12 samples\n",
            "  (Using difference cover)\n",
            "  Multikey QSorting samples time: 00:00:00\n",
            "Calculating bucket sizes\n",
            "  Invoking Larsson-Sadakane on ranks time: 00:00:02\n",
            "  Sanity-checking and returning\n",
            "Building samples\n",
            "Reserving space for 12 sample suffixes\n",
            "Generating random suffixes\n",
            "QSorting 12 sample offsets, eliminating duplicates\n",
            "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
            "Multikey QSorting 12 samples\n",
            "  (Using difference cover)\n",
            "  Multikey QSorting samples time: 00:00:00\n",
            "Calculating bucket sizes\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 2, merged 6; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Avg bucket size: 1.48528e+07 (target: 22279193)\n",
            "Converting suffix-array elements to index image\n",
            "Allocating ftab, absorbFtab\n",
            "Entering Ebwt loop\n",
            "Getting block 1 of 8\n",
            "  Reserving size (22279194) for bucket 1\n",
            "  Calculating Z arrays for bucket 1\n",
            "  Entering block accumulator loop for bucket 1:\n",
            "  bucket 1: 10%\n",
            "  bucket 1: 20%\n",
            "  bucket 1: 30%\n",
            "  bucket 1: 40%\n",
            "  bucket 1: 50%\n",
            "  bucket 1: 60%\n",
            "  bucket 1: 70%\n",
            "  bucket 1: 80%\n",
            "  bucket 1: 90%\n",
            "  bucket 1: 100%\n",
            "  Sorting block of length 16105101 for bucket 1\n",
            "  (Using difference cover)\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 2, merged 1; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 2, merged 2; iterating...\n",
            "  Sorting block time: 00:00:11\n",
            "Returning block of 16105102 for bucket 1\n",
            "Getting block 2 of 8\n",
            "  Reserving size (22279194) for bucket 2\n",
            "  Calculating Z arrays for bucket 2\n",
            "  Entering block accumulator loop for bucket 2:\n",
            "  bucket 2: 10%\n",
            "  bucket 2: 20%\n",
            "  bucket 2: 30%\n",
            "  bucket 2: 40%\n",
            "  bucket 2: 50%\n",
            "  bucket 2: 60%\n",
            "  bucket 2: 70%\n",
            "  bucket 2: 80%\n",
            "  bucket 2: 90%\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 2; iterating...\n",
            "  bucket 2: 100%\n",
            "  Sorting block of length 9511017 for bucket 2\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:07\n",
            "Returning block of 9511018 for bucket 2\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 1; iterating...\n",
            "Avg bucket size: 1.32025e+07 (target: 22279193)\n",
            "Converting suffix-array elements to index image\n",
            "Allocating ftab, absorbFtab\n",
            "Entering Ebwt loop\n",
            "Getting block 1 of 9\n",
            "  Reserving size (22279194) for bucket 1\n",
            "  Calculating Z arrays for bucket 1\n",
            "  Entering block accumulator loop for bucket 1:\n",
            "  bucket 1: 10%\n",
            "  bucket 1: 20%\n",
            "  bucket 1: 30%\n",
            "  bucket 1: 40%\n",
            "  bucket 1: 50%\n",
            "Getting block 3 of 8\n",
            "  Reserving size (22279194) for bucket 3\n",
            "  Calculating Z arrays for bucket 3\n",
            "  Entering block accumulator loop for bucket 3:\n",
            "  bucket 1: 60%\n",
            "  bucket 3: 10%\n",
            "  bucket 1: 70%\n",
            "  bucket 1: 80%\n",
            "  bucket 3: 20%\n",
            "  bucket 1: 90%\n",
            "  bucket 3: 30%\n",
            "  bucket 1: 100%\n",
            "  Sorting block of length 14872531 for bucket 1\n",
            "  (Using difference cover)\n",
            "  bucket 3: 40%\n",
            "  bucket 3: 50%\n",
            "  bucket 3: 60%\n",
            "  bucket 3: 70%\n",
            "  bucket 3: 80%\n",
            "  bucket 3: 90%\n",
            "  bucket 3: 100%\n",
            "  Sorting block of length 15942497 for bucket 3\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:14\n",
            "Returning block of 14872532 for bucket 1\n",
            "  Sorting block time: 00:00:13\n",
            "Returning block of 15942498 for bucket 3\n",
            "Getting block 2 of 9\n",
            "  Reserving size (22279194) for bucket 2\n",
            "  Calculating Z arrays for bucket 2\n",
            "  Entering block accumulator loop for bucket 2:\n",
            "  bucket 2: 10%\n",
            "  bucket 2: 20%\n",
            "  bucket 2: 30%\n",
            "  bucket 2: 40%\n",
            "  bucket 2: 50%\n",
            "  bucket 2: 60%\n",
            "Getting block 4 of 8\n",
            "  Reserving size (22279194) for bucket 4\n",
            "  Calculating Z arrays for bucket 4\n",
            "  Entering block accumulator loop for bucket 4:\n",
            "  bucket 2: 70%\n",
            "  bucket 4: 10%\n",
            "  bucket 2: 80%\n",
            "  bucket 4: 20%\n",
            "  bucket 2: 90%\n",
            "  bucket 4: 30%\n",
            "  bucket 2: 100%\n",
            "  Sorting block of length 22149652 for bucket 2\n",
            "  (Using difference cover)\n",
            "  bucket 4: 40%\n",
            "  bucket 4: 50%\n",
            "  bucket 4: 60%\n",
            "  bucket 4: 70%\n",
            "  bucket 4: 80%\n",
            "  bucket 4: 90%\n",
            "  bucket 4: 100%\n",
            "  Sorting block of length 20515027 for bucket 4\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:17\n",
            "Returning block of 20515028 for bucket 4\n",
            "  Sorting block time: 00:00:22\n",
            "Returning block of 22149653 for bucket 2\n",
            "Getting block 5 of 8\n",
            "  Reserving size (22279194) for bucket 5\n",
            "  Calculating Z arrays for bucket 5\n",
            "  Entering block accumulator loop for bucket 5:\n",
            "  bucket 5: 10%\n",
            "Getting block 3 of 9\n",
            "  Reserving size (22279194) for bucket 3\n",
            "  Calculating Z arrays for bucket 3\n",
            "  Entering block accumulator loop for bucket 3:\n",
            "  bucket 5: 20%\n",
            "  bucket 3: 10%\n",
            "  bucket 5: 30%\n",
            "  bucket 3: 20%\n",
            "  bucket 5: 40%\n",
            "  bucket 3: 30%\n",
            "  bucket 5: 50%\n",
            "  bucket 3: 40%\n",
            "  bucket 5: 60%\n",
            "  bucket 3: 50%\n",
            "  bucket 5: 70%\n",
            "  bucket 3: 60%\n",
            "  bucket 5: 80%\n",
            "  bucket 3: 70%\n",
            "  bucket 5: 90%\n",
            "  bucket 3: 80%\n",
            "  bucket 5: 100%\n",
            "  Sorting block of length 17789231 for bucket 5\n",
            "  (Using difference cover)\n",
            "  bucket 3: 90%\n",
            "  bucket 3: 100%\n",
            "  Sorting block of length 20254773 for bucket 3\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:14\n",
            "Returning block of 17789232 for bucket 5\n",
            "Getting block 6 of 8\n",
            "  Reserving size (22279194) for bucket 6\n",
            "  Calculating Z arrays for bucket 6\n",
            "  Entering block accumulator loop for bucket 6:\n",
            "  bucket 6: 10%\n",
            "  bucket 6: 20%\n",
            "  Sorting block time: 00:00:20\n",
            "Returning block of 20254774 for bucket 3\n",
            "  bucket 6: 30%\n",
            "  bucket 6: 40%\n",
            "  bucket 6: 50%\n",
            "  bucket 6: 60%\n",
            "  bucket 6: 70%\n",
            "  bucket 6: 80%\n",
            "  bucket 6: 90%\n",
            "  bucket 6: 100%\n",
            "  Sorting block of length 7043684 for bucket 6\n",
            "  (Using difference cover)\n",
            "Getting block 4 of 9\n",
            "  Reserving size (22279194) for bucket 4\n",
            "  Calculating Z arrays for bucket 4\n",
            "  Entering block accumulator loop for bucket 4:\n",
            "  bucket 4: 10%\n",
            "  bucket 4: 20%\n",
            "  bucket 4: 30%\n",
            "  bucket 4: 40%\n",
            "  bucket 4: 50%\n",
            "  bucket 4: 60%\n",
            "  bucket 4: 70%\n",
            "  Sorting block time: 00:00:05\n",
            "Returning block of 7043685 for bucket 6\n",
            "  bucket 4: 80%\n",
            "  bucket 4: 90%\n",
            "  bucket 4: 100%\n",
            "  Sorting block of length 8422961 for bucket 4\n",
            "  (Using difference cover)\n",
            "Getting block 7 of 8\n",
            "  Reserving size (22279194) for bucket 7\n",
            "  Calculating Z arrays for bucket 7\n",
            "  Entering block accumulator loop for bucket 7:\n",
            "  bucket 7: 10%\n",
            "  bucket 7: 20%\n",
            "  bucket 7: 30%\n",
            "  bucket 7: 40%\n",
            "  bucket 7: 50%\n",
            "  bucket 7: 60%\n",
            "  bucket 7: 70%\n",
            "  bucket 7: 80%\n",
            "  bucket 7: 90%\n",
            "  bucket 7: 100%\n",
            "  Sorting block of length 21444904 for bucket 7\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:08\n",
            "Returning block of 8422962 for bucket 4\n",
            "Getting block 5 of 9\n",
            "  Reserving size (22279194) for bucket 5\n",
            "  Calculating Z arrays for bucket 5\n",
            "  Entering block accumulator loop for bucket 5:\n",
            "  bucket 5: 10%\n",
            "  bucket 5: 20%\n",
            "  bucket 5: 30%\n",
            "  bucket 5: 40%\n",
            "  bucket 5: 50%\n",
            "  bucket 5: 60%\n",
            "  bucket 5: 70%\n",
            "  bucket 5: 80%\n",
            "  bucket 5: 90%\n",
            "  bucket 5: 100%\n",
            "  Sorting block of length 17366401 for bucket 5\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:17\n",
            "Returning block of 21444905 for bucket 7\n",
            "Getting block 8 of 8\n",
            "  Reserving size (22279194) for bucket 8\n",
            "  Calculating Z arrays for bucket 8\n",
            "  Entering block accumulator loop for bucket 8:\n",
            "  bucket 8: 10%\n",
            "  bucket 8: 20%\n",
            "  bucket 8: 30%\n",
            "  bucket 8: 40%\n",
            "  bucket 8: 50%\n",
            "  bucket 8: 60%\n",
            "  Sorting block time: 00:00:16\n",
            "Returning block of 17366402 for bucket 5\n",
            "  bucket 8: 70%\n",
            "  bucket 8: 80%\n",
            "  bucket 8: 90%\n",
            "  bucket 8: 100%\n",
            "  Sorting block of length 10470901 for bucket 8\n",
            "  (Using difference cover)\n",
            "Getting block 6 of 9\n",
            "  Reserving size (22279194) for bucket 6\n",
            "  Calculating Z arrays for bucket 6\n",
            "  Entering block accumulator loop for bucket 6:\n",
            "  bucket 6: 10%\n",
            "  bucket 6: 20%\n",
            "  bucket 6: 30%\n",
            "  bucket 6: 40%\n",
            "  bucket 6: 50%\n",
            "  bucket 6: 60%\n",
            "  Sorting block time: 00:00:07\n",
            "Returning block of 10470902 for bucket 8\n",
            "  bucket 6: 70%\n",
            "  bucket 6: 80%\n",
            "  bucket 6: 90%\n",
            "  bucket 6: 100%\n",
            "  Sorting block of length 5506757 for bucket 6\n",
            "  (Using difference cover)\n",
            "Exited Ebwt loop\n",
            "fchr[A]: 0\n",
            "fchr[C]: 59455040\n",
            "fchr[G]: 85484199\n",
            "fchr[T]: 85484199\n",
            "fchr[$]: 118822369\n",
            "Exiting Ebwt::buildToDisk()\n",
            "Returning from initFromVector\n",
            "Wrote 43801984 bytes to primary EBWT file: BS_GA.1.bt2.tmp\n",
            "Wrote 29705600 bytes to secondary EBWT file: BS_GA.2.bt2.tmp\n",
            "Re-opening _in1 and _in2 as input streams\n",
            "Returning from Ebwt constructor\n",
            "Headers:\n",
            "    len: 118822369\n",
            "    bwtLen: 118822370\n",
            "    sz: 29705593\n",
            "    bwtSz: 29705593\n",
            "    lineRate: 6\n",
            "    offRate: 4\n",
            "    offMask: 0xfffffff0\n",
            "    ftabChars: 10\n",
            "    eftabLen: 20\n",
            "    eftabSz: 80\n",
            "    ftabLen: 1048577\n",
            "    ftabSz: 4194308\n",
            "    offsLen: 7426399\n",
            "    offsSz: 29705596\n",
            "    lineSz: 64\n",
            "    sideSz: 64\n",
            "    sideBwtSz: 48\n",
            "    sideBwtLen: 192\n",
            "    numSides: 618867\n",
            "    numLines: 618867\n",
            "    ebwtTotLen: 39607488\n",
            "    ebwtTotSz: 39607488\n",
            "    color: 0\n",
            "    reverse: 0\n",
            "Total time for call to driver() for forward index: 00:03:05\n",
            "Reading reference sizes\n",
            "  Time reading reference sizes: 00:00:04\n",
            "Calculating joined length\n",
            "Writing header\n",
            "Reserving space for joined string\n",
            "Joining reference sequences\n",
            "  Time to join reference sequences: 00:00:02\n",
            "  Time to reverse reference sequence: 00:00:00\n",
            "bmax according to bmaxDivN setting: 29705592\n",
            "Using parameters --bmax 22279194 --dcv 1024\n",
            "  Doing ahead-of-time memory usage test\n",
            "  Passed!  Constructing with these parameters: --bmax 22279194 --dcv 1024\n",
            "Constructing suffix-array element generator\n",
            "Building DifferenceCoverSample\n",
            "  Building sPrime\n",
            "  Building sPrimeOrder\n",
            "  V-Sorting samples\n",
            "  Sorting block time: 00:00:08\n",
            "Returning block of 5506758 for bucket 6\n",
            "Getting block 7 of 9\n",
            "  Reserving size (22279194) for bucket 7\n",
            "  Calculating Z arrays for bucket 7\n",
            "  Entering block accumulator loop for bucket 7:\n",
            "  bucket 7: 10%\n",
            "  bucket 7: 20%\n",
            "  bucket 7: 30%\n",
            "  bucket 7: 40%\n",
            "  bucket 7: 50%\n",
            "  bucket 7: 60%\n",
            "  V-Sorting samples time: 00:00:05\n",
            "  Allocating rank array\n",
            "  Ranking v-sort output\n",
            "  bucket 7: 70%\n",
            "  bucket 7: 80%\n",
            "  bucket 7: 90%\n",
            "  Ranking v-sort output time: 00:00:02\n",
            "  Invoking Larsson-Sadakane on ranks\n",
            "  bucket 7: 100%\n",
            "  Sorting block of length 5313064 for bucket 7\n",
            "  (Using difference cover)\n",
            "  Invoking Larsson-Sadakane on ranks time: 00:00:02\n",
            "  Sanity-checking and returning\n",
            "Building samples\n",
            "Reserving space for 12 sample suffixes\n",
            "Generating random suffixes\n",
            "QSorting 12 sample offsets, eliminating duplicates\n",
            "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
            "Multikey QSorting 12 samples\n",
            "  (Using difference cover)\n",
            "  Multikey QSorting samples time: 00:00:00\n",
            "Calculating bucket sizes\n",
            "  Sorting block time: 00:00:06\n",
            "Returning block of 5313065 for bucket 7\n",
            "Getting block 8 of 9\n",
            "  Reserving size (22279194) for bucket 8\n",
            "  Calculating Z arrays for bucket 8\n",
            "  Entering block accumulator loop for bucket 8:\n",
            "  bucket 8: 10%\n",
            "  bucket 8: 20%\n",
            "  bucket 8: 30%\n",
            "  bucket 8: 40%\n",
            "  bucket 8: 50%\n",
            "  bucket 8: 60%\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 7; iterating...\n",
            "  bucket 8: 70%\n",
            "  bucket 8: 80%\n",
            "  bucket 8: 90%\n",
            "  bucket 8: 100%\n",
            "  Sorting block of length 20667562 for bucket 8\n",
            "  (Using difference cover)\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 1; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 1; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Avg bucket size: 1.69746e+07 (target: 22279193)\n",
            "Converting suffix-array elements to index image\n",
            "Allocating ftab, absorbFtab\n",
            "Entering Ebwt loop\n",
            "Getting block 1 of 7\n",
            "  Reserving size (22279194) for bucket 1\n",
            "  Calculating Z arrays for bucket 1\n",
            "  Entering block accumulator loop for bucket 1:\n",
            "  bucket 1: 10%\n",
            "  bucket 1: 20%\n",
            "  bucket 1: 30%\n",
            "  Sorting block time: 00:00:18\n",
            "Returning block of 20667563 for bucket 8\n",
            "  bucket 1: 40%\n",
            "  bucket 1: 50%\n",
            "  bucket 1: 60%\n",
            "  bucket 1: 70%\n",
            "  bucket 1: 80%\n",
            "  bucket 1: 90%\n",
            "  bucket 1: 100%\n",
            "  Sorting block of length 14924270 for bucket 1\n",
            "  (Using difference cover)\n",
            "Getting block 9 of 9\n",
            "  Reserving size (22279194) for bucket 9\n",
            "  Calculating Z arrays for bucket 9\n",
            "  Entering block accumulator loop for bucket 9:\n",
            "  bucket 9: 10%\n",
            "  bucket 9: 20%\n",
            "  bucket 9: 30%\n",
            "  bucket 9: 40%\n",
            "  bucket 9: 50%\n",
            "  bucket 9: 60%\n",
            "  bucket 9: 70%\n",
            "  bucket 9: 80%\n",
            "  bucket 9: 90%\n",
            "  bucket 9: 100%\n",
            "  Sorting block of length 4268660 for bucket 9\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:03\n",
            "Returning block of 4268661 for bucket 9\n",
            "Exited Ebwt loop\n",
            "fchr[A]: 0\n",
            "fchr[C]: 33431002\n",
            "fchr[G]: 33431002\n",
            "fchr[T]: 59455040\n",
            "fchr[$]: 118822369\n",
            "Exiting Ebwt::buildToDisk()\n",
            "Returning from initFromVector\n",
            "Wrote 43801984 bytes to primary EBWT file: BS_CT.1.bt2.tmp\n",
            "Wrote 29705600 bytes to secondary EBWT file: BS_CT.2.bt2.tmp\n",
            "Re-opening _in1 and _in2 as input streams\n",
            "Returning from Ebwt constructor\n",
            "Headers:\n",
            "    len: 118822369\n",
            "    bwtLen: 118822370\n",
            "    sz: 29705593\n",
            "    bwtSz: 29705593\n",
            "    lineRate: 6\n",
            "    offRate: 4\n",
            "    offMask: 0xfffffff0\n",
            "    ftabChars: 10\n",
            "    eftabLen: 20\n",
            "    eftabSz: 80\n",
            "    ftabLen: 1048577\n",
            "    ftabSz: 4194308\n",
            "    offsLen: 7426399\n",
            "    offsSz: 29705596\n",
            "    lineSz: 64\n",
            "    sideSz: 64\n",
            "    sideBwtSz: 48\n",
            "    sideBwtLen: 192\n",
            "    numSides: 618867\n",
            "    numLines: 618867\n",
            "    ebwtTotLen: 39607488\n",
            "    ebwtTotSz: 39607488\n",
            "    color: 0\n",
            "    reverse: 0\n",
            "Total time for call to driver() for forward index: 00:04:04\n",
            "Reading reference sizes\n",
            "  Sorting block time: 00:00:12\n",
            "Returning block of 14924271 for bucket 1\n",
            "  Time reading reference sizes: 00:00:02\n",
            "Calculating joined length\n",
            "Writing header\n",
            "Reserving space for joined string\n",
            "Joining reference sequences\n",
            "  Time to join reference sequences: 00:00:03\n",
            "  Time to reverse reference sequence: 00:00:00\n",
            "bmax according to bmaxDivN setting: 29705592\n",
            "Using parameters --bmax 22279194 --dcv 1024\n",
            "  Doing ahead-of-time memory usage test\n",
            "Getting block 2 of 7\n",
            "  Reserving size (22279194) for bucket 2\n",
            "  Calculating Z arrays for bucket 2\n",
            "  Entering block accumulator loop for bucket 2:\n",
            "  Passed!  Constructing with these parameters: --bmax 22279194 --dcv 1024\n",
            "Constructing suffix-array element generator\n",
            "Building DifferenceCoverSample\n",
            "  Building sPrime\n",
            "  Building sPrimeOrder\n",
            "  V-Sorting samples\n",
            "  bucket 2: 10%\n",
            "  bucket 2: 20%\n",
            "  bucket 2: 30%\n",
            "  bucket 2: 40%\n",
            "  bucket 2: 50%\n",
            "  bucket 2: 60%\n",
            "  bucket 2: 70%\n",
            "  bucket 2: 80%\n",
            "  bucket 2: 90%\n",
            "  bucket 2: 100%\n",
            "  Sorting block of length 9944205 for bucket 2\n",
            "  (Using difference cover)\n",
            "  V-Sorting samples time: 00:00:05\n",
            "  Allocating rank array\n",
            "  Ranking v-sort output\n",
            "  Ranking v-sort output time: 00:00:02\n",
            "  Invoking Larsson-Sadakane on ranks\n",
            "  Invoking Larsson-Sadakane on ranks time: 00:00:01\n",
            "  Sanity-checking and returning\n",
            "Building samples\n",
            "Reserving space for 12 sample suffixes\n",
            "Generating random suffixes\n",
            "QSorting 12 sample offsets, eliminating duplicates\n",
            "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
            "Multikey QSorting 12 samples\n",
            "  (Using difference cover)\n",
            "  Multikey QSorting samples time: 00:00:00\n",
            "Calculating bucket sizes\n",
            "  Sorting block time: 00:00:07\n",
            "Returning block of 9944206 for bucket 2\n",
            "Getting block 3 of 7\n",
            "  Reserving size (22279194) for bucket 3\n",
            "  Calculating Z arrays for bucket 3\n",
            "  Entering block accumulator loop for bucket 3:\n",
            "  bucket 3: 10%\n",
            "  bucket 3: 20%\n",
            "  bucket 3: 30%\n",
            "  bucket 3: 40%\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 7; iterating...\n",
            "  bucket 3: 50%\n",
            "  bucket 3: 60%\n",
            "  bucket 3: 70%\n",
            "  bucket 3: 80%\n",
            "  bucket 3: 90%\n",
            "  bucket 3: 100%\n",
            "  Sorting block of length 19992172 for bucket 3\n",
            "  (Using difference cover)\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 1; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 0; iterating...\n",
            "  Sorting block time: 00:00:16\n",
            "Returning block of 19992173 for bucket 3\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Avg bucket size: 1.69746e+07 (target: 22279193)\n",
            "Converting suffix-array elements to index image\n",
            "Allocating ftab, absorbFtab\n",
            "Entering Ebwt loop\n",
            "Getting block 1 of 7\n",
            "  Reserving size (22279194) for bucket 1\n",
            "  Calculating Z arrays for bucket 1\n",
            "  Entering block accumulator loop for bucket 1:\n",
            "  bucket 1: 10%\n",
            "  bucket 1: 20%\n",
            "  bucket 1: 30%\n",
            "  bucket 1: 40%\n",
            "  bucket 1: 50%\n",
            "  bucket 1: 60%\n",
            "  bucket 1: 70%\n",
            "  bucket 1: 80%\n",
            "  bucket 1: 90%\n",
            "  bucket 1: 100%\n",
            "  Sorting block of length 16138988 for bucket 1\n",
            "  (Using difference cover)\n",
            "Getting block 4 of 7\n",
            "  Reserving size (22279194) for bucket 4\n",
            "  Calculating Z arrays for bucket 4\n",
            "  Entering block accumulator loop for bucket 4:\n",
            "  bucket 4: 10%\n",
            "  bucket 4: 20%\n",
            "  bucket 4: 30%\n",
            "  bucket 4: 40%\n",
            "  bucket 4: 50%\n",
            "  bucket 4: 60%\n",
            "  bucket 4: 70%\n",
            "  bucket 4: 80%\n",
            "  bucket 4: 90%\n",
            "  bucket 4: 100%\n",
            "  Sorting block of length 12757883 for bucket 4\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:15\n",
            "Returning block of 16138989 for bucket 1\n",
            "  Sorting block time: 00:00:10\n",
            "Returning block of 12757884 for bucket 4\n",
            "Getting block 2 of 7\n",
            "  Reserving size (22279194) for bucket 2\n",
            "  Calculating Z arrays for bucket 2\n",
            "  Entering block accumulator loop for bucket 2:\n",
            "  bucket 2: 10%\n",
            "  bucket 2: 20%\n",
            "  bucket 2: 30%\n",
            "  bucket 2: 40%\n",
            "  bucket 2: 50%\n",
            "Getting block 5 of 7\n",
            "  Reserving size (22279194) for bucket 5\n",
            "  Calculating Z arrays for bucket 5\n",
            "  Entering block accumulator loop for bucket 5:\n",
            "  bucket 2: 60%\n",
            "  bucket 5: 10%\n",
            "  bucket 2: 70%\n",
            "  bucket 5: 20%\n",
            "  bucket 2: 80%\n",
            "  bucket 5: 30%\n",
            "  bucket 2: 90%\n",
            "  bucket 5: 40%\n",
            "  bucket 2: 100%\n",
            "  Sorting block of length 15591704 for bucket 2\n",
            "  (Using difference cover)\n",
            "  bucket 5: 50%\n",
            "  bucket 5: 60%\n",
            "  bucket 5: 70%\n",
            "  bucket 5: 80%\n",
            "  bucket 5: 90%\n",
            "  bucket 5: 100%\n",
            "  Sorting block of length 20338591 for bucket 5\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:16\n",
            "Returning block of 15591705 for bucket 2\n",
            "Getting block 3 of 7\n",
            "  Reserving size (22279194) for bucket 3\n",
            "  Calculating Z arrays for bucket 3\n",
            "  Entering block accumulator loop for bucket 3:\n",
            "  bucket 3: 10%\n",
            "  bucket 3: 20%\n",
            "  bucket 3: 30%\n",
            "  bucket 3: 40%\n",
            "  Sorting block time: 00:00:21\n",
            "Returning block of 20338592 for bucket 5\n",
            "  bucket 3: 50%\n",
            "  bucket 3: 60%\n",
            "  bucket 3: 70%\n",
            "  bucket 3: 80%\n",
            "  bucket 3: 90%\n",
            "  bucket 3: 100%\n",
            "  Sorting block of length 20385288 for bucket 3\n",
            "  (Using difference cover)\n",
            "Getting block 6 of 7\n",
            "  Reserving size (22279194) for bucket 6\n",
            "  Calculating Z arrays for bucket 6\n",
            "  Entering block accumulator loop for bucket 6:\n",
            "  bucket 6: 10%\n",
            "  bucket 6: 20%\n",
            "  bucket 6: 30%\n",
            "  bucket 6: 40%\n",
            "  bucket 6: 50%\n",
            "  bucket 6: 60%\n",
            "  bucket 6: 70%\n",
            "  bucket 6: 80%\n",
            "  bucket 6: 90%\n",
            "  bucket 6: 100%\n",
            "  Sorting block of length 19054512 for bucket 6\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:19\n",
            "Returning block of 20385289 for bucket 3\n",
            "Getting block 4 of 7\n",
            "  Reserving size (22279194) for bucket 4\n",
            "  Calculating Z arrays for bucket 4\n",
            "  Entering block accumulator loop for bucket 4:\n",
            "  bucket 4: 10%\n",
            "  bucket 4: 20%\n",
            "  Sorting block time: 00:00:18\n",
            "Returning block of 19054513 for bucket 6\n",
            "  bucket 4: 30%\n",
            "  bucket 4: 40%\n",
            "  bucket 4: 50%\n",
            "  bucket 4: 60%\n",
            "  bucket 4: 70%\n",
            "  bucket 4: 80%\n",
            "  bucket 4: 90%\n",
            "  bucket 4: 100%\n",
            "  Sorting block of length 17447222 for bucket 4\n",
            "  (Using difference cover)\n",
            "Getting block 7 of 7\n",
            "  Reserving size (22279194) for bucket 7\n",
            "  Calculating Z arrays for bucket 7\n",
            "  Entering block accumulator loop for bucket 7:\n",
            "  bucket 7: 10%\n",
            "  bucket 7: 20%\n",
            "  bucket 7: 30%\n",
            "  bucket 7: 40%\n",
            "  bucket 7: 50%\n",
            "  bucket 7: 60%\n",
            "  bucket 7: 70%\n",
            "  bucket 7: 80%\n",
            "  bucket 7: 90%\n",
            "  bucket 7: 100%\n",
            "  Sorting block of length 21810730 for bucket 7\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:16\n",
            "Returning block of 17447223 for bucket 4\n",
            "Getting block 5 of 7\n",
            "  Reserving size (22279194) for bucket 5\n",
            "  Calculating Z arrays for bucket 5\n",
            "  Entering block accumulator loop for bucket 5:\n",
            "  bucket 5: 10%\n",
            "  bucket 5: 20%\n",
            "  bucket 5: 30%\n",
            "  bucket 5: 40%\n",
            "  bucket 5: 50%\n",
            "  bucket 5: 60%\n",
            "  Sorting block time: 00:00:19\n",
            "Returning block of 21810731 for bucket 7\n",
            "  bucket 5: 70%\n",
            "  bucket 5: 80%\n",
            "  bucket 5: 90%\n",
            "  bucket 5: 100%\n",
            "  Sorting block of length 16483437 for bucket 5\n",
            "  (Using difference cover)\n",
            "Exited Ebwt loop\n",
            "fchr[A]: 0\n",
            "fchr[C]: 59455040\n",
            "fchr[G]: 85484199\n",
            "fchr[T]: 85484199\n",
            "fchr[$]: 118822369\n",
            "Exiting Ebwt::buildToDisk()\n",
            "Returning from initFromVector\n",
            "Wrote 43801984 bytes to primary EBWT file: BS_GA.rev.1.bt2.tmp\n",
            "Wrote 29705600 bytes to secondary EBWT file: BS_GA.rev.2.bt2.tmp\n",
            "Re-opening _in1 and _in2 as input streams\n",
            "Returning from Ebwt constructor\n",
            "Headers:\n",
            "    len: 118822369\n",
            "    bwtLen: 118822370\n",
            "    sz: 29705593\n",
            "    bwtSz: 29705593\n",
            "    lineRate: 6\n",
            "    offRate: 4\n",
            "    offMask: 0xfffffff0\n",
            "    ftabChars: 10\n",
            "    eftabLen: 20\n",
            "    eftabSz: 80\n",
            "    ftabLen: 1048577\n",
            "    ftabSz: 4194308\n",
            "    offsLen: 7426399\n",
            "    offsSz: 29705596\n",
            "    lineSz: 64\n",
            "    sideSz: 64\n",
            "    sideBwtSz: 48\n",
            "    sideBwtLen: 192\n",
            "    numSides: 618867\n",
            "    numLines: 618867\n",
            "    ebwtTotLen: 39607488\n",
            "    ebwtTotSz: 39607488\n",
            "    color: 0\n",
            "    reverse: 1\n",
            "Total time for backward call to driver() for mirror index: 00:03:33\n",
            "Renaming BS_GA.3.bt2.tmp to BS_GA.3.bt2\n",
            "Renaming BS_GA.4.bt2.tmp to BS_GA.4.bt2\n",
            "Renaming BS_GA.1.bt2.tmp to BS_GA.1.bt2\n",
            "Renaming BS_GA.2.bt2.tmp to BS_GA.2.bt2\n",
            "Renaming BS_GA.rev.1.bt2.tmp to BS_GA.rev.1.bt2\n",
            "Renaming BS_GA.rev.2.bt2.tmp to BS_GA.rev.2.bt2\n",
            "  Sorting block time: 00:00:12\n",
            "Returning block of 16483438 for bucket 5\n",
            "Getting block 6 of 7\n",
            "  Reserving size (22279194) for bucket 6\n",
            "  Calculating Z arrays for bucket 6\n",
            "  Entering block accumulator loop for bucket 6:\n",
            "  bucket 6: 10%\n",
            "  bucket 6: 20%\n",
            "  bucket 6: 30%\n",
            "  bucket 6: 40%\n",
            "  bucket 6: 50%\n",
            "  bucket 6: 60%\n",
            "  bucket 6: 70%\n",
            "  bucket 6: 80%\n",
            "  bucket 6: 90%\n",
            "  bucket 6: 100%\n",
            "  Sorting block of length 19416568 for bucket 6\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:15\n",
            "Returning block of 19416569 for bucket 6\n",
            "Getting block 7 of 7\n",
            "  Reserving size (22279194) for bucket 7\n",
            "  Calculating Z arrays for bucket 7\n",
            "  Entering block accumulator loop for bucket 7:\n",
            "  bucket 7: 10%\n",
            "  bucket 7: 20%\n",
            "  bucket 7: 30%\n",
            "  bucket 7: 40%\n",
            "  bucket 7: 50%\n",
            "  bucket 7: 60%\n",
            "  bucket 7: 70%\n",
            "  bucket 7: 80%\n",
            "  bucket 7: 90%\n",
            "  bucket 7: 100%\n",
            "  Sorting block of length 13359156 for bucket 7\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:10\n",
            "Returning block of 13359157 for bucket 7\n",
            "Exited Ebwt loop\n",
            "fchr[A]: 0\n",
            "fchr[C]: 33431002\n",
            "fchr[G]: 33431002\n",
            "fchr[T]: 59455040\n",
            "fchr[$]: 118822369\n",
            "Exiting Ebwt::buildToDisk()\n",
            "Returning from initFromVector\n",
            "Wrote 43801984 bytes to primary EBWT file: BS_CT.rev.1.bt2.tmp\n",
            "Wrote 29705600 bytes to secondary EBWT file: BS_CT.rev.2.bt2.tmp\n",
            "Re-opening _in1 and _in2 as input streams\n",
            "Returning from Ebwt constructor\n",
            "Headers:\n",
            "    len: 118822369\n",
            "    bwtLen: 118822370\n",
            "    sz: 29705593\n",
            "    bwtSz: 29705593\n",
            "    lineRate: 6\n",
            "    offRate: 4\n",
            "    offMask: 0xfffffff0\n",
            "    ftabChars: 10\n",
            "    eftabLen: 20\n",
            "    eftabSz: 80\n",
            "    ftabLen: 1048577\n",
            "    ftabSz: 4194308\n",
            "    offsLen: 7426399\n",
            "    offsSz: 29705596\n",
            "    lineSz: 64\n",
            "    sideSz: 64\n",
            "    sideBwtSz: 48\n",
            "    sideBwtLen: 192\n",
            "    numSides: 618867\n",
            "    numLines: 618867\n",
            "    ebwtTotLen: 39607488\n",
            "    ebwtTotSz: 39607488\n",
            "    color: 0\n",
            "    reverse: 1\n",
            "Total time for backward call to driver() for mirror index: 00:03:25\n",
            "Renaming BS_CT.3.bt2.tmp to BS_CT.3.bt2\n",
            "Renaming BS_CT.4.bt2.tmp to BS_CT.4.bt2\n",
            "Renaming BS_CT.1.bt2.tmp to BS_CT.1.bt2\n",
            "Renaming BS_CT.2.bt2.tmp to BS_CT.2.bt2\n",
            "Renaming BS_CT.rev.1.bt2.tmp to BS_CT.rev.1.bt2\n",
            "Renaming BS_CT.rev.2.bt2.tmp to BS_CT.rev.2.bt2\n",
            "\n",
            "=========================================\n",
            "\n",
            "Parallel genome indexing complete. Enjoy!\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Коллинг метилирования"
      ],
      "metadata": {
        "id": "LKBD3cwo1wLx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark_methylation_extractor --gzip --bedGraph --cytosine_report --genome_folder /content/drive/MyDrive/bioinf/mouse_chr /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OU_B6dy5koG2",
        "outputId": "f07cf5c4-0838-40b3-b4f9-8077228a82cd"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " *** Bismark methylation extractor version v0.24.2 ***\n",
            "\n",
            "Trying to determine the type of mapping from the SAM header line of file /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bam\n",
            "Treating file(s) as paired-end data (as extracted from @PG line)\n",
            "\n",
            "Setting option '--no_overlap' since this is (normally) the right thing to do for paired-end data\n",
            "\n",
            "\n",
            "Setting core usage to single-threaded (default). Consider using --multicore <int> to speed up the extraction process.\n",
            "\n",
            "Summarising Bismark methylation extractor parameters:\n",
            "===============================================================\n",
            "Bismark paired-end SAM format specified (default)\n",
            "Number of cores to be used: 1\n",
            "Output will be written to the current directory ('/content/drive/MyDrive/bioinf')\n",
            "\n",
            "\n",
            "Summarising bedGraph parameters:\n",
            "===============================================================\n",
            "Generating additional output in bedGraph and coverage format\n",
            "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
            "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
            "\n",
            "Using a cutoff of 1 read(s) to report cytosine positions\n",
            "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
            "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
            "\n",
            "\n",
            "Summarising genome-wide cytosine methylation report parameters:\n",
            "===============================================================\n",
            "Generating comprehensive genome-wide cytosine report\n",
            "(output format: <Chromosome> <Position> <Strand> <count methylated> <count non-methylated>  <C-context>  <trinucleotide context> )\n",
            "Reporting cytosine methylation in CpG context only (default)\n",
            "Using 1-based genomic coordinates (default)\n",
            "Genome folder was specified as /content/drive/MyDrive/bioinf/mouse_chr/\n",
            "\n",
            "Checking file >>/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "Now testing Bismark result file >/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bam< for positional sorting (which would be bad...)\t...passed!\n",
            "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Now reading in Bismark result file /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bam\n",
            "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping SAM header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h --threads 1 /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.bam\n",
            "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.19.2\tCL:/usr/local/bin/samtools view -bS --threads 1 -\n",
            "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bam\n",
            "Processed lines: 500000\n",
            "Processed lines: 1000000\n",
            "Processed lines: 1500000\n",
            "Processed lines: 2000000\n",
            "Processed lines: 2500000\n",
            "Processed lines: 3000000\n",
            "Processed lines: 3500000\n",
            "Processed lines: 4000000\n",
            "Processed lines: 4500000\n",
            "Processed lines: 5000000\n",
            "Processed lines: 5500000\n",
            "Processed lines: 6000000\n",
            "Processed lines: 6500000\n",
            "\n",
            "Processed 6846444 lines in total\n",
            "Total number of methylation call strings processed: 13692888\n",
            "\n",
            "Final Cytosine Methylation Report\n",
            "=================================\n",
            "Total number of C's analysed:\t331429490\n",
            "\n",
            "Total methylated C's in CpG context:\t11550199\n",
            "Total methylated C's in CHG context:\t1574513\n",
            "Total methylated C's in CHH context:\t4133143\n",
            "\n",
            "Total C to T conversions in CpG context:\t3463972\n",
            "Total C to T conversions in CHG context:\t71924830\n",
            "Total C to T conversions in CHH context:\t238782833\n",
            "\n",
            "C methylated in CpG context:\t76.9%\n",
            "C methylated in CHG context:\t2.1%\n",
            "C methylated in CHH context:\t1.7%\n",
            "\n",
            "\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Deleting unused files ...\n",
            "\n",
            "CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "\n",
            "\n",
            "Using these input files: CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CpG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CpG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CpG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHH_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHH_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Summary of parameters for bismark2bedGraph conversion:\n",
            "======================================================\n",
            "bedGraph output:\t\tSRR3824222_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "output directory:\t\t><\n",
            "remove whitespaces:\t\tno\n",
            "CX context:\t\t\tno (CpG context only, default)\n",
            "No-header selected:\t\tno\n",
            "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
            "Sort buffer size:\t\t2G\n",
            "Coverage threshold:\t\t1\n",
            "=============================================================================\n",
            "Methylation information will now be written into a bedGraph and coverage file\n",
            "=============================================================================\n",
            "\n",
            "Using the following files as Input:\n",
            "/content/drive/MyDrive/bioinf/CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing bedGraph to file: SRR3824222_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "Also writing out a coverage file including counts methylated and unmethylated residues to file: SRR3824222_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "\n",
            "Now writing methylation information for file >>CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Collecting temporary chromosome file information... Processing the following input file(s):\n",
            "CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Sorting input file CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp by positions (using -S of 2G)\n",
            "Successfully deleted the temporary input file CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Finished BedGraph conversion ...\n",
            "\n",
            "Summary of parameters for genome-wide cytosine report:\n",
            "==============================================================================\n",
            "Coverage infile:\t\t\t\tSRR3824222_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "Output directory:\t\t\t\t><\n",
            "Parent directory:\t\t\t\t>/content/drive/MyDrive/bioinf/<\n",
            "Genome directory:\t\t\t\t>/content/drive/MyDrive/bioinf/mouse_chr/<\n",
            "Coverage threshold\t\t\t\t0 (all positions will be reported, default)\n",
            "CX context:\t\t\t\t\tno (CpG context only, default)\n",
            "Genome coordinates used:\t\t\t1-based (default)\n",
            "GZIP compression:\t\t\t\tyes\n",
            "Split by chromosome:\t\t\t\tno\n",
            "\n",
            "\n",
            "Now reading in and storing sequence information of the genome specified in: /content/drive/MyDrive/bioinf/mouse_chr/\n",
            "\n",
            "chr chr11 (121973369 bp)\n",
            "\n",
            "Stored sequence information of 1 chromosomes/scaffolds in total\n",
            "\n",
            "==============================================================================\n",
            "Methylation information will now be written into a genome-wide cytosine report\n",
            "==============================================================================\n",
            "\n",
            "Adding context-specific methylation summaries\n",
            "\n",
            ">>> Writing genome-wide cytosine report to: SRR3824222_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz <<<\n",
            "\n",
            ">>> Writing all cytosine context summary file to: SRR3824222_1_bismark_bt2_pe.deduplicated.cytosine_context_summary.txt <<<\n",
            "\n",
            "Storing all covered cytosine positions for chromosome: chr11\n",
            "Writing cytosine report for last chromosome chr11 (stored 2250540 different covered positions)\n",
            "Finished writing out cytosine report for covered chromosomes (processed 1 chromosomes/scaffolds in total)\n",
            "\n",
            "Now processing chromosomes that were not covered by any methylation calls in the coverage file...\n",
            "All chromosomes in the genome were covered by at least some reads. coverage2cytosine processing complete.\n",
            "\n",
            "\n",
            "\n",
            "Finished generating genome-wide cytosine report\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark_methylation_extractor --gzip --bedGraph --cytosine_report --genome_folder /content/drive/MyDrive/bioinf/mouse_chr /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SXofnVGaurR2",
        "outputId": "8099a61a-3842-4119-e966-4ea6832473c9"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " *** Bismark methylation extractor version v0.24.2 ***\n",
            "\n",
            "Trying to determine the type of mapping from the SAM header line of file /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bam\n",
            "Treating file(s) as paired-end data (as extracted from @PG line)\n",
            "\n",
            "Setting option '--no_overlap' since this is (normally) the right thing to do for paired-end data\n",
            "\n",
            "\n",
            "Setting core usage to single-threaded (default). Consider using --multicore <int> to speed up the extraction process.\n",
            "\n",
            "Summarising Bismark methylation extractor parameters:\n",
            "===============================================================\n",
            "Bismark paired-end SAM format specified (default)\n",
            "Number of cores to be used: 1\n",
            "Output will be written to the current directory ('/content/drive/MyDrive/bioinf')\n",
            "\n",
            "\n",
            "Summarising bedGraph parameters:\n",
            "===============================================================\n",
            "Generating additional output in bedGraph and coverage format\n",
            "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
            "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
            "\n",
            "Using a cutoff of 1 read(s) to report cytosine positions\n",
            "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
            "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
            "\n",
            "\n",
            "Summarising genome-wide cytosine methylation report parameters:\n",
            "===============================================================\n",
            "Generating comprehensive genome-wide cytosine report\n",
            "(output format: <Chromosome> <Position> <Strand> <count methylated> <count non-methylated>  <C-context>  <trinucleotide context> )\n",
            "Reporting cytosine methylation in CpG context only (default)\n",
            "Using 1-based genomic coordinates (default)\n",
            "Genome folder was specified as /content/drive/MyDrive/bioinf/mouse_chr/\n",
            "\n",
            "Checking file >>/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "Now testing Bismark result file >/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bam< for positional sorting (which would be bad...)\t...passed!\n",
            "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Now reading in Bismark result file /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bam\n",
            "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping SAM header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h --threads 1 /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.bam\n",
            "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.19.2\tCL:/usr/local/bin/samtools view -bS --threads 1 -\n",
            "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bam\n",
            "Processed lines: 500000\n",
            "Processed lines: 1000000\n",
            "Processed lines: 1500000\n",
            "Processed lines: 2000000\n",
            "\n",
            "Processed 2335658 lines in total\n",
            "Total number of methylation call strings processed: 4671316\n",
            "\n",
            "Final Cytosine Methylation Report\n",
            "=================================\n",
            "Total number of C's analysed:\t104748620\n",
            "\n",
            "Total methylated C's in CpG context:\t2103916\n",
            "Total methylated C's in CHG context:\t270195\n",
            "Total methylated C's in CHH context:\t1236285\n",
            "\n",
            "Total C to T conversions in CpG context:\t2729982\n",
            "Total C to T conversions in CHG context:\t22951565\n",
            "Total C to T conversions in CHH context:\t75456677\n",
            "\n",
            "C methylated in CpG context:\t43.5%\n",
            "C methylated in CHG context:\t1.2%\n",
            "C methylated in CHH context:\t1.6%\n",
            "\n",
            "\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Deleting unused files ...\n",
            "\n",
            "CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "\n",
            "\n",
            "Using these input files: CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CpG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CpG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CpG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHH_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHH_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Summary of parameters for bismark2bedGraph conversion:\n",
            "======================================================\n",
            "bedGraph output:\t\tSRR5836473_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "output directory:\t\t><\n",
            "remove whitespaces:\t\tno\n",
            "CX context:\t\t\tno (CpG context only, default)\n",
            "No-header selected:\t\tno\n",
            "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
            "Sort buffer size:\t\t2G\n",
            "Coverage threshold:\t\t1\n",
            "=============================================================================\n",
            "Methylation information will now be written into a bedGraph and coverage file\n",
            "=============================================================================\n",
            "\n",
            "Using the following files as Input:\n",
            "/content/drive/MyDrive/bioinf/CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing bedGraph to file: SRR5836473_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "Also writing out a coverage file including counts methylated and unmethylated residues to file: SRR5836473_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "\n",
            "Now writing methylation information for file >>CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Collecting temporary chromosome file information... Processing the following input file(s):\n",
            "CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Sorting input file CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp by positions (using -S of 2G)\n",
            "Successfully deleted the temporary input file CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Finished BedGraph conversion ...\n",
            "\n",
            "Summary of parameters for genome-wide cytosine report:\n",
            "==============================================================================\n",
            "Coverage infile:\t\t\t\tSRR5836473_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "Output directory:\t\t\t\t><\n",
            "Parent directory:\t\t\t\t>/content/drive/MyDrive/bioinf/<\n",
            "Genome directory:\t\t\t\t>/content/drive/MyDrive/bioinf/mouse_chr/<\n",
            "Coverage threshold\t\t\t\t0 (all positions will be reported, default)\n",
            "CX context:\t\t\t\t\tno (CpG context only, default)\n",
            "Genome coordinates used:\t\t\t1-based (default)\n",
            "GZIP compression:\t\t\t\tyes\n",
            "Split by chromosome:\t\t\t\tno\n",
            "\n",
            "\n",
            "Now reading in and storing sequence information of the genome specified in: /content/drive/MyDrive/bioinf/mouse_chr/\n",
            "\n",
            "chr chr11 (121973369 bp)\n",
            "\n",
            "Stored sequence information of 1 chromosomes/scaffolds in total\n",
            "\n",
            "==============================================================================\n",
            "Methylation information will now be written into a genome-wide cytosine report\n",
            "==============================================================================\n",
            "\n",
            "Adding context-specific methylation summaries\n",
            "\n",
            ">>> Writing genome-wide cytosine report to: SRR5836473_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz <<<\n",
            "\n",
            ">>> Writing all cytosine context summary file to: SRR5836473_1_bismark_bt2_pe.deduplicated.cytosine_context_summary.txt <<<\n",
            "\n",
            "Storing all covered cytosine positions for chromosome: chr11\n",
            "Writing cytosine report for last chromosome chr11 (stored 1737533 different covered positions)\n",
            "Finished writing out cytosine report for covered chromosomes (processed 1 chromosomes/scaffolds in total)\n",
            "\n",
            "Now processing chromosomes that were not covered by any methylation calls in the coverage file...\n",
            "All chromosomes in the genome were covered by at least some reads. coverage2cytosine processing complete.\n",
            "\n",
            "\n",
            "\n",
            "Finished generating genome-wide cytosine report\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark_methylation_extractor --gzip --bedGraph --cytosine_report --genome_folder /content/drive/MyDrive/bioinf/mouse_chr /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VvduB6tGuq74",
        "outputId": "61bfc4b7-d26d-493e-dbb6-e0329698b056"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " *** Bismark methylation extractor version v0.24.2 ***\n",
            "\n",
            "Trying to determine the type of mapping from the SAM header line of file /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bam\n",
            "Treating file(s) as paired-end data (as extracted from @PG line)\n",
            "\n",
            "Setting option '--no_overlap' since this is (normally) the right thing to do for paired-end data\n",
            "\n",
            "\n",
            "Setting core usage to single-threaded (default). Consider using --multicore <int> to speed up the extraction process.\n",
            "\n",
            "Summarising Bismark methylation extractor parameters:\n",
            "===============================================================\n",
            "Bismark paired-end SAM format specified (default)\n",
            "Number of cores to be used: 1\n",
            "Output will be written to the current directory ('/content/drive/MyDrive/bioinf')\n",
            "\n",
            "\n",
            "Summarising bedGraph parameters:\n",
            "===============================================================\n",
            "Generating additional output in bedGraph and coverage format\n",
            "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
            "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
            "\n",
            "Using a cutoff of 1 read(s) to report cytosine positions\n",
            "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
            "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
            "\n",
            "\n",
            "Summarising genome-wide cytosine methylation report parameters:\n",
            "===============================================================\n",
            "Generating comprehensive genome-wide cytosine report\n",
            "(output format: <Chromosome> <Position> <Strand> <count methylated> <count non-methylated>  <C-context>  <trinucleotide context> )\n",
            "Reporting cytosine methylation in CpG context only (default)\n",
            "Using 1-based genomic coordinates (default)\n",
            "Genome folder was specified as /content/drive/MyDrive/bioinf/mouse_chr/\n",
            "\n",
            "Checking file >>/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "Now testing Bismark result file >/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bam< for positional sorting (which would be bad...)\t...passed!\n",
            "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Now reading in Bismark result file /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bam\n",
            "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping SAM header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR5836475_1.fq -2 SRR5836475_2.fq\"\n",
            "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h --threads 1 /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.bam\n",
            "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.19.2\tCL:/usr/local/bin/samtools view -bS --threads 1 -\n",
            "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.19.2\tCL:/usr/local/bin/samtools view -h /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bam\n",
            "Processed lines: 500000\n",
            "Processed lines: 1000000\n",
            "Processed lines: 1500000\n",
            "Processed lines: 2000000\n",
            "Processed lines: 2500000\n",
            "Processed lines: 3000000\n",
            "Processed lines: 3500000\n",
            "\n",
            "Processed 3791973 lines in total\n",
            "Total number of methylation call strings processed: 7583946\n",
            "\n",
            "Final Cytosine Methylation Report\n",
            "=================================\n",
            "Total number of C's analysed:\t177777254\n",
            "\n",
            "Total methylated C's in CpG context:\t1915876\n",
            "Total methylated C's in CHG context:\t298499\n",
            "Total methylated C's in CHH context:\t1434775\n",
            "\n",
            "Total C to T conversions in CpG context:\t6256205\n",
            "Total C to T conversions in CHG context:\t39157632\n",
            "Total C to T conversions in CHH context:\t128714267\n",
            "\n",
            "C methylated in CpG context:\t23.4%\n",
            "C methylated in CHG context:\t0.8%\n",
            "C methylated in CHH context:\t1.1%\n",
            "\n",
            "\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Deleting unused files ...\n",
            "\n",
            "CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
            "\n",
            "\n",
            "Using these input files: CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CpG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CpG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CpG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHH_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHH_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Summary of parameters for bismark2bedGraph conversion:\n",
            "======================================================\n",
            "bedGraph output:\t\tSRR5836475_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "output directory:\t\t><\n",
            "remove whitespaces:\t\tno\n",
            "CX context:\t\t\tno (CpG context only, default)\n",
            "No-header selected:\t\tno\n",
            "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
            "Sort buffer size:\t\t2G\n",
            "Coverage threshold:\t\t1\n",
            "=============================================================================\n",
            "Methylation information will now be written into a bedGraph and coverage file\n",
            "=============================================================================\n",
            "\n",
            "Using the following files as Input:\n",
            "/content/drive/MyDrive/bioinf/CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\t/content/drive/MyDrive/bioinf/CpG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Writing bedGraph to file: SRR5836475_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "Also writing out a coverage file including counts methylated and unmethylated residues to file: SRR5836475_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "\n",
            "Now writing methylation information for file >>CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "\n",
            "Collecting temporary chromosome file information... Processing the following input file(s):\n",
            "CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Sorting input file CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp by positions (using -S of 2G)\n",
            "Successfully deleted the temporary input file CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Finished BedGraph conversion ...\n",
            "\n",
            "Summary of parameters for genome-wide cytosine report:\n",
            "==============================================================================\n",
            "Coverage infile:\t\t\t\tSRR5836475_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "Output directory:\t\t\t\t><\n",
            "Parent directory:\t\t\t\t>/content/drive/MyDrive/bioinf/<\n",
            "Genome directory:\t\t\t\t>/content/drive/MyDrive/bioinf/mouse_chr/<\n",
            "Coverage threshold\t\t\t\t0 (all positions will be reported, default)\n",
            "CX context:\t\t\t\t\tno (CpG context only, default)\n",
            "Genome coordinates used:\t\t\t1-based (default)\n",
            "GZIP compression:\t\t\t\tyes\n",
            "Split by chromosome:\t\t\t\tno\n",
            "\n",
            "\n",
            "Now reading in and storing sequence information of the genome specified in: /content/drive/MyDrive/bioinf/mouse_chr/\n",
            "\n",
            "chr chr11 (121973369 bp)\n",
            "\n",
            "Stored sequence information of 1 chromosomes/scaffolds in total\n",
            "\n",
            "==============================================================================\n",
            "Methylation information will now be written into a genome-wide cytosine report\n",
            "==============================================================================\n",
            "\n",
            "Adding context-specific methylation summaries\n",
            "\n",
            ">>> Writing genome-wide cytosine report to: SRR5836475_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz <<<\n",
            "\n",
            ">>> Writing all cytosine context summary file to: SRR5836475_1_bismark_bt2_pe.deduplicated.cytosine_context_summary.txt <<<\n",
            "\n",
            "Storing all covered cytosine positions for chromosome: chr11\n",
            "Writing cytosine report for last chromosome chr11 (stored 2054296 different covered positions)\n",
            "Finished writing out cytosine report for covered chromosomes (processed 1 chromosomes/scaffolds in total)\n",
            "\n",
            "Now processing chromosomes that were not covered by any methylation calls in the coverage file...\n",
            "All chromosomes in the genome were covered by at least some reads. coverage2cytosine processing complete.\n",
            "\n",
            "\n",
            "\n",
            "Finished generating genome-wide cytosine report\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#как выглядел бы общий вид для трех команд\n",
        "! find /content/drive/MyDrive/bioinf -name \"*.deduplicated.bam\" | xargs -P 4 -tI{} bismark_methylation_extractor --gzip --bedGraph --cytosine_report --genome_folder //content/drive/MyDrive/bioinf/mouse_chr {}"
      ],
      "metadata": {
        "id": "x5pOD8frzoad"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%cd /content/drive/MyDrive/bioinf\n",
        "! ls -lh"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iNwzMUeAkQ4f",
        "outputId": "e60df09e-164c-405e-e05c-7807de8e2b20"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/bioinf\n",
            "total 13G\n",
            "-rw------- 1 root root   13K Feb 11 11:45 CHG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  7.7K Feb 11 12:20 CHG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  9.4K Feb 11 13:06 CHG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   14K Feb 11 11:45 CHG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  8.2K Feb 11 12:20 CHG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   10K Feb 11 13:06 CHG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  157M Feb 11 11:45 CHG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   51M Feb 11 12:20 CHG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   84M Feb 11 13:06 CHG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  157M Feb 11 11:45 CHG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   51M Feb 11 12:20 CHG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   85M Feb 11 13:06 CHG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  1.9M Feb 11 11:45 CHH_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  1.1M Feb 11 12:20 CHH_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  1.6M Feb 11 13:06 CHH_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  1.7M Feb 11 11:45 CHH_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root 1014K Feb 11 12:20 CHH_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  1.4M Feb 11 13:06 CHH_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  419M Feb 11 11:45 CHH_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  133M Feb 11 12:20 CHH_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  223M Feb 11 13:06 CHH_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  419M Feb 11 11:45 CHH_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  133M Feb 11 12:20 CHH_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  223M Feb 11 13:06 CHH_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  7.5K Feb 11 11:45 CpG_CTOB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  4.2K Feb 11 12:20 CpG_CTOB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  4.7K Feb 11 13:06 CpG_CTOB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  8.0K Feb 11 11:45 CpG_CTOT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  3.8K Feb 11 12:20 CpG_CTOT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root  4.6K Feb 11 13:06 CpG_CTOT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   49M Feb 11 11:45 CpG_OB_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   17M Feb 11 12:20 CpG_OB_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   28M Feb 11 13:06 CpG_OB_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   49M Feb 11 11:45 CpG_OT_SRR3824222_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   17M Feb 11 12:20 CpG_OT_SRR5836473_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "-rw------- 1 root root   28M Feb 11 13:06 CpG_OT_SRR5836475_1_bismark_bt2_pe.deduplicated.txt.gz\n",
            "drwx------ 3 root root  4.0K Feb 11 10:23 mouse_chr\n",
            "-rw------- 1 root root  528M Feb 10 20:07 sorted_8cell.bam\n",
            "-rw------- 1 root root  203K Feb 10 20:17 sorted_8cell.bam.bai\n",
            "-rw------- 1 root root  1.3G Feb 10 20:10 sorted_epiblast.bam\n",
            "-rw------- 1 root root  349K Feb 10 20:18 sorted_epiblast.bam.bai\n",
            "-rw------- 1 root root  777M Feb 10 20:15 sorted_icm.bam\n",
            "-rw------- 1 root root  341K Feb 10 20:18 sorted_icm.bam.bai\n",
            "-rw------- 1 root root  1.9G Feb 10 19:13 SRR3824222_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root  1.9G Feb 10 20:33 SRR3824222_1_bismark_bt2_pe.deduplicated.bam\n",
            "-rw------- 1 root root   14M Feb 11 11:48 SRR3824222_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "-rw------- 1 root root   14M Feb 11 11:48 SRR3824222_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "-rw------- 1 root root   11M Feb 11 11:52 SRR3824222_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz\n",
            "-rw------- 1 root root  1.7K Feb 11 11:52 SRR3824222_1_bismark_bt2_pe.deduplicated.cytosine_context_summary.txt\n",
            "-rw------- 1 root root   24K Feb 11 11:45 SRR3824222_1_bismark_bt2_pe.deduplicated.M-bias.txt\n",
            "-rw------- 1 root root   864 Feb 11 11:45 SRR3824222_1_bismark_bt2_pe.deduplicated_splitting_report.txt\n",
            "-rw------- 1 root root   312 Feb 10 20:33 SRR3824222_1_bismark_bt2_pe.deduplication_report.txt\n",
            "-rw------- 1 root root  3.1M Feb 11 11:56 SRR3824222_1_bismark_bt2_PE_report.html\n",
            "-rw------- 1 root root  1.8K Feb 10 19:16 SRR3824222_1_bismark_bt2_PE_report.txt\n",
            "-rw------- 1 root root  769M Feb 10 19:09 SRR5836473_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root  620M Feb 10 20:27 SRR5836473_1_bismark_bt2_pe.deduplicated.bam\n",
            "-rw------- 1 root root  9.8M Feb 11 12:21 SRR5836473_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "-rw------- 1 root root  9.4M Feb 11 12:21 SRR5836473_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "-rw------- 1 root root  9.9M Feb 11 12:25 SRR5836473_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz\n",
            "-rw------- 1 root root  1.6K Feb 11 12:25 SRR5836473_1_bismark_bt2_pe.deduplicated.cytosine_context_summary.txt\n",
            "-rw------- 1 root root   22K Feb 11 12:20 SRR5836473_1_bismark_bt2_pe.deduplicated.M-bias.txt\n",
            "-rw------- 1 root root   860 Feb 11 12:20 SRR5836473_1_bismark_bt2_pe.deduplicated_splitting_report.txt\n",
            "-rw------- 1 root root   313 Feb 10 20:27 SRR5836473_1_bismark_bt2_pe.deduplication_report.txt\n",
            "-rw------- 1 root root  3.1M Feb 11 12:26 SRR5836473_1_bismark_bt2_PE_report.html\n",
            "-rw------- 1 root root  1.8K Feb 10 19:16 SRR5836473_1_bismark_bt2_PE_report.txt\n",
            "-rw------- 1 root root  1.2G Feb 10 19:11 SRR5836475_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root  1.1G Feb 10 20:30 SRR5836475_1_bismark_bt2_pe.deduplicated.bam\n",
            "-rw------- 1 root root   12M Feb 11 13:08 SRR5836475_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
            "-rw------- 1 root root   12M Feb 11 13:08 SRR5836475_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
            "-rw------- 1 root root   11M Feb 11 13:12 SRR5836475_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz\n",
            "-rw------- 1 root root  1.6K Feb 11 13:12 SRR5836475_1_bismark_bt2_pe.deduplicated.cytosine_context_summary.txt\n",
            "-rw------- 1 root root   23K Feb 11 13:06 SRR5836475_1_bismark_bt2_pe.deduplicated.M-bias.txt\n",
            "-rw------- 1 root root   861 Feb 11 13:06 SRR5836475_1_bismark_bt2_pe.deduplicated_splitting_report.txt\n",
            "-rw------- 1 root root   312 Feb 10 20:30 SRR5836475_1_bismark_bt2_pe.deduplication_report.txt\n",
            "-rw------- 1 root root  1.8K Feb 10 19:16 SRR5836475_1_bismark_bt2_PE_report.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark2report --alignment_report /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_PE_report.txt -mbias_report /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.M-bias.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5Sc-qMKIvvRg",
        "outputId": "22505e41-f946-4bcc-ab62-7d8ebadff8a6"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Writing Bismark HTML report to >> SRR5836473_1_bismark_bt2_PE_report.html <<\n",
            "\n",
            "==============================================================================================================\n",
            "Using the following alignment report:\t\t> /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_PE_report.txt <\n",
            "Processing alignment report /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_PE_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following deduplication report:\t> /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplication_report.txt <\n",
            "Processing deduplication report /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplication_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following splitting report:\t\t> /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated_splitting_report.txt <\n",
            "Processing splitting report /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated_splitting_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following M-bias report:\t\t> /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.M-bias.txt <\n",
            "Processing M-bias report /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.M-bias.txt ...\n",
            "Complete\n",
            "\n",
            "No nucleotide coverage report file specified, skipping this step\n",
            "==============================================================================================================\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark2report --alignment_report /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_PE_report.txt -mbias_report /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.M-bias.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AKqGihDC4tPq",
        "outputId": "f986edab-0466-4bdb-ac4b-7475bee3573b"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Writing Bismark HTML report to >> SRR3824222_1_bismark_bt2_PE_report.html <<\n",
            "\n",
            "==============================================================================================================\n",
            "Using the following alignment report:\t\t> /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_PE_report.txt <\n",
            "Processing alignment report /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_PE_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following deduplication report:\t> /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplication_report.txt <\n",
            "Processing deduplication report /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplication_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following splitting report:\t\t> /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated_splitting_report.txt <\n",
            "Processing splitting report /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated_splitting_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following M-bias report:\t\t> /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.M-bias.txt <\n",
            "Processing M-bias report /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.M-bias.txt ...\n",
            "Complete\n",
            "\n",
            "No nucleotide coverage report file specified, skipping this step\n",
            "==============================================================================================================\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark2report --alignment_report /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_PE_report.txt -mbias_report //content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.M-bias.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YERW0CyM4tDV",
        "outputId": "378f0c07-833a-4d11-a8b3-0c4c05acbb39"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Writing Bismark HTML report to >> SRR5836475_1_bismark_bt2_PE_report.html <<\n",
            "\n",
            "==============================================================================================================\n",
            "Using the following alignment report:\t\t> /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_PE_report.txt <\n",
            "Processing alignment report /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_PE_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following deduplication report:\t> /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplication_report.txt <\n",
            "Processing deduplication report /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplication_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following splitting report:\t\t> /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated_splitting_report.txt <\n",
            "Processing splitting report /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated_splitting_report.txt ...\n",
            "Complete\n",
            "\n",
            "Using the following M-bias report:\t\t> //content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.M-bias.txt <\n",
            "Processing M-bias report //content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.M-bias.txt ...\n",
            "Complete\n",
            "\n",
            "No nucleotide coverage report file specified, skipping this step\n",
            "==============================================================================================================\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Уставновка необходимых пакетов"
      ],
      "metadata": {
        "id": "wSPyj5YWID_f"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "b6k9REsM6NXn",
        "outputId": "4606e643-a29b-4f15-8379-3c2afd5fbadb"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting pyGenomeTracks\n",
            "  Downloading pyGenomeTracks-3.8-py2.py3-none-any.whl (113 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m113.6/113.6 kB\u001b[0m \u001b[31m6.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting numpy>=1.20 (from pyGenomeTracks)\n",
            "  Downloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m61.0/61.0 kB\u001b[0m \u001b[31m4.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting matplotlib<=3.6.2,>=3.1.1 (from pyGenomeTracks)\n",
            "  Downloading matplotlib-3.6.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.8 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.8/11.8 MB\u001b[0m \u001b[31m51.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting intervaltree>=2.1.0 (from pyGenomeTracks)\n",
            "  Using cached intervaltree-3.1.0-py2.py3-none-any.whl\n",
            "Collecting pyBigWig>=0.3.16 (from pyGenomeTracks)\n",
            "  Downloading pyBigWig-0.3.22-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (209 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m209.7/209.7 kB\u001b[0m \u001b[31m14.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting future>=0.17.0 (from pyGenomeTracks)\n",
            "  Downloading future-0.18.3.tar.gz (840 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m840.9/840.9 kB\u001b[0m \u001b[31m34.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting hicmatrix>=15 (from pyGenomeTracks)\n",
            "  Downloading HiCMatrix-17.1-py3-none-any.whl.metadata (2.0 kB)\n",
            "Collecting pysam>=0.14 (from pyGenomeTracks)\n",
            "  Downloading pysam-0.22.0-cp310-cp310-manylinux_2_28_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting pytest (from pyGenomeTracks)\n",
            "  Downloading pytest-8.0.0-py3-none-any.whl.metadata (7.8 kB)\n",
            "Collecting gffutils>=0.9 (from pyGenomeTracks)\n",
            "  Downloading gffutils-0.12-py3-none-any.whl.metadata (1.5 kB)\n",
            "Collecting pybedtools>=0.8.1 (from pyGenomeTracks)\n",
            "  Using cached pybedtools-0.9.1-cp310-cp310-linux_x86_64.whl\n",
            "Requirement already satisfied: tqdm>=4.20 in /usr/local/lib/python3.10/site-packages (from pyGenomeTracks) (4.66.1)\n",
            "Collecting bx-python>=0.8.13 (from pyGenomeTracks)\n",
            "  Downloading bx_python-0.11.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.0 kB)\n",
            "Collecting pyfaidx>=0.1.3 (from pyGenomeTracks)\n",
            "  Downloading pyfaidx-0.8.1.1-py3-none-any.whl.metadata (25 kB)\n",
            "Collecting six>=1.12.0 (from gffutils>=0.9->pyGenomeTracks)\n",
            "  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)\n",
            "Collecting argh>=0.26.2 (from gffutils>=0.9->pyGenomeTracks)\n",
            "  Downloading argh-0.31.2-py3-none-any.whl.metadata (7.3 kB)\n",
            "Collecting argcomplete>=1.9.4 (from gffutils>=0.9->pyGenomeTracks)\n",
            "  Downloading argcomplete-3.2.2-py3-none-any.whl.metadata (16 kB)\n",
            "Collecting simplejson (from gffutils>=0.9->pyGenomeTracks)\n",
            "  Downloading simplejson-3.19.2-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.1 kB)\n",
            "Collecting scipy>=1.2 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading scipy-1.12.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (60 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m60.4/60.4 kB\u001b[0m \u001b[31m4.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting tables>=3.5 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading tables-3.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.3 kB)\n",
            "Collecting pandas>=0.25 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading pandas-2.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (19 kB)\n",
            "Collecting cooler>=0.8.9 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading cooler-0.9.3-py3-none-any.whl.metadata (8.2 kB)\n",
            "Collecting sortedcontainers<3.0,>=2.0 (from intervaltree>=2.1.0->pyGenomeTracks)\n",
            "  Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)\n",
            "Collecting contourpy>=1.0.1 (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading contourpy-1.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.8 kB)\n",
            "Collecting cycler>=0.10 (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)\n",
            "Collecting fonttools>=4.22.0 (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading fonttools-4.48.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (158 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m158.9/158.9 kB\u001b[0m \u001b[31m13.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting kiwisolver>=1.0.1 (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading kiwisolver-1.4.5-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl.metadata (6.4 kB)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/site-packages (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks) (23.2)\n",
            "Collecting pillow>=6.2.0 (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading pillow-10.2.0-cp310-cp310-manylinux_2_28_x86_64.whl.metadata (9.7 kB)\n",
            "Collecting pyparsing>=2.2.1 (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading pyparsing-3.1.1-py3-none-any.whl.metadata (5.1 kB)\n",
            "Collecting python-dateutil>=2.7 (from matplotlib<=3.6.2,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m247.7/247.7 kB\u001b[0m \u001b[31m18.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: setuptools in /usr/local/lib/python3.10/site-packages (from pyfaidx>=0.1.3->pyGenomeTracks) (68.2.2)\n",
            "Collecting importlib-metadata (from pyfaidx>=0.1.3->pyGenomeTracks)\n",
            "  Downloading importlib_metadata-7.0.1-py3-none-any.whl.metadata (4.9 kB)\n",
            "Collecting iniconfig (from pytest->pyGenomeTracks)\n",
            "  Downloading iniconfig-2.0.0-py3-none-any.whl (5.9 kB)\n",
            "Requirement already satisfied: pluggy<2.0,>=1.3.0 in /usr/local/lib/python3.10/site-packages (from pytest->pyGenomeTracks) (1.3.0)\n",
            "Collecting exceptiongroup>=1.0.0rc8 (from pytest->pyGenomeTracks)\n",
            "  Downloading exceptiongroup-1.2.0-py3-none-any.whl.metadata (6.6 kB)\n",
            "Collecting tomli>=1.0.0 (from pytest->pyGenomeTracks)\n",
            "  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)\n",
            "Collecting asciitree (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Using cached asciitree-0.3.3-py3-none-any.whl\n",
            "Collecting click>=7 (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading click-8.1.7-py3-none-any.whl.metadata (3.0 kB)\n",
            "Collecting cytoolz (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading cytoolz-0.12.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.6 kB)\n",
            "Collecting h5py>=2.5 (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading h5py-3.10.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.5 kB)\n",
            "Collecting multiprocess (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading multiprocess-0.70.16-py310-none-any.whl.metadata (7.2 kB)\n",
            "Collecting pyyaml (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading PyYAML-6.0.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.1 kB)\n",
            "Collecting pytz>=2020.1 (from pandas>=0.25->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)\n",
            "Collecting tzdata>=2022.7 (from pandas>=0.25->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading tzdata-2023.4-py2.py3-none-any.whl.metadata (1.4 kB)\n",
            "Collecting numexpr>=2.6.2 (from tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading numexpr-2.9.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (7.9 kB)\n",
            "Collecting py-cpuinfo (from tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading py_cpuinfo-9.0.0-py3-none-any.whl (22 kB)\n",
            "Collecting blosc2>=2.3.0 (from tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading blosc2-2.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (9.2 kB)\n",
            "Collecting zipp>=0.5 (from importlib-metadata->pyfaidx>=0.1.3->pyGenomeTracks)\n",
            "  Downloading zipp-3.17.0-py3-none-any.whl.metadata (3.7 kB)\n",
            "Collecting ndindex>=1.4 (from blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading ndindex-1.7-py3-none-any.whl (85 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m85.7/85.7 kB\u001b[0m \u001b[31m7.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting msgpack (from blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading msgpack-1.0.7-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (9.1 kB)\n",
            "Collecting toolz>=0.8.0 (from cytoolz->cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading toolz-0.12.1-py3-none-any.whl.metadata (5.1 kB)\n",
            "Collecting dill>=0.3.8 (from multiprocess->cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading dill-0.3.8-py3-none-any.whl.metadata (10 kB)\n",
            "Downloading bx_python-0.11.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.6/5.6 MB\u001b[0m \u001b[31m88.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading gffutils-0.12-py3-none-any.whl (1.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m64.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading HiCMatrix-17.1-py3-none-any.whl (47 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m47.4/47.4 kB\u001b[0m \u001b[31m3.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.2/18.2 MB\u001b[0m \u001b[31m72.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pyfaidx-0.8.1.1-py3-none-any.whl (28 kB)\n",
            "Downloading pysam-0.22.0-cp310-cp310-manylinux_2_28_x86_64.whl (21.9 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m21.9/21.9 MB\u001b[0m \u001b[31m49.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pytest-8.0.0-py3-none-any.whl (334 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m334.0/334.0 kB\u001b[0m \u001b[31m7.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading argcomplete-3.2.2-py3-none-any.whl (42 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m42.3/42.3 kB\u001b[0m \u001b[31m3.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading argh-0.31.2-py3-none-any.whl (44 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m44.8/44.8 kB\u001b[0m \u001b[31m3.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading contourpy-1.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (310 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m310.7/310.7 kB\u001b[0m \u001b[31m24.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading cooler-0.9.3-py3-none-any.whl (103 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m103.7/103.7 kB\u001b[0m \u001b[31m9.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading cycler-0.12.1-py3-none-any.whl (8.3 kB)\n",
            "Downloading exceptiongroup-1.2.0-py3-none-any.whl (16 kB)\n",
            "Downloading fonttools-4.48.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.6/4.6 MB\u001b[0m \u001b[31m93.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading kiwisolver-1.4.5-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m60.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pandas-2.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.0/13.0 MB\u001b[0m \u001b[31m98.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pillow-10.2.0-cp310-cp310-manylinux_2_28_x86_64.whl (4.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.5/4.5 MB\u001b[0m \u001b[31m83.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pyparsing-3.1.1-py3-none-any.whl (103 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m103.1/103.1 kB\u001b[0m \u001b[31m9.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading scipy-1.12.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (38.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m38.4/38.4 MB\u001b[0m \u001b[31m12.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading tables-3.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (7.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.5/7.5 MB\u001b[0m \u001b[31m87.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading importlib_metadata-7.0.1-py3-none-any.whl (23 kB)\n",
            "Downloading simplejson-3.19.2-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (137 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m137.9/137.9 kB\u001b[0m \u001b[31m12.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading blosc2-2.5.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.4/4.4 MB\u001b[0m \u001b[31m81.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading click-8.1.7-py3-none-any.whl (97 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m97.9/97.9 kB\u001b[0m \u001b[31m9.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading h5py-3.10.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.8 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.8/4.8 MB\u001b[0m \u001b[31m80.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading numexpr-2.9.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (375 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m375.2/375.2 kB\u001b[0m \u001b[31m29.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pytz-2024.1-py2.py3-none-any.whl (505 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m505.5/505.5 kB\u001b[0m \u001b[31m37.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading tzdata-2023.4-py2.py3-none-any.whl (346 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m346.6/346.6 kB\u001b[0m \u001b[31m28.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading zipp-3.17.0-py3-none-any.whl (7.4 kB)\n",
            "Downloading cytoolz-0.12.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.0/2.0 MB\u001b[0m \u001b[31m73.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading multiprocess-0.70.16-py310-none-any.whl (134 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m134.8/134.8 kB\u001b[0m \u001b[31m13.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading PyYAML-6.0.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (705 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m705.5/705.5 kB\u001b[0m \u001b[31m45.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dill-0.3.8-py3-none-any.whl (116 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m116.3/116.3 kB\u001b[0m \u001b[31m8.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading toolz-0.12.1-py3-none-any.whl (56 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.1/56.1 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading msgpack-1.0.7-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (530 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m530.8/530.8 kB\u001b[0m \u001b[31m32.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hBuilding wheels for collected packages: future\n",
            "  Building wheel for future (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for future: filename=future-0.18.3-py3-none-any.whl size=492024 sha256=0d04456684277faf0a8637df4874f1aeb9685c3722df172c7bca06f50502e9db\n",
            "  Stored in directory: /root/.cache/pip/wheels/5e/a9/47/f118e66afd12240e4662752cc22cefae5d97275623aa8ef57d\n",
            "Successfully built future\n",
            "Installing collected packages: sortedcontainers, pytz, py-cpuinfo, asciitree, zipp, tzdata, toolz, tomli, six, simplejson, pyyaml, pysam, pyparsing, pyBigWig, pillow, numpy, ndindex, msgpack, kiwisolver, intervaltree, iniconfig, future, fonttools, exceptiongroup, dill, cycler, click, argh, argcomplete, scipy, python-dateutil, pytest, pybedtools, numexpr, multiprocess, importlib-metadata, h5py, cytoolz, contourpy, bx-python, blosc2, tables, pyfaidx, pandas, matplotlib, gffutils, cooler, hicmatrix, pyGenomeTracks\n",
            "Successfully installed argcomplete-3.2.2 argh-0.31.2 asciitree-0.3.3 blosc2-2.5.1 bx-python-0.11.0 click-8.1.7 contourpy-1.2.0 cooler-0.9.3 cycler-0.12.1 cytoolz-0.12.3 dill-0.3.8 exceptiongroup-1.2.0 fonttools-4.48.1 future-0.18.3 gffutils-0.12 h5py-3.10.0 hicmatrix-17.1 importlib-metadata-7.0.1 iniconfig-2.0.0 intervaltree-3.1.0 kiwisolver-1.4.5 matplotlib-3.6.2 msgpack-1.0.7 multiprocess-0.70.16 ndindex-1.7 numexpr-2.9.0 numpy-1.26.4 pandas-2.2.0 pillow-10.2.0 py-cpuinfo-9.0.0 pyBigWig-0.3.22 pyGenomeTracks-3.8 pybedtools-0.9.1 pyfaidx-0.8.1.1 pyparsing-3.1.1 pysam-0.22.0 pytest-8.0.0 python-dateutil-2.8.2 pytz-2024.1 pyyaml-6.0.1 scipy-1.12.0 simplejson-3.19.2 six-1.16.0 sortedcontainers-2.4.0 tables-3.9.2 tomli-2.0.1 toolz-0.12.1 tzdata-2023.4 zipp-3.17.0\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
            "\u001b[0m"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "cycler",
                  "dateutil",
                  "kiwisolver",
                  "matplotlib",
                  "mpl_toolkits",
                  "six"
                ]
              }
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "! pip install pyGenomeTracks"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install -U ipykernel"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "fcsJWNmHo1i-",
        "outputId": "369f4c9c-410c-40a2-cf94-552a17a2438b"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting ipykernel\n",
            "  Downloading ipykernel-6.29.2-py3-none-any.whl.metadata (6.0 kB)\n",
            "Collecting comm>=0.1.1 (from ipykernel)\n",
            "  Downloading comm-0.2.1-py3-none-any.whl.metadata (3.7 kB)\n",
            "Collecting debugpy>=1.6.5 (from ipykernel)\n",
            "  Downloading debugpy-1.8.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.1 kB)\n",
            "Collecting ipython>=7.23.1 (from ipykernel)\n",
            "  Downloading ipython-8.21.0-py3-none-any.whl.metadata (5.9 kB)\n",
            "Collecting jupyter-client>=6.1.12 (from ipykernel)\n",
            "  Downloading jupyter_client-8.6.0-py3-none-any.whl.metadata (8.3 kB)\n",
            "Collecting jupyter-core!=5.0.*,>=4.12 (from ipykernel)\n",
            "  Downloading jupyter_core-5.7.1-py3-none-any.whl.metadata (3.4 kB)\n",
            "Collecting matplotlib-inline>=0.1 (from ipykernel)\n",
            "  Downloading matplotlib_inline-0.1.6-py3-none-any.whl (9.4 kB)\n",
            "Collecting nest-asyncio (from ipykernel)\n",
            "  Downloading nest_asyncio-1.6.0-py3-none-any.whl.metadata (2.8 kB)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/site-packages (from ipykernel) (23.2)\n",
            "Collecting psutil (from ipykernel)\n",
            "  Downloading psutil-5.9.8-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (21 kB)\n",
            "Collecting pyzmq>=24 (from ipykernel)\n",
            "  Downloading pyzmq-25.1.2-cp310-cp310-manylinux_2_28_x86_64.whl.metadata (4.9 kB)\n",
            "Collecting tornado>=6.1 (from ipykernel)\n",
            "  Downloading tornado-6.4-cp38-abi3-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.5 kB)\n",
            "Collecting traitlets>=5.4.0 (from ipykernel)\n",
            "  Downloading traitlets-5.14.1-py3-none-any.whl.metadata (10 kB)\n",
            "Collecting decorator (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading decorator-5.1.1-py3-none-any.whl (9.1 kB)\n",
            "Collecting jedi>=0.16 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading jedi-0.19.1-py2.py3-none-any.whl.metadata (22 kB)\n",
            "Collecting prompt-toolkit<3.1.0,>=3.0.41 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading prompt_toolkit-3.0.43-py3-none-any.whl.metadata (6.5 kB)\n",
            "Collecting pygments>=2.4.0 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading pygments-2.17.2-py3-none-any.whl.metadata (2.6 kB)\n",
            "Collecting stack-data (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading stack_data-0.6.3-py3-none-any.whl.metadata (18 kB)\n",
            "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/site-packages (from ipython>=7.23.1->ipykernel) (1.2.0)\n",
            "Collecting pexpect>4.3 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading pexpect-4.9.0-py2.py3-none-any.whl.metadata (2.5 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/site-packages (from jupyter-client>=6.1.12->ipykernel) (2.8.2)\n",
            "Requirement already satisfied: platformdirs>=2.5 in /usr/local/lib/python3.10/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel) (4.1.0)\n",
            "Collecting parso<0.9.0,>=0.8.3 (from jedi>=0.16->ipython>=7.23.1->ipykernel)\n",
            "  Downloading parso-0.8.3-py2.py3-none-any.whl (100 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m100.8/100.8 kB\u001b[0m \u001b[31m4.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting ptyprocess>=0.5 (from pexpect>4.3->ipython>=7.23.1->ipykernel)\n",
            "  Downloading ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)\n",
            "Collecting wcwidth (from prompt-toolkit<3.1.0,>=3.0.41->ipython>=7.23.1->ipykernel)\n",
            "  Downloading wcwidth-0.2.13-py2.py3-none-any.whl.metadata (14 kB)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/site-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel) (1.16.0)\n",
            "Collecting executing>=1.2.0 (from stack-data->ipython>=7.23.1->ipykernel)\n",
            "  Downloading executing-2.0.1-py2.py3-none-any.whl.metadata (9.0 kB)\n",
            "Collecting asttokens>=2.1.0 (from stack-data->ipython>=7.23.1->ipykernel)\n",
            "  Downloading asttokens-2.4.1-py2.py3-none-any.whl.metadata (5.2 kB)\n",
            "Collecting pure-eval (from stack-data->ipython>=7.23.1->ipykernel)\n",
            "  Downloading pure_eval-0.2.2-py3-none-any.whl (11 kB)\n",
            "Downloading ipykernel-6.29.2-py3-none-any.whl (116 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m116.1/116.1 kB\u001b[0m \u001b[31m9.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading comm-0.2.1-py3-none-any.whl (7.2 kB)\n",
            "Downloading debugpy-1.8.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.0/3.0 MB\u001b[0m \u001b[31m50.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading ipython-8.21.0-py3-none-any.whl (810 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m810.0/810.0 kB\u001b[0m \u001b[31m33.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading jupyter_client-8.6.0-py3-none-any.whl (105 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m105.9/105.9 kB\u001b[0m \u001b[31m8.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading jupyter_core-5.7.1-py3-none-any.whl (28 kB)\n",
            "Downloading pyzmq-25.1.2-cp310-cp310-manylinux_2_28_x86_64.whl (1.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.1 MB\u001b[0m \u001b[31m32.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading tornado-6.4-cp38-abi3-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (435 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m435.4/435.4 kB\u001b[0m \u001b[31m19.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading traitlets-5.14.1-py3-none-any.whl (85 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m85.4/85.4 kB\u001b[0m \u001b[31m7.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nest_asyncio-1.6.0-py3-none-any.whl (5.2 kB)\n",
            "Downloading psutil-5.9.8-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (288 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m288.2/288.2 kB\u001b[0m \u001b[31m8.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading jedi-0.19.1-py2.py3-none-any.whl (1.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m43.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pexpect-4.9.0-py2.py3-none-any.whl (63 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m63.8/63.8 kB\u001b[0m \u001b[31m639.9 kB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading prompt_toolkit-3.0.43-py3-none-any.whl (386 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m386.1/386.1 kB\u001b[0m \u001b[31m15.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pygments-2.17.2-py3-none-any.whl (1.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m25.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading stack_data-0.6.3-py3-none-any.whl (24 kB)\n",
            "Downloading asttokens-2.4.1-py2.py3-none-any.whl (27 kB)\n",
            "Downloading executing-2.0.1-py2.py3-none-any.whl (24 kB)\n",
            "Downloading wcwidth-0.2.13-py2.py3-none-any.whl (34 kB)\n",
            "Installing collected packages: wcwidth, pure-eval, ptyprocess, traitlets, tornado, pyzmq, pygments, psutil, prompt-toolkit, pexpect, parso, nest-asyncio, executing, decorator, debugpy, asttokens, stack-data, matplotlib-inline, jupyter-core, jedi, comm, jupyter-client, ipython, ipykernel\n",
            "Successfully installed asttokens-2.4.1 comm-0.2.1 debugpy-1.8.1 decorator-5.1.1 executing-2.0.1 ipykernel-6.29.2 ipython-8.21.0 jedi-0.19.1 jupyter-client-8.6.0 jupyter-core-5.7.1 matplotlib-inline-0.1.6 nest-asyncio-1.6.0 parso-0.8.3 pexpect-4.9.0 prompt-toolkit-3.0.43 psutil-5.9.8 ptyprocess-0.7.0 pure-eval-0.2.2 pygments-2.17.2 pyzmq-25.1.2 stack-data-0.6.3 tornado-6.4 traitlets-5.14.1 wcwidth-0.2.13\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
            "\u001b[0m"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "jedi",
                  "matplotlib_inline",
                  "parso",
                  "pexpect",
                  "prompt_toolkit",
                  "wcwidth",
                  "zmq"
                ]
              }
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! conda install bioconda/label/cf201901::ucsc-fetchchromsizes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UhIKyKz_t7yy",
        "outputId": "67f0d0d2-af6e-4e6c-d580-cf5fc447f8bb"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Channels:\n",
            " - conda-forge\n",
            " - bioconda/label/cf201901\n",
            " - bioconda\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Solving environment: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - bioconda/label/cf201901::ucsc-fetchchromsizes\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    libpng-1.6.42              |       h2797004_0         282 KB  conda-forge\n",
            "    mysql-connector-c-6.1.11   |    h659d440_1008         1.2 MB  conda-forge\n",
            "    ucsc-fetchchromsizes-357   |       h5eb252a_2           5 KB  bioconda/label/cf201901\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         1.5 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  libpng             conda-forge/linux-64::libpng-1.6.42-h2797004_0 \n",
            "  mysql-connector-c  conda-forge/linux-64::mysql-connector-c-6.1.11-h659d440_1008 \n",
            "  ucsc-fetchchromsi~ bioconda/label/cf201901/linux-64::ucsc-fetchchromsizes-357-h5eb252a_2 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "mysql-connector-c-6. | 1.2 MB    | :   0% 0/1 [00:00<?, ?it/s]\n",
            "libpng-1.6.42        | 282 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\n",
            "\n",
            "mysql-connector-c-6. | 1.2 MB    | :   1% 0.01291599165947316/1 [00:00<00:10, 10.31s/it]\n",
            "libpng-1.6.42        | 282 KB    | :   6% 0.05667243168453822/1 [00:00<00:02,  2.38s/it]\u001b[A\n",
            "libpng-1.6.42        | 282 KB    | : 100% 1.0/1 [00:00<00:00,  4.57it/s]                \u001b[A\n",
            "libpng-1.6.42        | 282 KB    | : 100% 1.0/1 [00:00<00:00,  4.57it/s]\u001b[A\n",
            "\n",
            "ucsc-fetchchromsizes | 5 KB      | : 100% 1.0/1 [00:00<00:00,  3.60it/s]\u001b[A\u001b[A\n",
            "\n",
            "                                                                        \n",
            "                                                                        \u001b[A\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\n",
            "Preparing transaction: \\ \b\bdone\n",
            "Verifying transaction: / \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Executing transaction: - \b\bdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! fetchChromSizes mm10 > m.chrom.sizes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LZpspVUvt4EY",
        "outputId": "fc4b260f-e24c-475f-a51e-89acd1f131d9"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO: trying WGET /usr/bin/wget for database mm10\n",
            "url: http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! conda install -c bioconda ucsc-bedgraphtobigwig"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "svn9iAuGtVxX",
        "outputId": "0e21cc23-e83a-45fe-cfef-a75b6ab00e87"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            " - bioconda/label/cf201901\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - ucsc-bedgraphtobigwig\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    ucsc-bedgraphtobigwig-455  |       h2a80c09_0         2.4 MB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         2.4 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  ucsc-bedgraphtobi~ bioconda/linux-64::ucsc-bedgraphtobigwig-455-h2a80c09_0 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "                                                                        \n",
            "Preparing transaction: \\ \b\bdone\n",
            "Verifying transaction: / \b\bdone\n",
            "Executing transaction: \\ \b\bdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! apt-get install bedtools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XEWV8p4ku-Hy",
        "outputId": "dc296f49-ea17-4349-a228-e085a94df35c"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "bedtools is already the newest version (2.30.0+dfsg-2ubuntu0.1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 32 not upgraded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Распаковка файлов"
      ],
      "metadata": {
        "id": "KNBHWG3fILm_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! gunzip /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
        "! gunzip /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
        "! gunzip /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
        "! gunzip /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
        "! gunzip /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
        "! gunzip /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bismark.cov.gz"
      ],
      "metadata": {
        "id": "tPxNkLdxHWlK"
      },
      "execution_count": 38,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! head /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bedGraph"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qofYnYciunf8",
        "outputId": "d0e269fa-b0e0-425a-f616-343e337e913f"
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "track type=bedGraph\n",
            "chr11\t3050030\t3050031\t50\n",
            "chr11\t3050047\t3050048\t66.6666666666667\n",
            "chr11\t3050048\t3050049\t50\n",
            "chr11\t3050223\t3050224\t87.5\n",
            "chr11\t3050224\t3050225\t53.3333333333333\n",
            "chr11\t3050285\t3050286\t66.6666666666667\n",
            "chr11\t3050286\t3050287\t33.3333333333333\n",
            "chr11\t3050309\t3050310\t62.5\n",
            "chr11\t3050310\t3050311\t50\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! head /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bismark.cov"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2MX4eZ31Sf8B",
        "outputId": "3a537422-4411-46e8-dfb5-0fa233547a19"
      },
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "chr11\t3050031\t3050031\t50\t1\t1\n",
            "chr11\t3050048\t3050048\t66.6666666666667\t2\t1\n",
            "chr11\t3050049\t3050049\t50\t1\t1\n",
            "chr11\t3050224\t3050224\t87.5\t7\t1\n",
            "chr11\t3050225\t3050225\t53.3333333333333\t8\t7\n",
            "chr11\t3050286\t3050286\t66.6666666666667\t6\t3\n",
            "chr11\t3050287\t3050287\t33.3333333333333\t5\t10\n",
            "chr11\t3050310\t3050310\t62.5\t5\t3\n",
            "chr11\t3050311\t3050311\t50\t7\t7\n",
            "chr11\t3050349\t3050349\t55.5555555555556\t5\t4\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Построим распределения метилирования цитозинов по хромосоме"
      ],
      "metadata": {
        "id": "jG7FxVYeE2D3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from matplotlib import pyplot as plt\n",
        "import seaborn as sns\n",
        "\n",
        "sns.set_style(\"darkgrid\")"
      ],
      "metadata": {
        "id": "9gohVysTEnlc"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def hist_plot(ax, path, title):\n",
        "    bedGraph = pd.read_csv(path,  delimiter='\\t', skiprows=1, header=None)\n",
        "\n",
        "    ax.hist(bedGraph[3], bins=100)\n",
        "    ax.set_title(title)\n",
        "    ax.set_xlabel('Процент метилированных цитозинов')\n",
        "    ax.set_ylabel('Частота')"
      ],
      "metadata": {
        "id": "CgQQuvOEEiSi"
      },
      "execution_count": 52,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "paths = ['/content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bedGraph',\n",
        "         '/content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bedGraph',\n",
        "         '/content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bedGraph']\n",
        "names = ['Распределение 8_cell',\n",
        "         'Распределение ICM',\n",
        "         'Распределение epiblast']\n",
        "fig, axs = plt.subplots(3, figsize=(10, 18))\n",
        "for i in range(3):\n",
        "    hist_plot(axs[i], paths[i], names[i])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "BI7s17KqErQc",
        "outputId": "3d07e7ad-2ffc-4eef-8e8a-ca44541f2b85"
      },
      "execution_count": 53,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1000x1800 with 3 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA2wAAAW/CAYAAADQFnDqAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAADzS0lEQVR4nOzdeXyNd/7//+c5WWoJkSaWMaGUiiWJBNM0Ec1UVamaFkUXlA+1lKKlqGrtiap2LFWU2lOqtdXaop3+GKFoUqWWYhRlyEZIkO38/vDNNU4TnJOmzpXmcb/dcrs51/t9XdcrOa+GZ6/reh+LzWazCQAAAABgOlZXFwAAAAAAKBiBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAA/CmsWrVKAQEBOnPmjLGta9eu6tq1qwurAoDfx93VBQAAXGvVqlV64403jNeenp6qWrWqmjZtqpdffll+fn4urA5mtXPnTs2aNUtHjx5VTk6OatSooS5duujpp592dWkA8KdCYAMASJIGDhwof39/ZWZmat++fVq2bJm+/fZbrV+/XqVLl3Z1eTCRbdu2qX///goJCdErr7wii8WiTZs2afjw4bp48aK6d+/u6hIB4E+DwAYAkCQ9/PDDCgoKkiR17NhRFSpU0IIFC7Rt2zY9+eSTLq4OZhIbG6uKFStq8eLF8vT0lCR17txZrVu31qpVqwhsAFCEeIYNAFCghx56SJKM54EuXryod955R23btlVoaKgaNWqkXr166fDhw/n2vX79umbMmKHHH39cQUFBioyM1IABA3Tq1CnjmAEBAbf8uvmZo927dysgIEAbN27U+++/r6ZNmyokJER9+/bVuXPn8p37hx9+UM+ePdW4cWM1bNhQXbp00b59+wr8Hrt27Vrg+WfMmJFv7tq1a9W+fXsFBwfrwQcf1Kuvvlrg+W/3vd0sNzdXCxcuVJs2bRQUFKSIiAi9/fbbunTpkt285s2bq0+fPvnOM27cuHzHLKj2efPm5fuZSlJmZqamT5+uxx57TIGBgYqKitLkyZOVmZlZ4M/qZleuXJG3t7cR1iTJ3d1dPj4+KlWq1B33/61vv/1WXbp0MfqqQ4cOWrdund0cZ95XAPgz4QobAKBAeeGqQoUKkqTTp09r69atatWqlfz9/ZWUlKRPP/1UXbp00YYNG1S5cmVJUk5Ojvr06aO4uDi1adNG3bp1U3p6uv7973/r6NGjql69unGOJ598Ug8//LDded9///0C65k1a5YsFoteeuklJScna9GiRerevbvWrl1rhIS4uDi99NJLCgwM1IABA2SxWLRq1Sq9+OKL+uSTTxQcHJzvuFWqVNFrr70mScrIyNCYMWMKPPe0adPUunVrPfPMM0pJSdHSpUv1wgsvaM2aNSpfvny+fTp37qzGjRtLkrZs2aItW7bYjb/99ttavXq12rdvr65du+rMmTOKjY3VTz/9pGXLlsnDw6PAn4Mz0tLS9NFHH+Xbnpubq379+mnfvn3q1KmTatWqpaNHj2rRokU6efKkPvzww9se98EHH9TcuXM1depUtWvXThaLRevWrdOBAwc0depUp2pctWqVRo4cqQceeEB9+vRRuXLldOjQIW3fvl1t27aVVLj3FQD+LAhsAABJN66apKSkKDMzU99//71mzpypUqVK6ZFHHpF04+rNl19+Kav1fzdnPPXUU2rdurU+//xz9e/fX5K0Zs0axcXF6Y033rC7Na53796y2Wx256xfv76eeuopu21z584tsL5Lly5p48aN8vLyMvYdPHiwVqxYoW7duslms2nMmDEKCwvTvHnzZLFYJEnPPvus2rRpo6lTp2r+/Pl2x8zKylL58uWNGlJSUvIFtl9//VUzZszQ4MGD1bdvX2N7y5Yt1a5dO33yySd223NyciRJjRo1Mo576tQpu8C2d+9effbZZ5oyZYoRSiQpLCxMvXr10ubNm+22F9acOXPk7u6uBg0a2G1ft26ddu7cqSVLlqhJkybG9gceeECjR4/W999/r0aNGt3yuC+//LLOnDmj2bNna9asWZKk0qVLa/r06WrRooXD9V2+fFkTJkxQcHCwlixZonvuuccYy+uVwryvAPBnwi2RAABJUvfu3RUeHq6oqCi9+uqrKlu2rD744APjypmnp6cR1nJycpSamqoyZcqoZs2a+umnn4zjfPXVV/Lx8VGXLl3ynSPvH9uF8fTTTxthTZJatWqlihUr6ttvv5UkHTp0SCdPnlTbtm2VmpqqlJQUpaSkKCMjQ+Hh4dqzZ49yc3PtjpmZmWl3W19BtmzZotzcXLVu3do4ZkpKivz8/HTfffdp9+7ddvOzsrIk6bbH3bx5s8qVK6emTZvaHbNBgwYqU6ZMvmNmZ2fbzUtJSdH169dvW/f58+e1dOlSvfzyyypbtmy+89eqVUv333+/3THzboP97fl/y9PTUzVq1NDjjz+u999/X++++64CAwP1+uuvKyEh4bb73uzf//630tPT1bt3b7uwJv2vVwrzvgLAnwlX2AAAkm7colezZk25ubnJz89PNWvWtLualpubq8WLF+uTTz7RmTNnjCtJ0v9um5RuXE2qWbOm3N2L9q+Y++67z+61xWLRfffdp19//VWSdPLkSUnS8OHDb3mMy5cvy9vb23idmpqa77i/dfLkSdlsNrVs2bLA8d9+n2lpaZKkMmXK3PKYv/zyiy5fvqzw8PACx5OTk+1e79ix45Zzb2X69OmqVKmSOnfurC+//DLf+Y8fP+7w+X9r3Lhx+uGHH7R69WqjR1q3bq0nn3xSEydO1GeffeZQjXm33T7wwAO3nFOY9xUA/kwIbAAASVJwcLCxSmRBZs+erWnTpqlDhw4aNGiQvL29ZbVaFR0dne9WR1fIq2HYsGGqV69egXNuDlGZmZlKTExURETEbY+bm5sri8WiuXPnys3N7bbHlKSkpCRJuu3n1+Xm5srX11dTpkwpcPzee++1e92wYUMNHjzYbtvSpUu1bdu2Avc/fvy4Vq9erXfffbfAZ+Fyc3NVp04du8/fu1mVKlVuWXtmZqZWrlypXr162QV6Dw8PNWvWTLGxsQ5duXSUs+8rAPzZENgAAA758ssvFRYWpujoaLvtaWlp8vHxMV5Xr15dP/zwg7Kysopk4Yw8v/zyi91rm82mX375xVgpsVq1apIkLy+vO4YwSTp8+LCysrIUGBh423nVq1eXzWaTv7+/atasecfjHjt2TBaL5bZzq1evrri4ODVq1MihVRV9fHzyfU9bt2695fz33ntPdevW1RNPPHHL8x8+fFjh4eFO36Z68eJFZWdn211hzZOdna3c3FyHb1HMW4Dm559/vuWVTmffVwD4s+EZNgCAQ9zc3PJdSdu0aZPOnz9vt61ly5ZKTU1VbGxsvmP8nitxa9as0ZUrV4zXmzdvVmJiorHKZGBgoKpXr6758+crPT093/4pKSl2rzdv3iw3NzdjUZVbadmypdzc3PTBBx/kq99msyk1NdV4nZ2dra+++krBwcH5nhu7WevWrZWTk1PgaozZ2dnGbZWFkZCQoG3btmno0KG3DGOtW7fW+fPntWLFinxj165dU0ZGxi2P7+vrq/Lly2vLli12HwGQnp6ub775Rvfff7/DS/tHRkaqbNmymjNnTr5n8vJ+1s6+rwDwZ8MVNgCAQ/7+979r5syZeuONNxQaGqqjR49q3bp1xhWQPE8//bTWrFmjmJgY7d+/X40bN9bVq1cVFxen5557zqlVBG/m7e2t559/Xu3btzeW9b/vvvvUqVMnSZLVatWECRP00ksv6cknn1T79u1VuXJlnT9/Xrt375aXl5dmz56tjIwMxcbGasmSJapRo4bdAht5QeXIkSOKj49XaGioqlevrsGDB+u9997Tr7/+qhYtWqhs2bI6c+aMtm7dqk6dOqlnz57auXOnpk2bpiNHjmj27Nm3/V4efPBBde7cWXPmzNGhQ4fUtGlTeXh46OTJk9q8ebPefPNNtWrVqlA/px07dqhp06a3vRr11FNPadOmTRo9erR2796tRo0aKScnRydOnNDmzZs1b968W94e6+bmpv/7v//T1KlT1blzZz311FPKzc3V559/rv/+97969913Ha7Vy8tLb7zxhkaNGqVnnnlGTz75pMqXL6/Dhw/r2rVreueddxx+XwHgz4rABgBwSN++fXX16lWtW7dOGzduVP369TVnzhy99957dvPc3Nw0d+5czZo1S+vXr9dXX32lChUqqFGjRvk+6NnZ8x85ckQfffSR0tPTFR4ertGjR6t06dLGnLCwMH366af68MMPtXTpUmVkZKhixYoKDg5W586dJd24IpP37Njx48c1bNiwfOfasmWLvLy8FBoaKunGRxLUqFFDCxcu1MyZMyXdeM6radOmat68uSTp66+/loeHhz766CM1a9bsjt/PuHHjFBgYqOXLl+uf//yn3Nzc9Ne//lX/+Mc/bruk/p1YLBYNGTLktnOsVqtmzpyphQsXau3atdqyZYtKly4tf39/de3a9Y63fvbr10/+/v5avHixZs6cqczMTAUEBGj69Ol6/PHHnaq3Y8eO8vX11UcffaQPP/xQ7u7uuv/+++0+EsKR9xUA/qwsNjM8KQ4AwC3s3r1b3bp107Rp0wp91elmZ86c0aOPPqpt27bJ39+/wDkzZszQr7/+qkmTJv3u8wEA8HvwDBsAAAAAmBS3RAIASpQyZcqobdu2t10KPiAgQJUqVbqLVf35pKSkFLiSZB4PDw+7z+8DABSMwAYAKFHuvffeW37+WZ5bfUg2HPfMM88YH2pekAcffFBLliy5ixUBQPHEM2wAAKDI7du3L99S/TcrX778HT8DDwBAYAMAAAAA02LREQAAAAAwKQIbAAAAAJgUgQ0AAAAATIpVIl0gOfmyXP3koMUi+fqWM0UtKB7oGTiLnoGz6Bk4i56Bs8zUM3m13AmBzQVsNrm8QfKYqRYUD/QMnEXPwFn0DJxFz8BZxalnuCUSAAAAAEyKwAYAAAAAJkVgAwAAAACTcmlgy8nJ0dSpU9W8eXMFBwerRYsWmjlzpm7+LG+bzaZp06YpMjJSwcHB6t69u06ePGl3nIsXL2rIkCFq1KiRmjRpopEjRyo9Pd1uzuHDh/X8888rKChIUVFRmjt3br56Nm3apFatWikoKEht27bVt99+azfuSC0AAAAAUFRcGtjmzp2rZcuW6e2339bGjRs1dOhQzZs3T0uWLLGbs2TJEo0ZM0YrVqxQ6dKl1bNnT12/ft2YM3ToUB07dkwLFizQ7NmztXfvXr399tvG+JUrV9SzZ09VrVpVq1at0rBhw/TBBx/o008/NeZ8//33GjJkiJ555hmtWbNGjz76qPr376+jR486VQsAAAAAFBWXrhIZHx+vRx99VH//+98lSf7+/tqwYYP2798v6cYVrcWLF6tfv35q0aKFJGny5MmKiIjQ1q1b1aZNGx0/flzbt2/X559/rqCgIEnSqFGj1Lt3bw0bNkyVK1fWF198oaysLEVHR8vT01MPPPCADh06pAULFqhz586SpMWLF6tZs2bq1auXJGnw4MHauXOnli5dqnHjxjlUi6MsliL58f0ueTWYoRYUD/QMnEXPwFn0DJxFz8BZZuoZR2twaWALDQ3VihUr9J///Ec1a9bU4cOHtW/fPo0YMUKSdObMGSUmJioiIsLYp1y5cmrYsKHi4+PVpk0bxcfHq3z58kZYk6SIiAhZrVbt379fjz32mBISEtSkSRN5enoacyIjIzV37lxdunRJ3t7eSkhIUPfu3e3qi4yM1NatWx2uxVGOfN7C3WKmWlA80DNwFj0DZ9EzcBY9A2cVp55xaWDr3bu3rly5otatW8vNzU05OTl69dVX9Y9//EOSlJiYKEny9fW128/X11dJSUmSpKSkJN1777124+7u7vL29jb2T0pKkr+/v90cPz8/Y8zb21tJSUnGtoLO40gtjjLTB/WZoRYUD/QMnEXPwFn0DJxFz8BZZuqZYvHB2Zs2bdK6dev03nvvqXbt2jp06JBiYmJUqVIltWvXzpWl/aHM9EF9ZqoFxQM9A2fRM3AWPQNn0TNwVnHqGZcuOjJ58mT17t1bbdq0UUBAgJ5++mm9+OKLmjNnjiSpYsWKkqTk5GS7/ZKTk42rYX5+fkpJSbEbz87O1qVLl4z9/fz88l0Fy3t983F+O+fm8zhSCwAAAAAUJZcGtmvXrsnym6ft3NzcjGX9/f39VbFiRcXFxRnjV65c0Q8//KDQ0FBJN56DS0tL04EDB4w5u3btUm5uroKDgyVJISEh2rt3r7Kysow5O3fuVM2aNeXt7W3M2bVrl10tO3fuVEhIiMO1AAAAAEBRcmlge+SRRzR79mz961//0pkzZ7RlyxYtWLDAWIXRYrGoW7dumjVrlrZt26YjR45o2LBhqlSpkjGnVq1aatasmd566y3t379f+/bt0/jx49WmTRtVrlxZktS2bVt5eHjozTff1M8//6yNGzdq8eLF6tGjh1FLt27dtH37ds2fP1/Hjx/XjBkzdODAAXXp0sXhWgAAAACgKFlsNtfdvXnlyhVNmzZNW7duVXJysipVqqQ2bdqof//+xoqONptN06dP14oVK5SWlqbGjRtr9OjRqlmzpnGcixcvavz48fr6669ltVrVsmVLjRo1SmXLljXmHD58WOPGjdOPP/4oHx8fdenSRb1797arZ9OmTZo6dap+/fVX1ahRQ6+//rqioqKMcUdqcURSkjkecvTzK2eKWlA80DNwFj0DZ9EzcBY9A2eZqWfyarnjPFcGtpLKTA1ihlpQPNAzcBY9A2fRM3AWPQNnmalnHA1sLr0lEgAAAABway5d1h+u5+ZWcGbPzbUpN5f/VQUAAAC4EoGthLJaLcrJtcnHp2yB49k5ubp0MYPQBgAAALgQga2EslgscrNaNGh5vI5duGI3VruSl6Y9Gyqr1UJgAwAAAFyIwFbCHbtwRQfPprm6DAAAAAAFYNERAAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADAplwa25s2bKyAgIN/X2LFjJUnXr1/X2LFjFRYWptDQUL3yyitKSkqyO8bZs2fVu3dvNWzYUOHh4XrnnXeUnZ1tN2f37t1q166dAgMD9dhjj2nVqlX5aomNjVXz5s0VFBSkjh07av/+/XbjjtQCAAAAAEXJpYHt888/144dO4yvBQsWSJJatWolSYqOjtY333yjqVOnasmSJbpw4YIGDBhg7J+Tk6M+ffooKytLy5cv16RJk7R69WpNnz7dmHP69Gn16dNHYWFhWrt2rV588UWNGjVK27dvN+Zs3LhRMTEx6t+/v1avXq26deuqZ8+eSk5ONubcqRYAAAAAKGouDWz33nuvKlasaHx98803ql69uh588EFdvnxZK1eu1IgRIxQeHq7AwEBFR0crPj5eCQkJkqQdO3bo2LFjevfdd1WvXj1FRUVp0KBBio2NVWZmpiRp+fLl8vf314gRI1SrVi116dJFjz/+uBYuXGjUsWDBAnXq1EkdOnRQ7dq1NXbsWJUqVUorV66UJIdqAQAAAICi5u7qAvJkZmbqiy++UI8ePWSxWHTgwAFlZWUpIiLCmFOrVi1VrVpVCQkJCgkJUUJCgurUqSM/Pz9jTmRkpMaMGaNjx46pfv36SkhIUHh4uN25IiMjFR0dbZz34MGD6tOnjzFutVoVERGh+Ph4SXKoFmdYLE5N/0M4WoMZaoU55PUCPQFH0TNwFj0DZ9EzcJaZesbRGkwT2LZu3arLly+rXbt2kqSkpCR5eHiofPnydvN8fX2VmJhozLk5rEkyXt9pzpUrV3Tt2jVdunRJOTk58vX1zXeeEydOOFyLM3x9yzm9jyv4+JR1dQkwoeLSvzAPegbOomfgLHoGzipOPWOawLZy5Uo9/PDDqly5sqtL+cMlJ1+WzebaGtzdrapQ4faBLDU1XTk5uXepIpidxXLjl5sZ+hfFAz0DZ9EzcBY9A2eZqWfyarkTUwS2X3/9VTt37tSMGTOMbX5+fsrKylJaWprdla3k5GRVrFjRmPPb1RzzVm68ec5vV3NMSkqSl5eXSpUqJavVKjc3N7sFRvLOk3dlzpFanGGzyeUN4uj5XV0nzMcM/YvihZ6Bs+gZOIuegbOKU8+Y4nPYVq1aJV9fX/397383tgUGBsrDw0NxcXHGthMnTujs2bPGM2MhISE6evSoXdjauXOnvLy8VLt2bWPOrl277M63c+dO4xienp5q0KCB3Xlyc3MVFxen0NBQh2sBAAAAgKLm8itsubm5WrVqlZ5++mm5u/+vnHLlyqlDhw6aNGmSvL295eXlpQkTJig0NNQISZGRkapdu7aGDRum119/XYmJiZo6dapeeOEFeXp6SpKeffZZxcbGavLkyerQoYN27dqlTZs2ac6cOca5evTooeHDhyswMFDBwcFatGiRrl69qvbt2ztcCwAAAAAUNZcHtp07d+rs2bPq0KFDvrGRI0fKarVq4MCByszMVGRkpEaPHm2Mu7m5afbs2RozZow6d+6s0qVLq127dho4cKAxp1q1apozZ45iYmK0ePFiValSRRMmTFCzZs2MOU888YRSUlI0ffp0JSYmql69epo3b57dYiV3qgUAAAAAiprFZisud2/+eSQluf4hR3d3q3x8yqrN9O06eDbNbqxB1fLaMLCZUlPTlZ3NoiO4wWKR/PzKmaJ/UTzQM3AWPQNn0TNwlpl6Jq+WOzHFM2wAAAAAgPwIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAk3J5YDt//ryGDh2qsLAwBQcHq23btvrxxx+NcZvNpmnTpikyMlLBwcHq3r27Tp48aXeMixcvasiQIWrUqJGaNGmikSNHKj093W7O4cOH9fzzzysoKEhRUVGaO3duvlo2bdqkVq1aKSgoSG3bttW3335rN+5ILQAAAABQVFwa2C5duqTnnntOHh4emjt3rjZs2KDhw4fL29vbmDN37lwtWbJEY8aM0YoVK1S6dGn17NlT169fN+YMHTpUx44d04IFCzR79mzt3btXb7/9tjF+5coV9ezZU1WrVtWqVas0bNgwffDBB/r000+NOd9//72GDBmiZ555RmvWrNGjjz6q/v376+jRo07VAgAAAABFxaWBbe7cuapSpYpiYmIUHBysatWqKTIyUtWrV5d044rW4sWL1a9fP7Vo0UJ169bV5MmTdeHCBW3dulWSdPz4cW3fvl0TJkxQw4YN1aRJE40aNUobNmzQ+fPnJUlffPGFsrKyFB0drQceeEBt2rRR165dtWDBAqOWxYsXq1mzZurVq5dq1aqlwYMHq379+lq6dKnDtQAAAABAUXJ35cm//vprRUZGauDAgdqzZ48qV66s559/Xp06dZIknTlzRomJiYqIiDD2KVeunBo2bKj4+Hi1adNG8fHxKl++vIKCgow5ERERslqt2r9/vx577DElJCSoSZMm8vT0NOZERkZq7ty5unTpkry9vZWQkKDu3bvb1RcZGWmEMUdqcZTF4tSP6Q/haA1mqBXmkNcL9AQcRc/AWfQMnEXPwFlm6hlHa3BpYDt9+rSWLVumHj16qG/fvvrxxx81YcIEeXh4qF27dkpMTJQk+fr62u3n6+urpKQkSVJSUpLuvfdeu3F3d3d5e3sb+yclJcnf399ujp+fnzHm7e2tpKQkY1tB53GkFkf5+pZzar6r+PiUdXUJMKHi0r8wD3oGzqJn4Cx6Bs4qTj3j0sBms9kUGBio1157TZJUv359/fzzz1q+fLnatWvnytL+UMnJl2WzubYGd3erKlS4fSBLTU1XTk7uXaoIZmex3PjlZob+RfFAz8BZ9AycRc/AWWbqmbxa7sSlga1ixYqqVauW3bb7779fX375pTEuScnJyapUqZIxJzk5WXXr1pV040pZSkqK3TGys7N16dIlY38/P798V8HyXuddVStoTnJysjHuSC2Ostnk8gZx9PyurhPmY4b+RfFCz8BZ9AycRc/AWcWpZ1y66EijRo30n//8x27byZMn9de//lWS5O/vr4oVKyouLs4Yv3Llin744QeFhoZKkkJDQ5WWlqYDBw4Yc3bt2qXc3FwFBwdLkkJCQrR3715lZWUZc3bu3KmaNWsaK1KGhIRo165ddrXs3LlTISEhDtcCAAAAAEXJpYHtxRdf1A8//KDZs2frl19+0bp167RixQo9//zzkiSLxaJu3bpp1qxZ2rZtm44cOaJhw4apUqVKatGihSSpVq1aatasmd566y3t379f+/bt0/jx49WmTRtVrlxZktS2bVt5eHjozTff1M8//6yNGzdq8eLF6tGjh1FLt27dtH37ds2fP1/Hjx/XjBkzdODAAXXp0sXhWgAAAACgKFlsNtdeDPzmm2/0/vvv6+TJk/L391ePHj2MVSKlG8+5TZ8+XStWrFBaWpoaN26s0aNHq2bNmsacixcvavz48fr6669ltVrVsmVLjRo1SmXL/u8ZrcOHD2vcuHH68ccf5ePjoy5duqh37952tWzatElTp07Vr7/+qho1auj1119XVFSUU7U4IinJ9ffMurtb5eNTVm2mb9fBs2l2Yw2qlteGgc2Umpqu7GyeYcMNFovk51fOFP2L4oGegbPoGTiLnoGzzNQzebXccZ6rA1tJZIYGIbDBWWb6BYfigZ6Bs+gZOIuegbPM1DOOBjaX3hIJAAAAALg1AhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKRcGthmzJihgIAAu69WrVoZ49evX9fYsWMVFham0NBQvfLKK0pKSrI7xtmzZ9W7d281bNhQ4eHheuedd5SdnW03Z/fu3WrXrp0CAwP12GOPadWqVflqiY2NVfPmzRUUFKSOHTtq//79duOO1AIAAAAARcnlV9geeOAB7dixw/j65JNPjLHo6Gh98803mjp1qpYsWaILFy5owIABxnhOTo769OmjrKwsLV++XJMmTdLq1as1ffp0Y87p06fVp08fhYWFae3atXrxxRc1atQobd++3ZizceNGxcTEqH///lq9erXq1q2rnj17Kjk52eFaAAAAAKCouTywubm5qWLFisbXvffeK0m6fPmyVq5cqREjRig8PFyBgYGKjo5WfHy8EhISJEk7duzQsWPH9O6776pevXqKiorSoEGDFBsbq8zMTEnS8uXL5e/vrxEjRqhWrVrq0qWLHn/8cS1cuNCoYcGCBerUqZM6dOig2rVra+zYsSpVqpRWrlzpcC0AAAAAUNTcXV3AL7/8osjISN1zzz0KCQnRkCFDVLVqVR04cEBZWVmKiIgw5taqVUtVq1ZVQkKCQkJClJCQoDp16sjPz8+YExkZqTFjxujYsWOqX7++EhISFB4ebnfOyMhIRUdHS5IyMzN18OBB9enTxxi3Wq2KiIhQfHy8JDlUizMsFqem/yEcrcEMtcIc8nqBnoCj6Bk4i56Bs+gZOMtMPeNoDS4NbMHBwYqJiVHNmjWVmJiomTNn6oUXXtC6deuUlJQkDw8PlS9f3m4fX19fJSYmSpKSkpLswpok4/Wd5ly5ckXXrl3TpUuXlJOTI19f33znOXHihHGMO9XiDF/fck7v4wo+PmVdXQJMqLj0L8yDnoGz6Bk4i56Bs4pTz7g0sEVFRRl/rlu3rho2bKhHHnlEmzZtUqlSpVxY2R8rOfmybDbX1uDublWFCrcPZKmp6crJyb1LFcHsLJYbv9zM0L8oHugZOIuegbPoGTjLTD2TV8uduPyWyJuVL19eNWrU0KlTpxQREaGsrCylpaXZXdlKTk5WxYoVJd24Uvbb1RzzVm68ec5vV3NMSkqSl5eXSpUqJavVKjc3N7sFRvLOk3dlzs/P7461OMNmk8sbxNHzu7pOmI8Z+hfFCz0DZ9EzcBY9A2cVp55x+aIjN0tPT9fp06dVsWJFBQYGysPDQ3Fxccb4iRMndPbsWeOZsZCQEB09etQubO3cuVNeXl6qXbu2MWfXrl1259m5c6dxDE9PTzVo0MDuPLm5uYqLi1NoaKgkOVQLAAAAABQ1l15he+edd/TII4+oatWqunDhgmbMmCGr1aonn3xS5cqVU4cOHTRp0iR5e3vLy8tLEyZMUGhoqBGSIiMjVbt2bQ0bNkyvv/66EhMTNXXqVL3wwgvy9PSUJD377LOKjY3V5MmT1aFDB+3atUubNm3SnDlzjDp69Oih4cOHKzAwUMHBwVq0aJGuXr2q9u3bS5JDtQAAAABAUXNpYPvvf/+r1157TRcvXtS9996rxo0ba8WKFcbS/iNHjpTVatXAgQOVmZmpyMhIjR492tjfzc1Ns2fP1pgxY9S5c2eVLl1a7dq108CBA4051apV05w5cxQTE6PFixerSpUqmjBhgpo1a2bMeeKJJ5SSkqLp06crMTFR9erV07x58+wWK7lTLQAAAABQ1Cw2W3G5e/PPIynJ9Q85urtb5eNTVm2mb9fBs2l2Yw2qlteGgc2Umpqu7GwWHcENFovk51fOFP2L4oGegbPoGTiLnoGzzNQzebXciameYQMAAAAA/A+BDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJuRd2x4yMDO3Zs0dnz55VVlaW3Vi3bt1+d2EAAAAAUNIVKrD99NNP6t27t65evaqrV6/K29tbqampKl26tO69914CGwAAAAAUgULdEhkTE6NHHnlEe/bs0T333KMVK1bom2++UYMGDTR8+PCirhEAAAAASqRCBbZDhw6pR48eslqtcnNzU2Zmpv7yl7/o9ddf1/vvv1/UNQIAAABAiVSowObu7i6r9cauvr6+Onv2rCTJy8tL//3vf4uuOgAAAAAowQr1DFv9+vX1448/qkaNGvrb3/6m6dOnKzU1VWvXrtUDDzxQ1DUCAAAAQIlUqCtsr776qipWrGj8uXz58hozZoxSU1M1fvz4Ii0QAAAAAEqqQl1hCwoKMv7s6+urjz/+uMgKAgAAAIDfw2q1yGq1uLqMIlGoK2zdunVTWlpaUdcCAAAAAL+L1WqRd4Uy8vEpW+BXTq6tWIW5Ql1h++677/J9WDYAAAAAuJrVapG7m1WDlsfr2IUrdmO1K3lp2rOhslgskmyuKdBJhQpskv7fNwkAAAAA5nPswhUdPFv87wosdGDr37+/PDw8ChxbvHhxoQsCAAAAANxQ6MAWEhKismXLFmUtAAAAAICbFCqwWSwW9erVS76+vkVdDwAAAADg/ynUKpE2W/F4QA8AAAAAirNCBbYBAwaoTJkyRV0LAAAAAOAmhbolcsCAAZKklJQUnThxQpJ0//3369577y26ygAAAACghCtUYLt69arGjRunL774Qjk5OZIkNzc3PfXUU3rrrbdUunTpIi0SAAAAAEqiQt0SGRMToz179ujDDz/U3r17tXfvXn344Yfas2ePJk2aVNQ1AgAAAECJVKjA9uWXX2rixImKioqSl5eXvLy8FBUVpfHjx+vLL78s6hoBAAAAoEQqVGC7du2a/Pz88m339fXVtWvXfndRAAAAAIBCBraQkBBNnz5d169fN7Zdu3ZNH3zwgUJCQoqqNgAAAAAo0Qq16MjIkSPVq1cvPfzww6pbt64k6fDhw7rnnnv08ccfF2mBAAAAAFBSFSqwBQQE6KuvvtK6deuMZf2ffPJJtW3bVqVKlSrSAgEAAACgpCpUYNuzZ49CQ0PVqVOnoq4HAAAAAPD/FOoZtm7duunSpUtFXQsAAAAA4CaFCmw2m62o69BHH32kgIAATZw40dh2/fp1jR07VmFhYQoNDdUrr7yipKQku/3Onj2r3r17q2HDhgoPD9c777yj7Oxsuzm7d+9Wu3btFBgYqMcee0yrVq3Kd/7Y2Fg1b95cQUFB6tixo/bv32837kgtAAAAAFCUChXYJCk+Pl579uwp8MtZ+/fv1/LlyxUQEGC3PTo6Wt98842mTp2qJUuW6MKFCxowYIAxnpOToz59+igrK0vLly/XpEmTtHr1ak2fPt2Yc/r0afXp00dhYWFau3atXnzxRY0aNUrbt2835mzcuFExMTHq37+/Vq9erbp166pnz55KTk52uBYAAAAAKGqFeoZN0i3DisVi0aFDhxw+Tnp6ul5//XVNmDBBs2bNMrZfvnxZK1eu1JQpUxQeHi7pRmh64oknlJCQoJCQEO3YsUPHjh3TggUL5Ofnp3r16mnQoEGaMmWKBgwYIE9PTy1fvlz+/v4aMWKEJKlWrVrat2+fFi5cqGbNmkmSFixYoE6dOqlDhw6SpLFjx+pf//qXVq5cqd69eztUizMsFqem/yEcrcEMtcIc8nqBnoCj6Bk4i56Bs+gZFJbF4vq+cfT8hQ5s//73v+Xr61vY3Q3jxo1TVFSUIiIi7ALbgQMHlJWVpYiICGNbrVq1VLVqVSMkJSQkqE6dOnYf4h0ZGakxY8bo2LFjql+/vhISEoyQdfOc6OhoSVJmZqYOHjyoPn36GONWq1URERGKj493uBZn+PqWc2q+q/j4lHV1CTCh4tK/MA96Bs6iZ+AsegbOqlCh+Pw7t1CBzVJEcXTDhg366aef9Pnnn+cbS0pKkoeHh8qXL2+33dfXV4mJicacm8OaJOP1neZcuXJF165d06VLl5STk5MvfPr6+hofWeBILc5ITr6sP+AxQKe4u1vv2KipqenKycm9SxXB7CyWG38hmqF/UTzQM3AWPQNn0TMoiJub9Y4XHi5eTFd2tmv/nZvXv3dSqMBWFIuOnDt3ThMnTtT8+fN1zz33/O7jFSc2m1z+S8XR87u6TpiPGfoXxQs9A2fRM3AWPQNnFaeeKVRgO3z48O8+8cGDB5WcnKz27dsb23JycrRnzx7Fxsbq448/VlZWltLS0uyubCUnJ6tixYqSblwp++1qjnkrN94857erOSYlJcnLy0ulSpWS1WqVm5ub3QIjeefJuzLn5+d3x1oAAAAAoKgVapXIOXPmFHgb4+eff66PPvrIoWM89NBDWrdundasWWN8BQYGqm3btsafPTw8FBcXZ+xz4sQJnT171nhmLCQkREePHrULWzt37pSXl5dq165tzNm1a5fduXfu3Gkcw9PTUw0aNLA7T25uruLi4hQaGipJDtUCAAAAAEWtUIHt008/1f33359v+wMPPKDly5c7dAwvLy/VqVPH7qtMmTKqUKGC6tSpo3LlyqlDhw6aNGmSdu3apQMHDmjkyJEKDQ01QlJkZKRq166tYcOG6fDhw9q+fbumTp2qF154QZ6enpKkZ599VqdPn9bkyZN1/PhxxcbGatOmTerevbtRS48ePbRixQqtXr1ax48f15gxY3T16lXj6p8jtQAAAABAUSvULZGJiYkF3gp47733FmoRjlsZOXKkrFarBg4cqMzMTEVGRmr06NHGuJubm2bPnq0xY8aoc+fOKl26tNq1a6eBAwcac6pVq6Y5c+YoJiZGixcvVpUqVTRhwgRjSX9JeuKJJ5SSkqLp06crMTFR9erV07x58+wWK7lTLQAAAABQ1Cy2Qqwg0rJlS/Xv319PPfWU3fY1a9ZoxowZ2rZtW5EV+GeUlOT6lYzc3W+sntNm+nYdPJtmN9aganltGNhMqamuXz0H5mGxSH5+5UzRvyge6Bk4i56Bs+gZFKS4/Ds3r3/vpFBX2Dp27Kjo6GhlZ2froYcekiTFxcXp3Xff1f/93/8V5pAAAAAAgN8oVGDr1auXLl68qLFjxyorK0uSdM8996hXr152H0ANAAAAACi8Qn9w9uuvv66XX35Zx48fV6lSpVSjRg1joQ8AAAAAwO9XqMCWp2zZsgoODi6qWgAAAAAANyl0YPvxxx+1adMmnTt3zrgtMs8HH3zwuwsDAAAAgJKuUJ/DtmHDBj333HM6ceKEtmzZouzsbP3888/atWuXypW780onAAAAAIA7K1Rgmz17tt544w3Nnj1bHh4eevPNN7V582a1bt1af/nLX4q6RgAAAAAokQoV2E6fPq2oqChJkqenpzIyMmSxWNS9e3etWLGiSAsEAAAAgJKqUIGtfPnySk9PlyRVqlRJP//8syQpLS1NV69eLbrqAAAAAKAEK9SiI3/729+0c+dOBQQEqFWrVpo4caJ27dqlnTt3Kjw8vKhrBAAAAIASqVCB7a233tL169clSf369ZOHh4e+//57tWzZUv369SvSAgEAAACgpHIqsF25cuXGTu7ucnd3N14///zzev7554u+OgAAAAAowZwKbE2aNJHFYrnjvEOHDhW6IAAAAADADU4FtsWLF9u9ttls6t27tyZMmKDKlSsXaWEAAAAAUNI5FdgefPDBfNusVqtCQkJUrVq1IisKAAAAAFDIZf0BAAAAAH+83xXYzp07p2vXrqlChQpFVA4AAAAAIE+hn2FLTU3Vhg0b9NBDD6lcuXJFXhgAAAAAlHROBbaFCxdKkiwWi3x8fPTII4/wuWsAAAAA8AdxKrB9/fXXf1QdAAAAAIDfYNERAAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmJRLA9snn3yitm3bqlGjRmrUqJE6d+6sb7/91hi/fv26xo4dq7CwMIWGhuqVV15RUlKS3THOnj2r3r17q2HDhgoPD9c777yj7Oxsuzm7d+9Wu3btFBgYqMcee0yrVq3KV0tsbKyaN2+uoKAgdezYUfv377cbd6QWAAAAAChKLg1sVapU0dChQ7Vq1SqtXLlSDz30kPr376+ff/5ZkhQdHa1vvvlGU6dO1ZIlS3ThwgUNGDDA2D8nJ0d9+vRRVlaWli9frkmTJmn16tWaPn26Mef06dPq06ePwsLCtHbtWr344osaNWqUtm/fbszZuHGjYmJi1L9/f61evVp169ZVz549lZycbMy5Uy0AAAAAUNRcGtiaN2+uqKgo1ahRQzVr1tSrr76qMmXKKCEhQZcvX9bKlSs1YsQIhYeHKzAwUNHR0YqPj1dCQoIkaceOHTp27Jjeffdd1atXT1FRURo0aJBiY2OVmZkpSVq+fLn8/f01YsQI1apVS126dNHjjz+uhQsXGnUsWLBAnTp1UocOHVS7dm2NHTtWpUqV0sqVKyXJoVoAAAAAoKi5u7qAPDk5Odq8ebMyMjIUGhqqAwcOKCsrSxEREcacWrVqqWrVqkpISFBISIgSEhJUp04d+fn5GXMiIyM1ZswYHTt2TPXr11dCQoLCw8PtzhUZGano6GhJUmZmpg4ePKg+ffoY41arVREREYqPj5ckh2pxhsXi1PQ/hKM1mKFWmENeL9ATcBQ9A2fRM3AWPYPCslhc3zeOnt/lge3IkSN69tlndf36dZUpU0YzZ85U7dq1dejQIXl4eKh8+fJ28319fZWYmChJSkpKsgtrkozXd5pz5coVXbt2TZcuXVJOTo58fX3znefEiRPGMe5UizN8fcs5vY8r+PiUdXUJMKHi0r8wD3oGzqJn4Cx6Bs6qUKH4/DvX5YGtZs2aWrNmjS5fvqwvv/xSw4cP19KlS11d1h8qOfmybDbX1uDubr1jo6ampisnJ/cuVQSzs1hu/IVohv5F8UDPwFn0DJxFz6Agbm7WO154uHgxXdnZrv13bl7/3onLA5unp6fuu+8+SVJgYKB+/PFHLV68WK1bt1ZWVpbS0tLsrmwlJyerYsWKkm5cKfvtao55KzfePOe3qzkmJSXJy8tLpUqVktVqlZubm90CI3nnybsy5+fnd8danGGzyeW/VBw9v6vrhPmYoX9RvNAzcBY9A2fRM3BWceoZ030OW25urjIzMxUYGCgPDw/FxcUZYydOnNDZs2eNZ8ZCQkJ09OhRu7C1c+dOeXl5qXbt2sacXbt22Z1j586dxjE8PT3VoEEDu/Pk5uYqLi5OoaGhkuRQLQAAAABQ1Fx6he29997Tww8/rL/85S9KT0/X+vXr9d133+njjz9WuXLl1KFDB02aNEne3t7y8vLShAkTFBoaaoSkyMhI1a5dW8OGDdPrr7+uxMRETZ06VS+88II8PT0lSc8++6xiY2M1efJkdejQQbt27dKmTZs0Z84co44ePXpo+PDhCgwMVHBwsBYtWqSrV6+qffv2kuRQLQAAAABQ1Fwa2JKTkzV8+HBduHBB5cqVU0BAgD7++GM1bdpUkjRy5EhZrVYNHDhQmZmZioyM1OjRo4393dzcNHv2bI0ZM0adO3dW6dKl1a5dOw0cONCYU61aNc2ZM0cxMTFavHixqlSpogkTJqhZs2bGnCeeeEIpKSmaPn26EhMTVa9ePc2bN89usZI71QIAAAAARc1isxWXuzf/PJKSXP9grLv7jYcx20zfroNn0+zGGlQtrw0Dmyk11fUPY8I8LBbJz6+cKfoXxQM9A2fRM3AWPYOCFJd/5+b1752Y7hk2AAAAAMANBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJuXu6gIAAChprFaLrFZLgWO5uTbl5truckUAALMisAEAcBdZrRZ5Vygjd7eCb3LJzsnVpYsZhDYAgCQX3xI5Z84cdejQQaGhoQoPD9fLL7+sEydO2M25fv26xo4dq7CwMIWGhuqVV15RUlKS3ZyzZ8+qd+/eatiwocLDw/XOO+8oOzvbbs7u3bvVrl07BQYG6rHHHtOqVavy1RMbG6vmzZsrKChIHTt21P79+52uBQCA27FaLXJ3s2rQ8ni1mb7d7mvQ8ni5u1lvefUNAFDyuDSwfffdd3rhhRe0YsUKLViwQNnZ2erZs6cyMjKMOdHR0frmm280depULVmyRBcuXNCAAQOM8ZycHPXp00dZWVlavny5Jk2apNWrV2v69OnGnNOnT6tPnz4KCwvT2rVr9eKLL2rUqFHavn27MWfjxo2KiYlR//79tXr1atWtW1c9e/ZUcnKyw7UAAOCoYxeu6ODZNLuvYxeuuLosAIDJuDSwffzxx2rfvr0eeOAB1a1bV5MmTdLZs2d18OBBSdLly5e1cuVKjRgxQuHh4QoMDFR0dLTi4+OVkJAgSdqxY4eOHTumd999V/Xq1VNUVJQGDRqk2NhYZWZmSpKWL18uf39/jRgxQrVq1VKXLl30+OOPa+HChUYtCxYsUKdOndShQwfVrl1bY8eOValSpbRy5UqHawEAAACAomSqZ9guX74sSfL29pYkHThwQFlZWYqIiDDm1KpVS1WrVlVCQoJCQkKUkJCgOnXqyM/Pz5gTGRmpMWPG6NixY6pfv74SEhIUHh5ud67IyEhFR0dLkjIzM3Xw4EH16dPHGLdarYqIiFB8fLzDtTjKYoI7XRytwQy1whzyeoGegKPomd+nJP7c6Bk4i55BYVksru8bR89vmsCWm5ur6OhoNWrUSHXq1JEkJSUlycPDQ+XLl7eb6+vrq8TERGPOzWFNkvH6TnOuXLmia9eu6dKlS8rJyZGvr2++8+Q9U+dILY7y9S3n1HxX8fEp6+oSYELFpX9hHvSM80r67196Bs6iZ+CsChWKz+9Z0wS2sWPH6ueff9Ynn3zi6lL+cMnJl2Vz8eJf7u7WOzZqamq6cnJy71JFMDuL5cZfiGboXxQP9EzB3NysdwxkJfX3Lz0DZ9EzKIgjv2cvXkxXdrZrf8/m9e+dmCKwjRs3Tv/617+0dOlSValSxdju5+enrKwspaWl2V3ZSk5OVsWKFY05v13NMW/lxpvn/HY1x6SkJHl5ealUqVKyWq1yc3OzW2Ak7zx5V+YcqcVRNptc/kvF0fO7uk6Yjxn6F8ULPVM4JflnRs/AWfQMnFWcesali47YbDaNGzdOW7Zs0aJFi1StWjW78cDAQHl4eCguLs7YduLECZ09e9Z4ZiwkJERHjx61C1s7d+6Ul5eXateubczZtWuX3bF37txpHMPT01MNGjSwO09ubq7i4uIUGhrqcC0AAAAAUJRceoVt7NixWr9+vT788EOVLVvWeBasXLlyKlWqlMqVK6cOHTpo0qRJ8vb2lpeXlyZMmKDQ0FAjJEVGRqp27doaNmyYXn/9dSUmJmrq1Kl64YUX5OnpKUl69tlnFRsbq8mTJ6tDhw7atWuXNm3apDlz5hi19OjRQ8OHD1dgYKCCg4O1aNEiXb16Ve3btzdqulMtAAAAAFCUXBrYli1bJknq2rWr3faYmBgjKI0cOVJWq1UDBw5UZmamIiMjNXr0aGOum5ubZs+erTFjxqhz584qXbq02rVrp4EDBxpzqlWrpjlz5igmJkaLFy9WlSpVNGHCBDVr1syY88QTTyglJUXTp09XYmKi6tWrp3nz5tktVnKnWgAAAACgKFlstuJy9+afR1KS6x+MdXe/8TBmm+nbdfBsmt1Yg6rltWFgM6Wmuv5hTJiHxSL5+ZUzRf+ieKBnCsbv31ujZ+AsegYFKS6/Z/P6905c+gwbAAAAAODWCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrABgAAAAAmRWADAAAAAJMisAEAAACASRHYAAAAAMCkCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrABgAAAAAmRWADAAAAAJMisAEAAACASRHYAAAAAMCkCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrABgAAAAAmRWADAAAAAJMisAEAAACASRHYAAAAAMCkCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrABgAAAAAmRWADAAAAAJMisAEAAACASRHYAAAAAMCkCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrABgAAAAAmRWADAAAAAJNyaWDbs2eP+vbtq8jISAUEBGjr1q124zabTdOmTVNkZKSCg4PVvXt3nTx50m7OxYsXNWTIEDVq1EhNmjTRyJEjlZ6ebjfn8OHDev755xUUFKSoqCjNnTs3Xy2bNm1Sq1atFBQUpLZt2+rbb791uhYAAAAAKEouDWwZGRkKCAjQ6NGjCxyfO3eulixZojFjxmjFihUqXbq0evbsqevXrxtzhg4dqmPHjmnBggWaPXu29u7dq7ffftsYv3Llinr27KmqVatq1apVGjZsmD744AN9+umnxpzvv/9eQ4YM0TPPPKM1a9bo0UcfVf/+/XX06FGnagEAAACAouTSwBYVFaVXX31Vjz32WL4xm82mxYsXq1+/fmrRooXq1q2ryZMn68KFC8aVuOPHj2v79u2aMGGCGjZsqCZNmmjUqFHasGGDzp8/L0n64osvlJWVpejoaD3wwANq06aNunbtqgULFhjnWrx4sZo1a6ZevXqpVq1aGjx4sOrXr6+lS5c6XAsAAAAAFDV3VxdwK2fOnFFiYqIiIiKMbeXKlVPDhg0VHx+vNm3aKD4+XuXLl1dQUJAxJyIiQlarVfv379djjz2mhIQENWnSRJ6ensacyMhIzZ07V5cuXZK3t7cSEhLUvXt3u/NHRkYaYcyRWpxhsTg1/Q/haA1mqBXmkNcL9AQcRc/8PiXx50bPwFn0DArLYnF93zh6ftMGtsTEREmSr6+v3XZfX18lJSVJkpKSknTvvffajbu7u8vb29vYPykpSf7+/nZz/Pz8jDFvb28lJSUZ2wo6jyO1OMPXt5zT+7iCj09ZV5cAEyou/QvzoGecV9J//9IzcBY9A2dVqFB8fs+aNrD9mSUnX5bN5toa3N2td2zU1NR05eTk3qWKYHYWy42/EM3Qvyge6JmCublZ7xjISurvX3oGzqJnUBBHfs9evJiu7GzX/p7N6987MW1gq1ixoiQpOTlZlSpVMrYnJyerbt26km5cKUtJSbHbLzs7W5cuXTL29/Pzy3cVLO913lW1guYkJycb447U4gybTS7/peLo+V1dJ8zHDP2L4oWeKZyS/DOjZ+AsegbOKk49Y9rPYfP391fFihUVFxdnbLty5Yp++OEHhYaGSpJCQ0OVlpamAwcOGHN27dql3NxcBQcHS5JCQkK0d+9eZWVlGXN27typmjVrytvb25iza9cuu/Pv3LlTISEhDtcCAAAAAEXNpYEtPT1dhw4d0qFDhyTdWNzj0KFDOnv2rCwWi7p166ZZs2Zp27ZtOnLkiIYNG6ZKlSqpRYsWkqRatWqpWbNmeuutt7R//37t27dP48ePV5s2bVS5cmVJUtu2beXh4aE333xTP//8szZu3KjFixerR48eRh3dunXT9u3bNX/+fB0/flwzZszQgQMH1KVLF0lyqBYAAAAAKGouvSXywIED6tatm/E6JiZGktSuXTtNmjRJL730kq5evaq3335baWlpaty4sebNm6d77rnH2GfKlCkaP368XnzxRVmtVrVs2VKjRo0yxsuVK6ePP/5Y48aNU/v27eXj46OXX35ZnTt3NuY0atRIU6ZM0dSpU/X++++rRo0amjlzpurUqWPMcaQWAAAAAChKLg1sYWFhOnLkyC3HLRaLBg0apEGDBt1yToUKFfTee+/d9jx169bVJ598cts5rVu3VuvWrX9XLQAAAABQlEy76AgAlHRWq0VWa8Ef0pKba1NubjF5WhoAABQagQ0ATMhqtci7Qhm5uxX8qHF2Tq4uXcwgtAEA8CdHYAMAE7JaLXJ3s2rQ8ngdu3DFbqx2JS9NezZUVquFwAYAwJ8cgQ0ATOzYhSs6eDbN1WUAAAAXMe3nsAEAAABASUdgAwAAAACT4pZIAAAAJ7GKK4C7hcAGAADgBFZxBXA3EdgAAACcwCquAO4mAhsAAEAhsIorgLuBRUcAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJubu6AAAoDKvVIqvVUuBYbq5Nubm2u1wRAABA0SOwASh2rFaLvCuUkbtbwTcJZOfk6tLFDEIbAAAo9ghsAIodq9UidzerBi2P17ELV+zGalfy0rRnQ2W1WghsAACg2COwASi2jl24ooNn01xdBgAAwB+GwAYAAACn8BwxcPcQ2AAAAOAwniMG7i4CGwAAABzGc8TA3UVgAwAAgNN4jhi4O/jgbAAAAAAwKQIbAAAAAJgUt0QCcIrbLR4yZ1Uw3Ao9A2fRMwDwPwQ2AA6xWi3KybXJx6dsgeOsCobfomfgLHoGhUXIx58Zgc1JsbGx+vjjj5WYmKi6devqrbfeUnBwsKvLAv5wFotFblYLq4LBYfQMnEXPwFmEfJQEBDYnbNy4UTExMRo7dqwaNmyoRYsWqWfPntq8ebN8fX1dXR5wV7AqGJxFz8BZ9AwcRchHSUBgc8KCBQvUqVMndejQQZI0duxY/etf/9LKlSvVu3dvF1eHkshqtchqtRQ4xm0gzuPnaR68F8UD7xPMoqhDPr0NMyGwOSgzM1MHDx5Unz59jG1Wq1URERGKj4936lhWq2Rz8X/nlv/3O6hB1fIq7elmN3a/343bCm51P7jN9r/9i+uY2eopzJjFYpFXuVJyv8X7lJ2Tq/Qr1wr8S6Uw53N3v3EeM/RM3nmKspY/6ud5p/G7+T3e7bHC9szd7u27PfZHvbd/RK23G/sj3icz/Z653dif4b/P3zNmpu//j+iZP/vvoOI0Vth9HelRi+XGv8ld6Xbft908m83V0aF4OH/+vB5++GEtX75coaGhxvbJkydrz549+uyzz1xYHQAAAIA/Iz6HDQAAAABMisDmIB8fH7m5uSk5Odlue3Jysvz8/FxUFQAAAIA/MwKbgzw9PdWgQQPFxcUZ23JzcxUXF2d3iyQAAAAAFBUWHXFCjx49NHz4cAUGBio4OFiLFi3S1atX1b59e1eXBgAAAOBPiMDmhCeeeEIpKSmaPn26EhMTVa9ePc2bN49bIgEAAAD8IVglEgAAAABMimfYAAAAAMCkCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrAVkLFxsaqefPmCgoKUseOHbV//35XlwQTmDNnjjp06KDQ0FCFh4fr5Zdf1okTJ+zmXL9+XWPHjlVYWJhCQ0P1yiuvKCkpyUUVw2w++ugjBQQEaOLEicY2ega/df78eQ0dOlRhYWEKDg5W27Zt9eOPPxrjNptN06ZNU2RkpIKDg9W9e3edPHnSdQXDpXJycjR16lQ1b95cwcHBatGihWbOnKmbFzqnZ0q2PXv2qG/fvoqMjFRAQIC2bt1qN+5If1y8eFFDhgxRo0aN1KRJE40cOVLp6el38bu4NQJbCbRx40bFxMSof//+Wr16terWrauePXsqOTnZ1aXBxb777ju98MILWrFihRYsWKDs7Gz17NlTGRkZxpzo6Gh98803mjp1qpYsWaILFy5owIABLqwaZrF//34tX75cAQEBdtvpGdzs0qVLeu655+Th4aG5c+dqw4YNGj58uLy9vY05c+fO1ZIlSzRmzBitWLFCpUuXVs+ePXX9+nUXVg5XmTt3rpYtW6a3335bGzdu1NChQzVv3jwtWbLEbg49U3JlZGQoICBAo0ePLnDckf4YOnSojh07pgULFmj27Nnau3ev3n777bv1LdyeDSXOM888Yxs7dqzxOicnxxYZGWmbM2eOC6uCGSUnJ9vq1Klj++6772w2m82WlpZma9CggW3Tpk3GnGPHjtnq1Klji4+Pd1GVMIMrV67YWrZsafv3v/9t69Kli23ChAk2m42eQX7vvvuu7bnnnrvleG5urq1p06a2efPmGdvS0tJsgYGBtvXr19+NEmEyvXv3tr3xxht22wYMGGAbMmSIzWajZ2CvTp06ti1bthivHemPvL+X9u/fb8z59ttvbQEBAbb//ve/d6/4W+AKWwmTmZmpgwcPKiIiwthmtVoVERGh+Ph4F1YGM7p8+bIkGf/n+8CBA8rKyrLrn1q1aqlq1apKSEhwRYkwiXHjxikqKsquNyR6Bvl9/fXXCgwM1MCBAxUeHq6nn35aK1asMMbPnDmjxMREu54pV66cGjZsyN9TJVRoaKh27dql//znP5Kkw4cPa9++fXr44Ycl0TO4PUf6Iz4+XuXLl1dQUJAxJyIiQlar1RSPDbm7ugDcXampqcrJyZGvr6/ddl9f33zPKqFky83NVXR0tBo1aqQ6depIkpKSkuTh4aHy5cvbzfX19VViYqIryoQJbNiwQT/99JM+//zzfGP0DH7r9OnTWrZsmXr06KG+ffvqxx9/1IQJE+Th4aF27doZfVHQ31M8+1gy9e7dW1euXFHr1q3l5uamnJwcvfrqq/rHP/4hSfQMbsuR/khKStK9995rN+7u7i5vb29T/F1FYANQoLFjx+rnn3/WJ5984upSYGLnzp3TxIkTNX/+fN1zzz2uLgfFgM1mU2BgoF577TVJUv369fXzzz9r+fLlateunYurgxlt2rRJ69at03vvvafatWvr0KFDiomJUaVKlegZlAjcElnC+Pj4yM3NLd8CI8nJyfLz83NRVTCbcePG6V//+pcWLVqkKlWqGNv9/PyUlZWltLQ0u/nJycmqWLHi3S4TJnDw4EElJyerffv2ql+/vurXr6/vvvtOS5YsUf369ekZ5FOxYkXVqlXLbtv999+vs2fPGuOS+HsKhsmTJ6t3795q06aNAgIC9PTTT+vFF1/UnDlzJNEzuD1H+sPPz08pKSl249nZ2bp06ZIp/q4isJUwnp6eatCggeLi4oxtubm5iouLU2hoqAsrgxnYbDaNGzdOW7Zs0aJFi1StWjW78cDAQHl4eNj1z4kTJ3T27FmFhITc5WphBg899JDWrVunNWvWGF+BgYFq27at8Wd6Bjdr1KiR8SxSnpMnT+qvf/2rJMnf318VK1a065krV67ohx9+4O+pEuratWuyWCx229zc3Ixl/ekZ3I4j/REaGqq0tDQdOHDAmLNr1y7l5uYqODj4rtf8W9wSWQL16NFDw4cPV2BgoIKDg7Vo0SJdvXpV7du3d3VpcLGxY8dq/fr1+vDDD1W2bFnjvu1y5cqpVKlSKleunDp06KBJkybJ29tbXl5emjBhgkJDQ/nHdwnl5eVlPOOYp0yZMqpQoYKxnZ7BzV588UU999xzmj17tlq3bq39+/drxYoVGjdunCTJYrGoW7dumjVrlu677z75+/tr2rRpqlSpklq0aOHi6uEKjzzyiGbPnq2qVasat0QuWLBAHTp0kETPQEpPT9epU6eM12fOnNGhQ4fk7e2tqlWr3rE/atWqpWbNmumtt97S2LFjlZWVpfHjx6tNmzaqXLmyq74tg8Vmu+lTB1FiLF26VB9//LESExNVr149jRo1Sg0bNnR1WXCx335+Vp6YmBgj0F+/fl2TJk3Shg0blJmZqcjISI0ePdoUtwzAHLp27aq6devqzTfflETPIL9vvvlG77//vk6ePCl/f3/16NFDnTp1MsZtNpumT5+uFStWKC0tTY0bN9bo0aNVs2ZNF1YNV7ly5YqmTZumrVu3Kjk5WZUqVVKbNm3Uv39/eXp6SqJnSrrdu3erW7du+ba3a9dOkyZNcqg/Ll68qPHjx+vrr7+W1WpVy5YtNWrUKJUtW/ZufisFIrABAAAAgEnxDBsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApd1cXAAB/RiNGjNDq1atvOb5nzx6VL1/+LlYEAACKIwIbAPxBmjVrppiYGLtt8fHxeuWVV1xUEQAAKG64JRIA/iCenp6qWLGi3Ze3t7fdnFWrVqlJkybaunWrWrZsqaCgIPXs2VPnzp2zm/fJJ5+oRYsWCgwM1OOPP641a9bYjQcEBGjr1q1227p27aqJEyfabfvss8/UunVrBQUFqVWrVoqNjTXGzpw5o4CAAB06dMhun+bNm2vhwoXGnwMCAgr8WrVqVYE/hxEjRiggIEALFiyw296/f/98+507d06DBg1SkyZN9OCDD6pfv346c+aMJGnGjBm3PHfXrl2Nc7388svG8VJTU/W3v/1NTZo0sTt33vf626+0tDTjXE899ZQxPzMzU4899pjdnKJ87/K+GjVqpB49eujUqVPG+Jo1a9S+fXuFhoaqadOmGjJkiJKTk43x3bt329V183HzesKR9/a3+/zWU089pRkzZhjnDAwM1N69e43xuXPnKjw8XElJSQXun/fzullBdf22hs8++0wBAQFGL3ft2vWWfZBX36VLlzRs2DD97W9/U8OGDdWrVy+dPHnSOObGjRvVokULBQUFKSwsTAMHDlRKSorDNRT0s5Py919mZqYmTJig8PBwBQUF6bnnntP+/fuN8bz3LiAgQHXr1lV4eLhGjhyp69evF/gzBFAyEdgAwMWuXbumWbNm6Z133tGyZcuUlpamV1991RjfsmWLoqOj1aNHD61bt07PPvusRo4cqV27djl1ni+++ELTpk3Tq6++qo0bN+q1117T9OnTb3vr5m99/vnn2rFjh3bs2KEqVapo5MiRxusnnnjilvtVrlxZK1asMF6fP39e33//vUqXLm1sy8rKUs+ePVW2bFnFxsZq2bJlKlOmjHr16qXMzEz93//9n3Gu//u//1NoaKjxOu8f6r81c+ZMZWdn59tus9kkSQsXLrzt/nliY2MLDCJF9d7FxMRox44dio2NVXJyst5//31jLDs7W4MGDdIXX3yhmTNn6tdff9WIESNuW+8fLSwsTN26ddOwYcN0+fJl/fTTT5o2bZomTJggPz+/IjtPRkaGpk2bpjJlyhjbZsyYYbzvoaGh+fpCuhGcDhw4oFmzZunTTz+VzWZT7969lZWVJUm6//77NWnSJG3evFnz5s3TmTNn9O677zpcg6MmT56sL7/8UpMmTdLq1at13333qVevXrp48aLdvM2bN+v/+//+P7377rvatGmTVq5c6fS5APx5cUskALhYVlaW3n77bTVs2FCSNGnSJD3xxBPav3+/goOD9fHHH6tdu3Z64YUXJEk1a9ZUQkKC5s+fr4ceesjh88yYMUMjRoxQy5YtJUnVqlXTsWPH9Omnn6pdu3YOHePee+81/uzm5qZy5cqpYsWKd9wvMDBQZ86c0d69e9WkSROtXLlSbdq0sbvatHHjRuXm5mrixImyWCySbgSZv/3tb/ruu+8UGRmpsmXLSpLKlCkjDw+P2577P//5j1auXKnu3btryZIldmN5Ic7Pz6/AK583u3jxombNmqWXXnpJ06ZNsxsrqveufPnyqlixokqVKqWyZcuqXLlyxtgzzzxj/LlatWp688039cwzzyg9Pd34ebjC4MGDtXPnTr311lv6+eef1a5dOz366KNFeo558+apdu3aysnJMbZVqFDB+LOHh4fKlClj1wcnT57U119/rWXLlqlRo0aSpClTpujvf/+7tm7dqtatW6tu3brGfG9vb/n4+Cg3N9fhGhyRkZGh5cuXKyYmRlFRUZKk8ePH69///rc+//xz9erVy5jr6+ur8uXL6+rVq/Lw8LB7/wGAwAYALubu7q6goCDjda1atVS+fHkdP35cwcHBOnHihDp37my3T6NGjbR48WKHz5GRkaFTp07pzTff1FtvvWVsz87OzvePw2effVZW6/9uwLh69aqz31KBOnXqpE8//VSNGjXS559/rlmzZtkFtsOHD+vUqVPGP7LzXL9+3e4WQUe9++676ty5s6pVq5Zv7MqVK5Jkd4XvVmbOnKmwsDA1btw431hRvXevvfaa3NzcdPXqVdWpU0evvfaaMXbgwAF98MEHOnz4sC5dumRcHTx37pxq165tzMsLBbfjyHubV0vZsmVVv359DRs2zO48eTw9PTVlyhT94x//UNWqVfXGG2/c8fzOOH/+vBYsWKBly5blu7X3do4fPy53d3cjREuSj4+PatasqePHjxvb9u7dq5deeklXr15V48aNNXz4cKdrmDJlil2Iz8zMNN6HU6dOKSsry66fPTw8FBwcbFeHdOO9s9lsunr1qtq2basnn3zS4e8XwJ8fgQ0ASoCMjAxJN/4P/83/kJVk9w94SfrnP/+pWrVqGa/zng/7vf7xj39oxowZ2rBhgypWrKiAgIB8NTZo0EBTpkzJt+/NV/Yc8d1332nfvn2KiYnRtm3b8o1fuHBBVqv1jlcHT548qc8//1xr1qzRf//7X6dqcMYbb7yhiIgIpaWlaerUqXrjjTc0e/ZsZWRkqGfPnoqMjNSUKVPk4+Ojc+fOqWfPnsbtfXliY2PtrrjlXUm9mSPv7c21/POf/9TgwYO1fv36AuuOj4+XdOOZsUuXLhXqtsFbmTp1qlq1amV3NawoBQYGavXq1Tp79qwmTpyozz77TH369HGqhp49e6p9+/bG6ylTpjh9JU7633t35swZvfXWW1qwYIFxeycA8AwbALhYdna2Dhw4YLw+ceKE0tLSjH9Y33///fr+++/t9vn+++8LvOpxK35+fqpUqZJOnz6t++67z+7rt1eg/vKXv9iNu7sXzf/bK1++vJo3b67Ro0erY8eO+cYbNGigX375Rb6+vvlqdOYWMZvNpkmTJqlfv363vNXxxx9/1P3336977rnntseaMmWKnnnmGd13330FjhfVe1exYkXdd999CgoKUpcuXfSvf/1LWVlZOnHihC5evKihQ4eqSZMmqlWrlt2CIzfz9/e3+5kVxJH39uZaunXrpmPHjuULh9KNK0jR0dEaP368goODNXz48FveVuisQ4cOafPmzRo8eLDT+9aqVUvZ2dn64YcfjG2pqan6z3/+Y/dzL1WqlGrUqKGIiAh16tRJ69atc7oGHx8fu5/nzYG5evXq8vDwsHv/s7Ky9OOPP+Z7//Peu6ZNm+rxxx/Xli1bnP6+Afx5cYUNAFzMw8ND48eP16hRo+Tm5qbx48crJCREwcHBkqRevXpp8ODBqlevniIiIvTNN99oy5Yt+VZdzMrKsltdLjc3V7m5ucrOzpa7u7sGDhyoCRMmqFy5cmrWrJkyMzN14MABpaWlqUePHnfle33ppZdUs2bNAhcoadu2rT7++GP169dPgwYNUuXKlXX27Flt2bJFvXr1UpUqVRw6x65du1SxYkXjubGbZWZmauPGjVq4cOEdP17h1KlTOnfunL766qtbzimq9y4tLU2JiYm6fPmyPv/8c1WrVk0eHh6qWrWqPDw8tGTJEj333HM6evSoPvzwQ4d+DoWV10dpaWn64osvVKNGDXl4eNjNycnJ0euvv65mzZqpQ4cOatasmdq2bav58+fbPZtVkJt7NDMz0zinzWYznl2cP3++evToocqVKztdf40aNfToo4/qrbfe0tixY+Xl5aUpU6aocuXKxjN2GzZsULVq1eTn56dffvlFy5cvV2BgoN1xfk8N0o3nLJ977jlNnjxZ3t7eqlq1qubNm6dr167ZPZcoScnJybp+/brOnj2rr7/+Ot9qmgBKNgIbALhYqVKl9NJLL2nIkCE6f/68mjRpYve8TIsWLTRy5EjNnz9f0dHR+utf/6ro6GiFhYXZHaegKwF79+5VhQoV9Morr6hjx44qVaqUPv74Y02ePFllypRRnTp19OKLL/7R36Lh/vvvV+/evQscK126tJYuXaopU6ZowIABSk9PV+XKlRUeHi4vLy+Hz5GRkaEhQ4bkCxmSdPToUX3wwQd6+eWX7xhSMzIyNHz4cLtFLn6rqN67vOe/ypYtqwYNGmj69OmSbtwKOmnSJL3//vtasmSJGjRooOHDh6tfv36O/jicltdHebVMnTo135xZs2bp119/1ezZsyVJlSpV0vjx4/Xaa68pMjLylrcQXr582QizN+vYsaO2bdsmf39/49x3Cn63ExMTo4kTJ6pv377KyspSkyZN9NFHHxk9cfz4cb377rtKSkqSj4+PHn744XzPsP3eGiRp6NChstlsGjZsmNLT0xUYGKh58+blu/LbqlUrSTeu2DVt2lSvv/767zovgD8Xiy3v6WUAwF23atUqRUdH232eVVFauHChLl++zId1/wH+6PeuJHnqqac0c+ZMI7ABAP6HZ9gA4E/M3d1dbm5uri4DuC1PT0/jdkgAgD1uiQSAP7EuXbq4ugTgjj777DNXlwAApsUtkQAAAABgUtwSCQAAAAAmRWADAAAAAJMisAEAAACASRHYAAAAAMCkCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrABgAAAAAmRWADAAAAAJMisAEAAACASRHYAAAAAMCkCGwAAAAAYFIENgAAAAAwKQIbAAAAAJgUgQ0AAAAATIrABgC4o1WrVikgIMD4CgoK0uOPP65x48YpKSnJ1eXBhEaMGKHQ0NACx7Zs2aJevXopLCxMgYGBioyM1KBBgxQXF2fM2b17t9Fva9euLfA4zz77rAICAvTkk0/+Id8DAJiBu6sLAAAUHwMHDpS/v78yMzO1b98+LVu2TN9++63Wr1+v0qVLu7o8mJzNZtPIkSO1atUq1a9fXz169JCfn58SExO1ZcsWde/eXcuWLVOjRo2Mfe655x6tX79eTz31lN2xzpw5o/j4eN1zzz13+9sAgLuKwAYAcNjDDz+soKAgSVLHjh1VoUIFLViwQNu2beMqB+5o/vz5WrVqlV588UW98cYbslgsxli/fv20Zs0aubvb/9MkKipKX3/9tVJSUnTvvfca29evXy8/Pz/dd999SktLu2vfAwDcbdwSCQAotIceekjSjasdknTx4kW98847atu2rUJDQ9WoUSP16tVLhw8fzrfv9evXNWPGDD3++OMKCgpSZGSkBgwYoFOnThnHvPk2zN9+de3a1ThW3u1zGzdu1Pvvv6+mTZsqJCREffv21blz5/Kd+4cfflDPnj3VuHFjNWzYUF26dNG+ffsK/B67du1a4PlnzJiRb+7atWvVvn17BQcH68EHH9Srr75a4Plv973dLDc3VwsXLlSbNm0UFBSkiIgIvf3227p06ZLdvObNm6tPnz75zjNu3Lh8xyyo9nnz5uX7mUpSZmampk+frscee0yBgYGKiorS5MmTlZmZWeDP6nauXbumjz76SPfff7+GDx9uF9byPP300woODrbb9uijj8rT01ObN2+2275+/Xq1bt1abm5uTtcCAMUJV9gAAIWWF64qVKggSTp9+rS2bt2qVq1ayd/fX0lJSfr000/VpUsXbdiwQZUrV5Yk5eTkqE+fPoqLi1ObNm3UrVs3paen69///reOHj2q6tWrG+d48skn9fDDD9ud9/333y+wnlmzZsliseill15ScnKyFi1apO7du2vt2rUqVaqUJCkuLk4vvfSSAgMDNWDAAFksFuOqzyeffJIvMEhSlSpV9Nprr0mSMjIyNGbMmALPPW3aNLVu3VrPPPOMUlJStHTpUr3wwgtas2aNypcvn2+fzp07q3HjxpJuPNe1ZcsWu/G3335bq1evVvv27dW1a1edOXNGsbGx+umnn7Rs2TJ5eHgU+HNwRlpamj766KN823Nzc9WvXz/t27dPnTp1Uq1atXT06FEtWrRIJ0+e1IcffujUefbt26eLFy+qW7duToWsUqVKqXnz5tqwYYOef/55SdLhw4f1888/a8KECTpy5IhTdQBAcUNgAwA47MqVK0pJSVFmZqa+//57zZw5U6VKldIjjzwi6cbVmy+//FJW6/9u4HjqqafUunVrff755+rfv78kac2aNYqLi9Mbb7yh7t27G3N79+4tm81md8769evne35p7ty5BdZ36dIlbdy4UV5eXsa+gwcP1ooVK9StWzfZbDaNGTNGYWFhmjdvnnGV59lnn1WbNm00depUzZ8/3+6YWVlZKl++vFFDSkpKvsD266+/asaMGRo8eLD69u1rbG/ZsqXatWunTz75xG57Tk6OJKlRo0bGcU+dOmUX2Pbu3avPPvtMU6ZMUdu2bY3tYWFh6tWrlzZv3my3vbDmzJkjd3d3NWjQwG77unXrtHPnTi1ZskRNmjQxtj/wwAMaPXq0vv/+e7tnze7k+PHjkpTvip8j2rZta1wt/ctf/qIvvvhC1apVU0hIiNPHAoDihlsiAQAO6969u8LDwxUVFaVXX31VZcuW1QcffGBcOfP09DTCWk5OjlJTU1WmTBnVrFlTP/30k3Gcr776Sj4+PurSpUu+cxR0q5yjnn76aSOsSVKrVq1UsWJFffvtt5KkQ4cO6eTJk2rbtq1SU1OVkpKilJQUZWRkKDw8XHv27FFubq7dMTMzM+Xp6Xnb827ZskW5ublq3bq1ccyUlBTjGavdu3fbzc/KypKk2x538+bNKleunJo2bWp3zAYNGqhMmTL5jpmdnW03LyUlRdevX79t3efPn9fSpUv18ssvq2zZsvnOX6tWLd1///12x8y7Dfa357+TK1euSFK+8ziiadOm8vb21oYNG2Sz2bRx40a1adPG6eMAQHHEFTYX2LNnjz7++GMdOHBAiYmJmjlzplq0aOHUMWw2m+bPn68VK1bo119/lY+Pj55//nn169fvD6oaAG7colezZk25ubnJz89PNWvWtLualpubq8WLF+uTTz7RmTNnjCtJ0v9um5RuXE2qWbNmvgUmfq/77rvP7rXFYtF9992nX3/9VZJ08uRJSdLw4cNveYzLly/L29vbeJ2amprvuL918uRJ2Ww2tWzZssDx336feYtklClT5pbH/OWXX3T58mWFh4cXOJ6cnGz3eseOHbeceyvTp09XpUqV1LlzZ3355Zf5zn/8+HGHz38neUE6PT3dqf0kycPDQ61atdL69esVHBysc+fOFcnVRQAoDghsLpCRkaGAgAB16NBBAwYMKNQxJk6cqB07dmjYsGGqU6eOLl26lO8hdAAoasHBwcYqkQWZPXu2pk2bpg4dOmjQoEHy9vaW1WpVdHR0vlsdXSGvhmHDhqlevXoFzrk5RGVmZioxMVERERG3PW5ubq4sFovmzp1b4PNZvw1meZ9d5+fnd9tj+vr6asqUKQWO37xioiQ1bNhQgwcPttu2dOlSbdu2rcD9jx8/rtWrV+vdd98t8Fm43Nxc1alTR2+88UaB+1epUuWWtRfk/vvvlyQdOXLE6f9JKd24LXL58uWaMWOG6tatq9q1azt9DAAojghsLhAVFaWoqKhbjmdmZuqf//yn1q9fr8uXL+uBBx7Q0KFDFRYWJunGX7LLli3TunXrjL8Aq1WrdldqB4Db+fLLLxUWFqbo6Gi77WlpafLx8TFeV69eXT/88IOysrKKZOGMPL/88ovda5vNpl9++cV4birvd6WXl9cdQ5h0Y3GLrKwsBQYG3nZe9erVZbPZ5O/vr5o1a97xuMeOHZPFYrnt3OrVqysuLk6NGjUyFky5HR8fn3zf09atW285/7333lPdunX1xBNP3PL8hw8fVnh4+O+6TTVP48aNjdsa+/bt6/Tqjo0bN1bVqlX13XffaejQob+7HgAoLniGzYTGjRun+Ph4/fOf/9QXX3yhVq1aqVevXsatPF9//bX8/f31r3/9S82bN1fz5s315ptv6uLFiy6tGwDc3NzyXUnbtGmTzp8/b7etZcuWSk1NVWxsbL5j/J4rcWvWrDGelZJuPIeVmJhorDIZGBio6tWra/78+QXempeSkmL3evPmzXJzczMWVbmVli1bys3NTR988EG++m02m1JTU43X2dnZ+uqrrxQcHHzb57lat26tnJycAldjzM7O/l2fPZaQkKBt27Zp6NChtwxjrVu31vnz57VixYp8Y9euXVNGRoZT5yxdurR69eql48ePa8qUKQW+z2vXrtX+/fsL3N9isejNN9/UgAED8i1CAwB/ZlxhM5mzZ89q1apV+uabb4yH+Hv27Knt27dr1apVeu2113T69GmdPXtWmzdv1uTJk5WTk6OYmBgNHDhQixcvdvF3AKAk+/vf/66ZM2fqjTfeUGhoqI4ePap169bluwvg6aef1po1axQTE6P9+/ercePGunr1quLi4vTcc88V6pY5SfL29tbzzz+v9u3bG8v633ffferUqZMkyWq1asKECXrppZf05JNPqn379qpcubLOnz+v3bt3y8vLS7Nnz1ZGRoZiY2O1ZMkS1ahRw26BjbygcuTIEcXHxys0NFTVq1fX4MGD9d577+nXX39VixYtVLZsWZ05c0Zbt25Vp06d1LNnT+3cuVPTpk3TkSNHNHv27Nt+Lw8++KA6d+6sOXPm6NChQ2ratKk8PDx08uRJbd68WW+++aZatWpVqJ/Tjh071LRp09teZXzqqae0adMmjR49Wrt371ajRo2Uk5OjEydOaPPmzZo3b95tb48tSK9evXTs2DHNnz9fu3fv1uOPPy4/Pz8lJSVp69at2r9/v5YvX37L/Vu0aFHo3gCA4orAZjJHjx5VTk5Ovr+EMzMzjQf2bTabMjMz9c477xi300ycOFHt27fXiRMnjNskAeBu69u3r65evap169Zp48aNql+/vubMmaP33nvPbp6bm5vmzp2rWbNmaf369frqq69UoUIFNWrUqFDLvt98/iNHjuijjz5Senq6wsPDNXr0aJUuXdqYExYWpk8//VQffvihli5dqoyMDFWsWFHBwcHq3LmzpBtX2vKeHTt+/LiGDRuW71xbtmyRl5eXQkNDJd34SIIaNWpo4cKFmjlzpqQbz3k1bdpUzZs3l3TjDgkPDw999NFHatas2R2/n3HjxikwMFDLly/XP//5T7m5uemvf/2r/vGPfzi1pP5vWSwWDRky5LZzrFarZs6cqYULF2rt2rXasmWLSpcuLX9/f3Xt2tWhWz8LOubkyZP16KOPasWKFZo/f76uXLkiHx8f/e1vf9Prr79u/DwBADdYbGZ4CrwECwgIsFslcuPGjRo6dKjWr1+f7/7+MmXKqGLFipo+fbrmzJmjgwcPGmPXrl1Tw4YNNX/+fDVt2vSufg8A4Gq7d+9Wt27dNG3atEJfdbrZmTNn9Oijj2rbtm3y9/cvcM6MGTP066+/atKkSb/7fAAA3ApX2EymXr16ysnJUUpKit0Hld6sUaNGys7O1qlTp1S9enVJ/1uqumrVqnerVAAAAAB/MAKbC6Snp+vUqVPG6zNnzujQoUPy9vZWzZo11bZtWw0bNkwjRoxQvXr1lJqaqri4OAUEBOjvf/+7IiIi1KBBA40cOVIjR45Ubm6uxo0bp6ZNmxbqFhUAgL0yZcqobdu2t/2ctICAAFWqVOkuVgUAKIm4JdIF8m7d+a127dpp0qRJysrK0qxZs7RmzRpduHBBFSpUUEhIiF555RXj2Y7z589rwoQJ2rFjh8qUKaOHH35Yw4cPt/tgWgAoKYr6lkgAAMyCwAYAAAAAJsXnsAEAAACASRHYAAAAAMCkCGwAAAAAYFIENgAAAAAwKZb1d4Hk5Mty9VIvFovk61vOFLWgeKBn4Cx6Bs6iZ+AsegbOMlPP5NVyJwQ2F7DZ5PIGyWOmWlA80DNwFj0DZ9EzcBY9A2cVp57hlkgAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApNxdXQBcy82t4Myem2tTbq7tLlcDAAAA4GYEthLKarUoJ9cmH5+yBY5n5+Tq0sUMQhsAAADgQgS2EspiscjNatGg5fE6duGK3VjtSl6a9myorFYLgQ0AAABwIQJbCXfswhUdPJvm6jIAAAAAFIBFRwAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSJTaw7dmzR3379lVkZKQCAgK0devW287/6quv1KNHDz300ENq1KiROnfurO3bt9+lagEAAACURCU2sGVkZCggIECjR492aP6ePXsUERGhjz76SKtWrVJYWJj69eunn3766Q+uFAAAAEBJ5e7qAlwlKipKUVFRDs9/88037V6/9tpr2rZtm77++mvVr1+/qMsDAAAAgJIb2H6v3Nxcpaenq0KFCk7va7EUfT1/VA1mqBXmkNcL9AQcRc/AWfQMnEXPwFlm6hlHayCwFdLHH3+sjIwMtW7d2ul9fX3L/QEVFT0fn7KuLgEmVFz6F+ZBz8BZ9AycRc/AWcWpZwhshbBu3TrNnDlTH374oXx9fZ3ePzn5smy2P6AwJ7i7W1Whwu0DWWpqunJycu9SRTA7i+XGLzcz9C+KB3oGzqJn4Cx6Bs4yU8/k1XInBDYnbdiwQaNGjdK0adMUERFRqGPYbHJ5gzh6flfXCfMxQ/+ieKFn4Cx6Bs6iZ+Cs4tQzJXaVyMJYv3693njjDb333nv6+9//7upyAAAAAPzJldgrbOnp6Tp16pTx+syZMzp06JC8vb1VtWpVvffeezp//rwmT54s6cZtkCNGjNDIkSPVsGFDJSYmSpJKlSqlcuWKzz2wAAAAAIqPEhvYDhw4oG7duhmvY2JiJEnt2rXTpEmTlJiYqHPnzhnjK1asUHZ2tsaNG6dx48YZ2/PmAwAAAEBRK7GBLSwsTEeOHLnl+G9D2JIlS/7okgAAAADADs+wAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJEdgAAAAAwKQIbAAAAABgUgQ2AAAAADApAhsAAAAAmBSBDQAAAABMisAGAAAAACZFYAMAAAAAkyKwAQAAAIBJldjAtmfPHvXt21eRkZEKCAjQ1q1b77jP7t271a5dOwUGBuqxxx7TqlWr7kKlAAAAAEqqEhvYMjIyFBAQoNGjRzs0//Tp0+rTp4/CwsK0du1avfjiixo1apS2b9/+B1cKAAAAoKRyd3UBrhIVFaWoqCiH5y9fvlz+/v4aMWKEJKlWrVrat2+fFi5cqGbNmv1RZQIAAAAowUpsYHNWQkKCwsPD7bZFRkYqOjra6WNZLEVVVeE5WoMZaoU55PUCPQFH0TNwFj0DZ9EzcJaZesbRGghsDkpKSpKfn5/dNj8/P125ckXXrl1TqVKlHD6Wr2+5oi7vD+HjU9bVJcCEikv/wjzoGTiLnoGz6Bk4qzj1DIHNBZKTL8tmc20N7u5WVahw+0CWmpqunJzcu1QRzM5iufHLzQz9i+KBnoGz6Bk4i56Bs8zUM3m13AmBzUF+fn5KSkqy25aUlCQvLy+nrq5Jks0mlzeIo+d3dZ0wHzP0L4oXegbOomfgLHoGzipOPVNiV4l0VkhIiHbt2mW3befOnQoJCXFNQQAAAAD+9EpsYEtPT9ehQ4d06NAhSdKZM2d06NAhnT17VpL03nvvadiwYcb8Z599VqdPn9bkyZN1/PhxxcbGatOmTerevbsrygcAAABQApTYWyIPHDigbt26Ga9jYmIkSe3atdOkSZOUmJioc+fOGePVqlXTnDlzFBMTo8WLF6tKlSqaMGECS/oDAAAA+MOU2MAWFhamI0eO3HJ80qRJBe6zZs2aP7AqAAAAAPifEntLJAAAAACYHYENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyKwAYAAAAAJkVgAwAAAACTIrABAAAAgEkR2AAAAADApAhsAAAAAGBSBDYAAAAAMCkCGwAAAACYFIENAAAAAEyqRAe22NhYNW/eXEFBQerYsaP2799/2/kLFy7U448/ruDgYEVFRSk6OlrXr1+/S9UCAAAAKGlKbGDbuHGjYmJi1L9/f61evVp169ZVz549lZycXOD8devW6b333tOAAQO0ceNGTZw4URs3btT7779/lysHAAAAUFKU2MC2YMECderUSR06dFDt2rU1duxYlSpVSitXrixwfnx8vBo1aqS2bdvK399fkZGRevLJJ+94VQ4AAAAACsvd1QW4QmZmpg4ePKg+ffoY26xWqyIiIhQfH1/gPqGhofriiy+0f/9+BQcH6/Tp0/r222/11FNPOX1+i6XQpRcZR2swQ60wh7xeoCfgKHoGzqJn4Cx6Bs4yU884WkOJDGypqanKycmRr6+v3XZfX1+dOHGiwH3atm2r1NRUPf/887LZbMrOztazzz6rvn37On1+X99yhar7bvPxKevqEmBCxaV/YR70DJxFz8BZ9AycVZx6pkQGtsLYvXu35syZo9GjRys4OFinTp3SxIkTNXPmTPXv39+pYyUnX5bN9gcV6iB3d6sqVLh9IEtNTVdOTu5dqghmZ7Hc+OVmhv5F8UDPwFn0DJxFz8BZZuqZvFrupEQGNh8fH7m5ueVbYCQ5OVl+fn4F7jNt2jT94x//UMeOHSVJAQEBysjI0Ntvv61+/frJanX8cUCbTS5vEEfP7+o6YT5m6F8UL/QMnEXPwFn0DJxVnHqmRC464unpqQYNGiguLs7Ylpubq7i4OIWGhha4z7Vr1/KFMjc3N0mSrbi82wAAAACKlRJ5hU2SevTooeHDhyswMFDBwcFatGiRrl69qvbt20uShg0bpsqVK2vIkCGSpEceeUQLFixQ/fr1jVsip02bpkceecQIbgAAAABQlEpsYHviiSeUkpKi6dOnKzExUfXq1dO8efOMWyLPnTtnd0WtX7//v737j6+5/v8/fj/nbPNjmNlmJfLzbdiPbJH8yjt5K7/qjSiFaEOiEqKUan5NfvQ2kh/5rUXLr5JRpHq/yxRRIn0Kya+9NTOGjf063z98d96Ozeyw7byOc7teLrtcnNfPxzmvh+3cz+v5ep3BMplMmjFjhk6ePKkqVaro/vvv14svvuispwAAAADgFmeyMp6v1J065fyLHD08zPL19Vanmf/RvhNpdvOCq1XShudbKzX1grKzuekILjOZJH//ioboX7gGegaOomfgKHoGjjJSz+TVcj1ueQ0bAAAAALgCAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKBc8ouz09PTtWPHDp04cUJZWVl28/r27eukqgAAAACgeLlcYPvll180cOBAZWRkKCMjQz4+PkpNTVW5cuVUpUoVAhsAAACAW4bLDYmMiYnR/fffrx07dqhMmTKKj4/Xl19+qeDgYI0ePdrZ5QEAAABAsXG5wLZ//371799fZrNZFotFmZmZuv322/XSSy/p7bffdnZ5AAAAAFBsXC6weXh4yGy+XLafn59OnDghSapQoYL++9//OrM0AAAAAChWLncNW6NGjfTzzz+rVq1aatq0qWbOnKnU1FR9/PHH+tvf/ubs8gAAAACg2LjcGbYXX3xRAQEBtn9XqlRJb775plJTUzV+/HgnVwcAAAAAxcflzrCFhoba/u3n56eFCxc6sRoAAAAAKDkud4atb9++SktLc3YZAAAAAFDiXC6wff/99/m+LBsAAAAAbkUuF9gkyWQyObsEAAAAAChxLncNmyQNGTJEnp6eBc5btmxZKVcDAAAAACXDJQNb48aN5e3t7ewyAAAAAKBEuVxgM5lMioqKkp+fn7NLAQAAAIAS5XLXsFmtVmeXAAAAAAClwuUC29ChQ1W+fHlnlwEAAAAAJc7lhkQOHTpUknT69GkdOnRIklSnTh1VqVLFmWUBAAAAQLFzucCWkZGhcePG6ZNPPlFOTo4kyWKx6JFHHtHYsWNVrlw5J1cIAAAAAMXD5YZExsTEaMeOHXr33Xe1c+dO7dy5U++++6527NihyZMnO7s8AAAAACg2LhfYPvvsM02cOFFt2rRRhQoVVKFCBbVp00bjx4/XZ5995uzyAAAAAKDYuFxgu3jxovz9/fNN9/Pz08WLF51QEQAAAACUDJcLbI0bN9bMmTN16dIl27SLFy/qnXfeUePGjZ1XGAAAAAAUM5e76ciYMWMUFRWl++67Tw0aNJAk/frrrypTpowWLlzo5OoAAAAAoPi4XGALCgrS559/rvXr19tu69+5c2d16dJFZcuWdXJ1AAAAAFB8XC6w7dixQ+Hh4erZs6ezSwEAAACAEuVy17D17dtXZ8+edXYZAAAAAFDiXC6wWa1WZ5cAAAAAAKXC5YZEStLu3bvl4+NT4LymTZuWcjUAAAAAUDJcMrANHTq0wOkmk0n79+8v5WoAAAAAoGS4ZGD79ttv5efn5+wyAAAAAKBEudw1bCaTydklAAAAAECpcLnAxk1HAAAAALgLlxsS+euvvzq7BAAAAAAoFS53hm3evHlatWpVvumrVq3S/PnznVARAAAAAJQMlwtsH374oerUqZNv+t/+9jetXLnSCRUBAAAAQMlwucCWnJysgICAfNOrVKmi5ORkJ1QEAAAAACXD5QLb7bffrl27duWb/sMPP6hq1apOqAgAAAAASobL3XSkR48emjRpkrKzs3XvvfdKkhITEzV16lQ9/fTTTq4OAAAAAIqPywW2qKgonTlzRtHR0crKypIklSlTRlFRURo0aJCTqwMAAACA4uNygc1kMumll17Ss88+q4MHD6ps2bKqVauWvLy8nF0aAAAAABQrlwtseby9vRUWFubsMgAAAACgxLhkYPv555+1ceNGJSUl2YZF5nnnnXecVBUAAAAAFC+Xu0vkhg0b1KtXLx06dEibN29Wdna2fv/9d23fvl0VK1Z0dnkAAAAAUGxcLrDNnTtXr7zyiubOnStPT0+9+uqr2rRpkzp06KDbb7/doW3FxcWpbdu2Cg0NVY8ePbRnz55Cl09LS1N0dLRatWqlkJAQPfjgg/r6669v5ukAAAAAwDW5XGA7evSo2rRpI0ny8vJSenq6TCaT+vXrp/j4+CJvJyEhQTExMRoyZIjWrl2rBg0aKDIyUikpKQUun5mZqf79++v48eOKjY3Vpk2bNH78eAUGBhbL8wIAAACAq7lcYKtUqZIuXLggSapatap+//13SZfPfmVkZBR5O4sXL1bPnj3VvXt31atXT9HR0SpbtqxWr15d4PKrV6/W2bNnNXv2bN19992qXr267rnnHjVo0ODmnxQAAAAAFMDlbjrStGlTbdu2TUFBQXrooYc0ceJEbd++Xdu2bVPz5s2LtI3MzEzt27fP7nvbzGazWrRood27dxe4ztatW9W4cWONGzdOX3zxhapUqaLOnTtrwIABslgsDj0Hk8mhxUtEUWswQq0whrxeoCdQVPQMHEXPwFH0DBxlpJ4pag0uF9jGjh2rS5cuSZIGDx4sT09P7dq1S+3bt9fgwYOLtI3U1FTl5OTIz8/Pbrqfn58OHTpU4DpHjx7V9u3b1aVLF82fP19HjhxRdHS0srOzNXToUIeeg5+fa9wcxdfX29klwIBcpX9hHPQMHEXPwFH0DBzlSj3jMoHt/PnzkiQPDw95eHjYHj/xxBN64oknSnz/VqtVfn5+Gj9+vCwWi0JCQnTy5EktXLjQ4cCWknJOVmsJFVpEHh5mVa5ceCBLTb2gnJzcUqoIRmcyXf7lZoT+hWugZ+AoegaOomfgKCP1TF4t1+Myga1JkyYyFeG84f79+6+7jK+vrywWS74bjKSkpMjf37/AdQICAuTh4WE3/LFOnTpKTk5WZmamvLy8rrvfPFarnN4gRd2/s+uE8Rihf+Fa6Bk4ip6Bo+gZOMqVesZlAtuyZcvsHlutVg0cOFATJkxw+E6NXl5eCg4OVmJiotq1aydJys3NVWJionr37l3gOhEREfr000+Vm5srs/nyvVoOHz6sgIAAh8IaAAAAABSVywS2e+65J980s9msxo0bq0aNGg5vr3///ho9erRCQkIUFhampUuXKiMjQ926dZMkjRo1SoGBgRoxYoQkqVevXnr//fc1ceJE9e7dW3/++afmzZunPn363NwTAwAAAIBrcJnAVtw6duyo06dPa+bMmUpOTlbDhg21YMEC25DIpKQk25k0Sbr99tu1cOFCxcTE6OGHH1ZgYKD69u2rAQMGOOspAAAAALjFuWxgS0pK0sWLF1W5cuUb3kbv3r2vOQRy+fLl+aaFh4c79OXcAAAAAHAzXCawXXkNW2pqqjZs2KB7771XFSu6zi05AQAAAMARLhPYlixZIkkymUzy9fXV/fffX+TvXQMAAAAAV+QygW3r1q3OLgEAAAAASpX5+osAAAAAAJyBwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQbh3Y4uLi1LZtW4WGhqpHjx7as2dPkdbbsGGDgoKC9Oyzz5ZwhQAAAADcmdsGtoSEBMXExGjIkCFau3atGjRooMjISKWkpBS63rFjx/TWW2+pSZMmpVQpAAAAAHfltoFt8eLF6tmzp7p376569eopOjpaZcuW1erVq6+5Tk5OjkaOHKnnnntONWrUKMVqAQAAALgjD2cX4AyZmZnat2+fBg0aZJtmNpvVokUL7d69+5rrzZ49W35+furRo4d++OGHG96/yXTDqxabotZghFphDHm9QE+gqOgZOIqegaPoGTjKSD1T1BrcMrClpqYqJydHfn5+dtP9/Px06NChAtfZuXOnVq1apXXr1t30/v38Kt70NkqDr6+3s0uAAblK/8I46Bk4ip6Bo+gZOMqVesYtA5ujzp8/r1GjRmn8+PGqUqXKTW8vJeWcrNZiKOwmeHiYVbly4YEsNfWCcnJyS6kiGJ3JdPmXmxH6F66BnoGj6Bk4ip6Bo4zUM3m1XI9bBjZfX19ZLJZ8NxhJSUmRv79/vuWPHj2q48ePa/DgwbZpubmXg0yjRo20adMm3XnnnUXev9UqpzdIUffv7DphPEboX7gWegaOomfgKHoGjnKlnnHLwObl5aXg4GAlJiaqXbt2ki4HsMTERPXu3Tvf8nXq1NH69evtps2YMUMXLlzQq6++qttuu61U6gYAAADgXtwysElS//79NXr0aIWEhCgsLExLly5VRkaGunXrJkkaNWqUAgMDNWLECJUpU0b169e3W79SpUqSlG86AAAAABQXtw1sHTt21OnTpzVz5kwlJyerYcOGWrBggW1IZFJSksxmt/3WAwAAAAAG4LaBTZJ69+5d4BBISVq+fHmh606ePLkkSgIAAAAAG04hAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKDcOrDFxcWpbdu2Cg0NVY8ePbRnz55rLhsfH68nnnhCTZs2VdOmTdWvX79ClwcAAACAm+W2gS0hIUExMTEaMmSI1q5dqwYNGigyMlIpKSkFLv/dd9+pU6dOWrZsmVauXKnbb79dTz/9tE6ePFnKlQMAAABwF24b2BYvXqyePXuqe/fuqlevnqKjo1W2bFmtXr26wOWnT5+uJ598Ug0bNlTdunU1YcIE5ebmKjExsZQrBwAAAOAuPJxdgDNkZmZq3759GjRokG2a2WxWixYttHv37iJtIyMjQ9nZ2fLx8XF4/yaTw6sUu6LWYIRaYQx5vUBPoKjoGTiKnoGj6Bk4ykg9U9Qa3DKwpaamKicnR35+fnbT/fz8dOjQoSJtY9q0aapatapatGjh8P79/Co6vI4z+Pp6O7sEGJCr9C+Mg56Bo+gZOIqegaNcqWfcMrDdrPnz5yshIUHLli1TmTJlHF4/JeWcrNYSKMwBHh5mVa5ceCBLTb2gnJzcUqoIRmcyXf7lZoT+hWugZ+AoegaOomfgKCP1TF4t1+OWgc3X11cWiyXfDUZSUlLk7+9f6LoLFy7U/PnztXjxYjVo0OCG9m+1yukNUtT9O7tOGI8R+heuhZ6Bo+gZOIqegaNcqWfc8qYjXl5eCg4OtrthSN4NRMLDw6+53nvvvad3331XCxYsUGhoaGmUCgAAAMCNueUZNknq37+/Ro8erZCQEIWFhWnp0qXKyMhQt27dJEmjRo1SYGCgRowYIenyMMiZM2dq+vTpuuOOO5ScnCxJKl++vLy9udYLAAAAQPFz28DWsWNHnT59WjNnzlRycrIaNmyoBQsW2IZEJiUlyWz+3wnIlStXKisrS88//7zddoYOHarnnnuuVGsHAAAA4B7cNrBJUu/evdW7d+8C5y1fvtzu8datW0ujJAAAAACwcctr2AAAAADAFbj1GTYAAJzBbDbJbC74G1Nzc63KzXWRW5cBAEocgQ0AgFJkNpvkU7m8PCwFD3LJzsnV2TPphDYAgCQCGwAApcpsNsnDYtYLK3frwF/n7ebVq1pBsY+Hy2w2EdgAAJIIbAAAOMWBv85r34k0Z5cBADA4bjoCAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAbl4ewCAAAFM5tNMptNBc7LzbUqN9dayhUBAIDSRmADAAMym03yqVxeHpaCB0Jk5+Tq7Jl0QhsAALc4AhsAGJDZbJKHxawXVu7Wgb/O282rV7WCYh8Pl9lsIrABAFCAwkapuBoCGwAY2IG/zmvfiTRnlwEAgMu43iiVnFyrzGaTcnJc40NPAhsAAACAW0ZRRqmYTCZJBDYAAAAAcIpbZZQKt/UHAAAAAIMisAEAAACAQRHYAAAAAMCgCGwAAAAAYFAENgAAAAAwKAIbAAAAABgUgQ0AAAAADIrABgAAAAAGxRdnA3BJZrNJZrOpwHm5uVbl5lpLuSIAAIDiR2AD4HLMZpN8KpeXh6XgQQLZObk6eyad0AYAAFwegQ2AyzGbTfKwmPXCyt068Nd5u3n1qlZQ7OPhMptNBDYAAODyCGwAXNaBv85r34k0Z5cBAABQYrjpCAAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEFxl0gAAAAHmc0mmc2mAufl5lr5WhEAxYbABgAA4ACz2SSfyuXlYSl4oFJ2Tq7OnkkntAEoFgQ2AA6xXOMNCp8oA3AXZrNJHhazXli5Wwf+Om83r17VCop9PFxms4nfiQCKBYENQJGYzSbl5Frl6+td4Hw+UQbgbg78dV77TqQ5uwwAtzgCG4AiMZlMsphNfKIMAABQighsABzCJ8oAAAClh9v6AwAAAIBBcYYNAFCiuFENAAA3jsAGACgR3KgGAICbR2ADXBhf3Aoj40Y1AADcPAIb4KL44la4Cm5UA9x6+MAQKD0ENsBF8cWtAG5VXPdobHxgCJQuAhtgADfzSSVnLwDcKrju0TXwgSFQutw6sMXFxWnhwoVKTk5WgwYNNHbsWIWFhV1z+Y0bNyo2NlbHjx9XrVq1NHLkSLVp06YUK8atiE8qjYMhPoBzcd2ja+EDQ6B0uG1gS0hIUExMjKKjo3XXXXdp6dKlioyM1KZNm+Tn55dv+V27dmnEiBEaPny47r//fq1fv15DhgzRmjVrVL9+fSc8AxjRjbzh55NKYyA4Gwvh2TWU1HEiCMBRDKPFrcxtA9vixYvVs2dPde/eXZIUHR2tr776SqtXr9bAgQPzLb9s2TK1bt1aUVFRkqRhw4Zp27Ztev/99zVu3LhSrd3ZCvsD7c5MJpMqVCxb6Bv+8+cuymq1/8OR90emsDcoBf0hutYfJ0eXKarS3t/N7sfRWiwW83WDs6enRTk5uQ5t90YZ6fW+UTf6HG70/5LR3MwbRaMfW6lkjpOr9L2r1HkzbjboFPfzL+xDz+sNo3WF3xc3qrDjxAdftw63DGyZmZnat2+fBg0aZJtmNpvVokUL7d69u8B1fvzxR/Xr189uWqtWrbRlyxaH9282S87+vWH6//9/g6tVUjkvi928Ov6Xf+ndyBupnFyrLNf45WCkeSW53blfHdSJsxl20/5WtYKeaFZTlSuXv+Z6BR2Lu++srJxcqypVKles693M8yvN/ZV2LZJUxsOcb5u3VSpzU9s00nN0xv+1G+3tG/m/ZKTnnp2TqwvnL+Z7U5T3u7Uk+tcZx7e4j5PkGn3vKnUWd/9KJdfDN1KPh4dZFrPJ5X9fFPfrUpQPVK51fK3W/71PdPa8G123sB7Ne59rMl1+T+5MhT1vu+Wst+pHDoU4efKk7rvvPq1cuVLh4eG26VOmTNGOHTv00Ucf5VsnJCREkydPVufOnW3T4uLiNHv2bG3btq1U6gYAAADgXlz7fD0AAAAA3MLcMrD5+vrKYrEoJSXFbnpKSor8/f0LXMff31+nTp0q8vIAAAAAcLPcMrB5eXkpODhYiYmJtmm5ublKTEy0GyJ5pcaNG2v79u1207Zt26bGjRuXZKkAAAAA3JhbBjZJ6t+/v+Lj47V27VodPHhQb775pjIyMtStWzdJ0qhRozR9+nTb8n379tV//vMfLVq0SAcPHtSsWbO0d+9e9e7d21lPAQAAAMAtzi3vEilJHTt21OnTpzVz5kwlJyerYcOGWrBggW2IY1JSksxX3DomIiJC06ZN04wZM/T222+rVq1amj17Nt/BBgAAAKDEuOVdIgEAAADAFbjtkEgAAAAAMDoCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgc1NxcXFq27atQkND1aNHD+3Zs8fZJcEA5s2bp+7duys8PFzNmzfXs88+q0OHDtktc+nSJUVHR6tZs2YKDw/Xc889l+9L5eG+5s+fr6CgIE2cONE2jZ7B1U6ePKmRI0eqWbNmCgsLU5cuXfTzzz/b5lutVsXGxqpVq1YKCwtTv379dPjwYecVDKfKycnRjBkz1LZtW4WFhaldu3aaPXu2rrxvHj3j3nbs2KFnnnlGrVq1UlBQkLZs2WI3vyj9cebMGY0YMUIRERFq0qSJxowZowsXLpTis7g2ApsbSkhIUExMjIYMGaK1a9eqQYMGioyMVEpKirNLg5N9//33evLJJxUfH6/FixcrOztbkZGRSk9Pty0zadIkffnll5oxY4aWL1+uv/76S0OHDnVi1TCKPXv2aOXKlQoKCrKbTs/gSmfPnlWvXr3k6emp9957Txs2bNDo0aPl4+NjW+a9997T8uXL9eabbyo+Pl7lypVTZGSkLl265MTK4SzvvfeeVqxYoddff10JCQkaOXKkFixYoOXLl9stQ8+4r/T0dAUFBemNN94ocH5R+mPkyJE6cOCAFi9erLlz52rnzp16/fXXS+spFM4Kt/Poo49ao6OjbY9zcnKsrVq1ss6bN8+JVcGIUlJSrPXr17d+//33VqvVak1LS7MGBwdbN27caFvmwIED1vr161t3797tpCphBOfPn7e2b9/e+u2331p79+5tnTBhgtVqpWeQ39SpU629evW65vzc3Fxry5YtrQsWLLBNS0tLs4aEhFg//fTT0igRBjNw4EDrK6+8Yjdt6NCh1hEjRlitVnoG9urXr2/dvHmz7XFR+iPv79KePXtsy3z99dfWoKAg63//+9/SK/4aOMPmZjIzM7Vv3z61aNHCNs1sNqtFixbavXu3EyuDEZ07d06SbJ987927V1lZWXb9U7duXVWrVk0//vijM0qEQYwbN05t2rSx6w2JnkF+W7duVUhIiJ5//nk1b95c//znPxUfH2+bf+zYMSUnJ9v1TMWKFXXXXXfxd8pNhYeHa/v27frjjz8kSb/++qt++OEH3XfffZLoGRSuKP2xe/duVapUSaGhobZlWrRoIbPZbIjLhjycXQBKV2pqqnJycuTn52c33c/PL9+1SnBvubm5mjRpkiIiIlS/fn1J0qlTp+Tp6alKlSrZLevn56fk5GRnlAkD2LBhg3755RetWrUq3zx6Blc7evSoVqxYof79++uZZ57Rzz//rAkTJsjT01Ndu3a19UVBf6e49tE9DRw4UOfPn1eHDh1ksViUk5OjF198UQ8//LAk0TMoVFH649SpU6pSpYrdfA8PD/n4+BjibxWBDUCBoqOj9fvvv+uDDz5wdikwsKSkJE2cOFGLFi1SmTJlnF0OXIDValVISIiGDx8uSWrUqJF+//13rVy5Ul27dnVydTCijRs3av369Zo+fbrq1aun/fv3KyYmRlWrVqVn4BYYEulmfH19ZbFY8t1gJCUlRf7+/k6qCkYzbtw4ffXVV1q6dKluu+0223R/f39lZWUpLS3NbvmUlBQFBASUdpkwgH379iklJUXdunVTo0aN1KhRI33//fdavny5GjVqRM8gn4CAANWtW9duWp06dXTixAnbfEn8nYLNlClTNHDgQHXq1ElBQUH65z//qaeeekrz5s2TRM+gcEXpD39/f50+fdpufnZ2ts6ePWuIv1UENjfj5eWl4OBgJSYm2qbl5uYqMTFR4eHhTqwMRmC1WjVu3Dht3rxZS5cuVY0aNezmh4SEyNPT065/Dh06pBMnTqhx48alXC2M4N5779X69eu1bt06209ISIi6dOli+zc9gytFRETYrkXKc/jwYd1xxx2SpOrVqysgIMCuZ86fP6+ffvqJv1Nu6uLFizKZTHbTLBaL7bb+9AwKU5T+CA8PV1pamvbu3WtbZvv27crNzVVYWFip13w1hkS6of79+2v06NEKCQlRWFiYli5dqoyMDHXr1s3ZpcHJoqOj9emnn+rdd9+Vt7e3bdx2xYoVVbZsWVWsWFHdu3fX5MmT5ePjowoVKmjChAkKDw/nzbebqlChgu0axzzly5dX5cqVbdPpGVzpqaeeUq9evTR37lx16NBBe/bsUXx8vMaNGydJMplM6tu3r+bMmaOaNWuqevXqio2NVdWqVdWuXTsnVw9nuP/++zV37lxVq1bNNiRy8eLF6t69uyR6BtKFCxd05MgR2+Njx45p//798vHxUbVq1a7bH3Xr1lXr1q01duxYRUdHKysrS+PHj1enTp0UGBjorKdlY7Jar/jWQbiN999/XwsXLlRycrIaNmyo1157TXfddZezy4KTXf39WXliYmJsgf7SpUuaPHmyNmzYoMzMTLVq1UpvvPGGIYYMwBj69OmjBg0a6NVXX5VEzyC/L7/8Um+//bYOHz6s6tWrq3///urZs6dtvtVq1cyZMxUfH6+0tDTdfffdeuONN1S7dm0nVg1nOX/+vGJjY7VlyxalpKSoatWq6tSpk4YMGSIvLy9J9Iy7++6779S3b99807t27arJkycXqT/OnDmj8ePHa+vWrTKbzWrfvr1ee+01eXt7l+ZTKRCBDQAAAAAMimvYAAAAAMCgCGwAAAAAYFAENgAAAAAwKAIbAAAAABgUgQ0AAAAADIrABgAAAAAGRWADAAAAAIMisAEAAACAQXk4uwAAuBW9/PLLWrt27TXn79ixQ5UqVSrFigAAgCsisAFACWndurViYmLspu3evVvPPfeckyoCAACuhiGRAFBCvLy8FBAQYPfj4+Njt8yaNWvUpEkTbdmyRe3bt1doaKgiIyOVlJRkt9wHH3ygdu3aKSQkRA8++KDWrVtnNz8oKEhbtmyxm9anTx9NnDjRbtpHH32kDh06KDQ0VA899JDi4uJs844dO6agoCDt37/fbp22bdtqyZIltn8HBQUV+LNmzZoCX4eXX35ZQUFBWrx4sd30IUOG5FsvKSlJL7zwgpo0aaJ77rlHgwcP1rFjxyRJs2bNuua++/TpY9vXs88+a9teamqqmjZtqiZNmtjtO++5Xv2TlpZm29cjjzxiWz4zM1P/+Mc/7JYpzmOX9xMREaH+/fvryJEjtvnr1q1Tt27dFB4erpYtW2rEiBFKSUmxzf/uu+/s6rpyu3k9UZRje/U6V3vkkUc0a9Ys2z5DQkK0c+dO2/z33ntPzZs316lTpwpcP+/1ulJBdV1dw0cffaSgoCBbL/fp0+eafZBX39mzZzVq1Cg1bdpUd911l6KionT48GHbNhMSEtSuXTuFhoaqWbNmev7553X69Oki11DQayfl77/MzExNmDBBzZs3V2hoqHr16qU9e/bY5ucdu6CgIDVo0EDNmzfXmDFjdOnSpQJfQwDuicAGAE528eJFzZkzR2+99ZZWrFihtLQ0vfjii7b5mzdv1qRJk9S/f3+tX79ejz/+uMaMGaPt27c7tJ9PPvlEsbGxevHFF5WQkKDhw4dr5syZhQ7dvNqqVav0zTff6JtvvtFtt92mMWPG2B537NjxmusFBgYqPj7e9vjkyZPatWuXypUrZ5uWlZWlyMhIeXt7Ky4uTitWrFD58uUVFRWlzMxMPf3007Z9Pf300woPD7c9znujfrXZs2crOzs733Sr1SpJWrJkSaHr54mLiyswiBTXsYuJidE333yjuLg4paSk6O2337bNy87O1gsvvKBPPvlEs2fP1vHjx/Xyyy8XWm9Ja9asmfr27atRo0bp3Llz+uWXXxQbG6sJEybI39+/2PaTnp6u2NhYlS9f3jZt1qxZtuMeHh6ery+ky8Fp7969mjNnjj788ENZrVYNHDhQWVlZkqQ6depo8uTJ2rRpkxYsWKBjx45p6tSpRa6hqKZMmaLPPvtMkydP1tq1a1WzZk1FRUXpzJkzdstt2rRJ//73vzV16lRt3LhRq1evdnhfAG5dDIkEACfLysrS66+/rrvuukuSNHnyZHXs2FF79uxRWFiYFi5cqK5du+rJJ5+UJNWuXVs//vijFi1apHvvvbfI+5k1a5ZefvlltW/fXpJUo0YNHThwQB9++KG6du1apG1UqVLF9m+LxaKKFSsqICDguuuFhITo2LFj2rlzp5o0aaLVq1erU6dOdmebEhISlJubq4kTJ8pkMkm6HGSaNm2q77//Xq1atZK3t7ckqXz58vL09Cx033/88YdWr16tfv36afny5Xbz8kKcv79/gWc+r3TmzBnNmTNHAwYMUGxsrN284jp2lSpVUkBAgMqWLStvb29VrFjRNu/RRx+1/btGjRp69dVX9eijj+rChQu218MZhg0bpm3btmns2LH6/fff1bVrVz3wwAPFuo8FCxaoXr16ysnJsU2rXLmy7d+enp4qX768XR8cPnxYW7du1YoVKxQRESFJmjZtmv7+979ry5Yt6tChgxo0aGBb3sfHR76+vsrNzS1yDUWRnp6ulStXKiYmRm3atJEkjR8/Xt9++61WrVqlqKgo27J+fn6qVKmSMjIy5OnpaXf8AYDABgBO5uHhodDQUNvjunXrqlKlSjp48KDCwsJ06NAhPfbYY3brREREaNmyZUXeR3p6uo4cOaJXX31VY8eOtU3Pzs7O9+bw8ccfl9n8vwEYGRkZjj6lAvXs2VMffvihIiIitGrVKs2ZM8cusP366686cuSI7U12nkuXLtkNESyqqVOn6rHHHlONGjXyzTt//rwk2Z3hu5bZs2erWbNmuvvuu/PNK65jN3z4cFksFmVkZKh+/foaPny4bd7evXv1zjvv6Ndff9XZs2dtZweTkpJUr14923J5oaAwRTm2ebV4e3urUaNGGjVqlN1+8nh5eWnatGl6+OGHVa1aNb3yyivX3b8jTp48qcWLF2vFihX5hvYW5uDBg/Lw8LCFaEny9fVV7dq1dfDgQdu0nTt3asCAAcrIyNDdd9+t0aNHO1zDtGnT7EJ8Zmam7TgcOXJEWVlZdv3s6empsLAwuzqky8fOarUqIyNDXbp0UefOnYv8fAHc+ghsAOAG0tPTJV3+hP/KN7KS7N7AS9K//vUv1a1b1/Y47/qwm/Xwww9r1qxZ2rBhgwICAhQUFJSvxuDgYE2bNi3fulee2SuK77//Xj/88INiYmL0xRdf5Jv/119/yWw2X/fs4OHDh7Vq1SqtW7dO//3vfx2qwRGvvPKKWrRoobS0NM2YMUOvvPKK5s6dq/T0dEVGRqpVq1aaNm2afH19lZSUpMjISNvwvjxxcXF2Z9zyzqReqSjH9spa/vWvf2nYsGH69NNPC6x79+7dki5fM3b27NkbGjZ4LTNmzNBDDz1kdzasOIWEhGjt2rU6ceKEJk6cqI8++kiDBg1yqIbIyEh169bN9njatGkOn4mT/nfsjh07prFjx2rx4sW24Z0AwDVsAOBk2dnZ2rt3r+3xoUOHlJaWZntjXadOHe3atctunV27dhV41uNa/P39VbVqVR09elQ1a9a0+7n6DNTtt99uN9/Do3g+26tUqZLatm2rN954Qz169Mg3Pzg4WH/++af8/Pzy1ejIEDGr1arJkydr8ODB1xzq+PPPP6tOnToqU6ZModuaNm2aHn30UdWsWbPA+cV17AICAlSzZk2Fhoaqd+/e+uqrr5SVlaVDhw7pzJkzGjlypJo0aaK6deva3XDkStWrV7d7zQpSlGN7ZS19+/bVgQMH8oVD6fIZpEmTJmn8+PEKCwvT6NGjrzms0FH79+/Xpk2bNGzYMIfXrVu3rrKzs/XTTz/ZpqWmpuqPP/6we93Lli2rWrVqqUWLFurZs6fWr1/vcA2+vr52r+eVgfnOO++Up6en3fHPysrSzz//nO/45x27li1b6sEHH9TmzZsdft4Abl2cYQMAJ/P09NT48eP12muvyWKxaPz48WrcuLHCwsIkSVFRURo2bJgaNmyoFi1a6Msvv9TmzZvz3XUxKyvL7u5yubm5ys3NVXZ2tjw8PPT8889rwoQJqlixolq3bq3MzEzt3btXaWlp6t+/f6k81wEDBqh27doF3qCkS5cuWrhwoQYPHqwXXnhBgYGBOnHihDZv3qyoqCjddtttRdrH9u3bFRAQYLtu7EqZmZlKSEjQkiVLrvv1CkeOHFFSUpI+//zzay5TXMcuLS1NycnJOnfunFatWqUaNWrI09NT1apVk6enp5YvX65evXrpt99+07vvvluk1+FG5fVRWlqaPvnkE9WqVUuenp52y+Tk5Oill15S69at1b17d7Vu3VpdunTRokWL7K7NKsiVPZqZmWnbp9VqtV27uGjRIvXv31+BgYEO11+rVi098MADGjt2rKKjo1WhQgVNmzZNgYGBtmvsNmzYoBo1asjf319//vmnVq5cqZCQELvt3EwN0uXrLHv16qUpU6bIx8dH1apV04IFC3Tx4kW76xIlKSUlRZcuXdKJEye0devWfHfTBODeCGwA4GRly5bVgAEDNGLECJ08eVJNmjSxu16mXbt2GjNmjBYtWqRJkybpjjvu0KRJk9SsWTO77RR0JmDnzp2qXLmynnvuOfXo0UNly5bVwoULNWXKFJUvX17169fXU089VdJP0aZOnToaOHBggfPKlSun999/X9OmTdPQoUN14cIFBQYGqnnz5qpQoUKR95Genq4RI0bkCxmS9Ntvv+mdd97Rs88+e92Qmp6ertGjR9vd5OJqxXXs8q7/8vb2VnBwsGbOnCnp8lDQyZMn6+2339by5csVHBys0aNHa/DgwUV9ORyW10d5tcyYMSPfMnPmzNHx48c1d+5cSVLVqlU1fvx4DR8+XK1atbrmEMJz587ZwuyVevTooS+++ELVq1e37ft6wa8wMTExmjhxop555hllZWWpSZMmmj9/vq0nDh48qKlTp+rUqVPy9fXVfffdl+8atputQZJGjhwpq9WqUaNG6cKFCwoJCdGCBQvynfl96KGHJF0+Y9eyZUu99NJLN7VfALcWkzXv6mUAQKlbs2aNJk2aZPd9VsVpyZIlOnfuHF/WXQJK+ti5k0ceeUSzZ8+2BTYAwP9wDRsA3MI8PDxksVicXQZQKC8vL9twSACAPYZEAsAtrHfv3s4uAbiujz76yNklAIBhMSQSAAAAAAyKIZEAAAAAYFAENgAAAAAwKAIbAAAAABgUgQ0AAAAADIrABgAAAAAGRWADAAAAAIMisAEAAACAQRHYAAAAAMCgCGwAAAAAYFAENgAAAAAwKAIbAAAAABgUgQ0AAAAADIrABgAAAAAGRWADAAAAAIMisAEAAACAQRHYAABAiWrbtq1efvll2+M1a9YoKChIP//883XX7dOnj/r06VOS5QGAoRHYAOAWlvfGOO8nNDRUDz74oMaNG6dTp045uzzAqQ4cOKBZs2bp2LFjzi4FAK7Jw9kFAABK3vPPP6/q1asrMzNTP/zwg1asWKGvv/5an376qcqVK+fs8nCL27Rpk0wmk7PLyOfAgQN65513dM8996h69erOLgcACkRgAwA3cN999yk0NFSS1KNHD1WuXFmLFy/WF198oc6dOzu5OtzqvLy8nF0CALgshkQCgBu69957Jck2FOzMmTN666231KVLF4WHhysiIkJRUVH69ddf86176dIlzZo1Sw8++KBCQ0PVqlUrDR06VEeOHLFt88phmFf/XHk90nfffaegoCAlJCTo7bffVsuWLdW4cWM988wzSkpKyrfvn376SZGRkbr77rt11113qXfv3vrhhx8KfI59+vQpcP+zZs3Kt+zHH3+sbt26KSwsTPfcc49efPHFAvdf2HO7Um5urpYsWaJOnTopNDRULVq00Ouvv66zZ8/aLde2bVsNGjQo337GjRuXb5sF1b5gwYJ8r6kkZWZmaubMmfrHP/6hkJAQtWnTRlOmTFFmZmaBr9XVivI6z5o1S0FBQTp48KBeeOEFRUREqFmzZpowYYIuXbqU73leeQ1bnosXL+r1119Xs2bNFBERoVGjRuV7ja6WmZmp2NhYdevWTXfffbcaN26sJ554Qtu3b8+37IYNG9StWzdbT3fp0kVLly6VdHm48AsvvCBJ6tu3r+04fvfdd0V6jQCgtHCGDQDcUF64qly5siTp6NGj2rJlix566CFVr15dp06d0ocffqjevXtrw4YNCgwMlCTl5ORo0KBBSkxMVKdOndS3b19duHBB3377rX777Tfdeeedtn107txZ9913n91+33777QLrmTNnjkwmkwYMGKCUlBQtXbpU/fr108cff6yyZctKkhITEzVgwACFhIRo6NChMplMWrNmjZ566il98MEHCgsLy7fd2267TcOHD5ckpaen68033yxw37GxserQoYMeffRRnT59Wu+//76efPJJrVu3TpUqVcq3zmOPPaa7775bkrR582Zt3rzZbv7rr7+utWvXqlu3burTp4+OHTumuLg4/fLLL1qxYoU8PT0LfB0ckZaWpvnz5+ebnpubq8GDB+uHH35Qz549VbduXf32229aunSpDh8+rHfffbfQ7Tr6Og8bNkx33HGHRowYoR9//FHLly9XWlqapkyZct3nMG7cOFWqVElDhw7VH3/8oRUrVujEiRNavnz5NYdQnj9/Xh999JE6d+6sHj166MKFC1q1apWioqL00UcfqWHDhpKkb7/9VsOHD1fz5s01cuRISdKhQ4e0a9cuPfXUU2ratKn69Omj5cuX65lnnlGdOnUkSXXr1r1u3QBQmghsAOAGzp8/r9OnTyszM1O7du3S7NmzVbZsWd1///2SLp+9+eyzz2Q2/2/gxSOPPKIOHTpo1apVGjJkiCRp3bp1SkxM1CuvvKJ+/frZlh04cKCsVqvdPhs1aqRHHnnEbtp7771XYH1nz55VQkKCKlSoYFt32LBhio+PV9++fWW1WvXmm2+qWbNmWrBgge3N/OOPP65OnTppxowZWrRokd02s7KyVKlSJVsNp0+fzhfYjh8/rlmzZmnYsGF65plnbNPbt2+vrl276oMPPrCbnpOTI0mKiIiwbffIkSN2gW3nzp366KOPNG3aNHXp0sU2vVmzZoqKitKmTZvspt+oefPmycPDQ8HBwXbT169fr23btmn58uVq0qSJbfrf/vY3vfHGG9q1a5ciIiIK3OaNvM7Vq1fXnDlzJElPPvmkKlSooA8++EBPP/20GjRoUOhz8PT01JIlS2wBtlq1apo6daq2bt2qBx54oMB1fHx8tHXrVrthlj179lSHDh20fPlyTZo0SZL01VdfqUKFClq4cKEsFku+7dSoUUNNmjTR8uXL1aJFCzVr1qzQWgHAWRgSCQBuoF+/fmrevLnatGmjF198Ud7e3nrnnXdsZ868vLxsYS0nJ0epqakqX768ateurV9++cW2nc8//1y+vr7q3bt3vn3czE0l/vnPf9rCmiQ99NBDCggI0Ndffy1J2r9/vw4fPqwuXbooNTVVp0+f1unTp5Wenq7mzZtrx44dys3NtdtmZmbmda+d2rx5s3Jzc9WhQwfbNk+fPi1/f3/VrFkz3/C4rKwsSYVfk7Vp0yZVrFhRLVu2tNtmcHCwypcvn2+b2dnZdsudPn0635DCq508eVLvv/++nn32WXl7e+fbf926dVWnTh27beYNgy1syN+NvM5PPvmk3eO83vj3v/9d6HOQLp+pvPJsY69eveTh4WE77gWxWCy21z83N1dnzpxRdna2QkJC7Hq1UqVKysjI0LfffnvdOgDAyDjDBgBu4PXXX1ft2rVlsVjk7++v2rVr251Ny83N1bJly/TBBx/o2LFjtjNJ0v+GTUqXzybVrl1bHh7F++ejZs2ado9NJpNq1qyp48ePS5IOHz4sSRo9evQ1t3Hu3Dn5+PjYHqempubb7tUOHz4sq9Wq9u3bFzj/6ueZlpYmSSpfvvw1t/nnn3/q3Llzat68eYHzU1JS7B5/880311z2WmbOnKmqVavqscce02effZZv/wcPHizy/q90I6/z1a/xnXfeKbPZXKRb5V+9rre3twICAmzH/VrWrl2rRYsW6Y8//rCFaEl2d3p84okntHHjRg0YMECBgYFq2bKlOnTokG+YLgAYHYHNCXbs2KGFCxdq7969Sk5O1uzZs9WuXTuHtmG1WrVo0SLFx8fr+PHj8vX11RNPPKHBgweXUNUAXFlYWJjtLpEFmTt3rmJjY9W9e3e98MIL8vHxkdls1qRJk/INdXSGvBpGjRplu0bpaleGqMzMTCUnJ6tFixaFbjc3N1cmk0nvvfdegcPmrg5med9d5+/vX+g2/fz8NG3atALnV6lSxe7xXXfdpWHDhtlNe//99/XFF18UuP7Bgwe1du1aTZ06tcBr4XJzc1W/fn298sorBa5/2223XbN2R1/ngpT07fs//vhjvfzyy2rXrp0iIyPl5+cni8WiefPm6ejRo7bl/Pz8tG7dOn3zzTf697//rX//+99as2aN/vnPf+qtt94q0RoBoDgR2JwgPT1dQUFB6t69u4YOHXpD25g4caK++eYbjRo1SvXr19fZs2eve2ctALiWzz77TM2aNbNd/5MnLS1Nvr6+tsd33nmnfvrpJ2VlZRXLjTPy/Pnnn3aPrVar/vzzT9udEmvUqCFJqlChwnVDmCT9+uuvysrKUkhISKHL3XnnnbJarapevbpq16593e0eOHBAJpOp0GXvvPNOJSYmKiIiwnbDlML4+vrme05btmy55vLTp09XgwYN1LFjx2vu/9dff1Xz5s0dDk+Ovs7S5WOXt17e49zc3CJ9r9mff/5pG6opSRcuXFBycnKhZ8E+++wz1ahRQ++8847d85s5c2a+Zb28vNS2bVu1bdtWubm5evPNN/Xhhx/q2WefVc2aNQ353XAAcDWuYXOCvGtI/vGPfxQ4PzMzU2+99ZZat26txo0bq0ePHnbXHBw8eFArVqzQu+++qwceeEA1atRQSEiIWrZsWVpPAcAtxmKx5DuTtnHjRp08edJuWvv27ZWamqq4uLh827iZM3Hr1q3T+fPnbY83bdpk98Y9JCREd955pxYtWqQLFy7kW//06dN2jzdt2iSLxWK7qcq1tG/fXhaLRe+8806++q1Wq1JTU22Ps7Oz9fnnnyssLCzfdWNX6tChg3Jycgq8G2N2drZtWOWN+PHHH/XFF19o5MiR1wwbHTp00MmTJxUfH59v3sWLF5Wenn7N7Tv6OkvK1wvvv/++JBVp6OGHH35oN6RxxYoVys7OLnTdvDOhVx6vn376ST/++KPdclceO0kym822DwDyvt4g70vjz507d91aAcBZOMNmQOPGjdOBAwf0r3/9S1WrVtXmzZsVFRWl9evXq1atWtq6dauqV6+ur776SlFRUZKk5s2b66WXXrK71gQAiurvf/+7Zs+erVdeeUXh4eH67bfftH79erszJ9Llm4OsW7dOMTEx2rNnj+6++25lZGQoMTFRvXr1cnh4dx4fHx898cQT6tatm+22/jVr1lTPnj0lXX6zPWHCBA0YMECdO3dWt27dFBgYqJMnT+q7775ThQoVNHfuXKWnpysuLk7Lly9XrVq17D7sygsq//d//6fdu3crPDxcd955p4YNG6bp06fr+PHjateunby9vXXs2DFt2bJFPXv2VGRkpLZt26bY2Fj93//9n+bOnVvoc7nnnnv02GOPad68edq/f79atmwpT09PHT58WJs2bdKrr76qhx566IZep2+++UYtW7Ys9OzXI488oo0bN+qNN97Qd999p4iICOXk5OjQoUPatGmTFixYcM3hsUV9na907NgxPfPMM2rdurV+/PFHffLJJ+rcufN17xApXb6JS79+/dShQwf98ccf+uCDD3T33Xdf8w6R0uVe/fzzzzVkyBD9/e9/17Fjx7Ry5UrVq1fPLoy+9tprOnv2rO69914FBgbqxIkTev/999WwYUPbrfsbNmwoi8Wi9957T+fOnZOXl5fuvfde+fn5Xbd2ACgtBDaDOXHihNasWaMvv/zSdve2yMhI/ec//9GaNWs0fPhwHT16VCdOnNCmTZs0ZcoU5eTkKCYmRs8//7yWLVvm5GcAwBU988wzysjI0Pr165WQkKBGjRpp3rx5mj59ut1yeW9u58yZo08//VSff/65KleurIiIiHxf9Ozo/v/v//5P8+fP14ULF9S8eXO98cYbtjMg0uXb4n/44Yd699139f777ys9PV0BAQEKCwvTY489JunyGaC8a8cOHjyoUaNG5dvX5s2bVaFCBYWHh0u6/JUEtWrV0pIlSzR79mxJl6/zatmypdq2bStJ2rp1qzw9PTV//ny1bt36us9n3LhxCgkJ0cqVK/Wvf/1LFotFd9xxhx5++OFr3lK/KEwmk0aMGFHoMmazWbNnz9aSJUv08ccfa/PmzSpXrpyqV6+uPn36XHfoZ1Fe5yvNmDFDsbGxmj59ujw8PNS7d+8CX/eCvP7661q/fr1mzpyprKwsderUSa+99lqhQxW7detm+57Ab775RvXq1dPUqVO1adMmff/997blHn74YcXHx+uDDz5QWlqaAgIC1KFDBz333HO2G+4EBAQoOjpa8+bN06uvvqqcnBwtW7aMwAbAUExWI1xN7saCgoLsbjry1VdfadCgQfku6s7MzNQ//vEPzZgxQ2PHjlV8fLw2bdpk+8O7b98+devWTRs3brR9+ScAGN13332nvn37KjY29obPOl3p2LFjeuCBB/TFF19c8xqqWbNm6fjx45o8efJN78+dzZo1S++8844SExPz3UgFAFB8OMNmMOnp6bJYLFq9enW+O5blhbiAgAB5eHjYfUqaN7wjKSmJwAYAAADcIghsBtOwYUPl5OTo9OnTatKkSYHLREREKDs7W0eOHNGdd94p6X/fnVOtWrXSKhUADKd8+fLq0qVLobeeDwoKUtWqVUuxKgAAbhyBzQkuXLigI0eO2B4fO3ZM+/fvl4+Pj2rXrq0uXbpo1KhRevnll9WwYUOlpqYqMTFRQUFB+vvf/64WLVooODhYY8aM0ZgxY5Sbm6tx48apZcuWRbotNQDcqqpUqXLN7z/Lc60vyQYAwIi4hs0J8q7ZuFrXrl01efJkZWVlac6cOVq3bp3++usvVa5cWY0bN9Zzzz1nu6j/5MmTmjBhgr755huVL19e9913n0aPHs1dIgEAAIBbCIENAAAAAAyKL84GAAAAAIMisAEAAACAQRHYAAAAAMCguEukE6SknJOzrxw0mSQ/v4qGqAWugZ6Bo+gZOIqegaPoGTjKSD2TV8v1ENicwGqV0xskj5FqgWugZ+AoegaOomfgKHoGjnKlnmFIJAAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQHs4uAAAAAACKk9lsktlscnYZxYLABgAAAOCWYTab5FO5vDwsBQ8mzMm1ymw2KSfHWsqV3RgCGwAAAIBbhtlskofFrBdW7taBv87bzatXtYJiHw+XyWSSRGADAAAAAKc48Nd57TuR5uwybho3HQEAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBuW1g27Fjh5555hm1atVKQUFB2rJly3XX+e6779S1a1eFhIToH//4h9asWVMKlQIAAABwV24b2NLT0xUUFKQ33nijSMsfPXpUgwYNUrNmzfTxxx/rqaee0muvvab//Oc/JVwpAAAAAHfl4ewCnKVNmzZq06ZNkZdfuXKlqlevrpdfflmSVLduXf3www9asmSJWrduXVJlAgAAAHBjbhvYHPXjjz+qefPmdtNatWqlSZMmObwtk6m4qrpxeTUYoRa4BnoGjqJn4Ch6Bo6iZ3CjTCbn901R909gK6JTp07J39/fbpq/v7/Onz+vixcvqmzZskXelp9fxeIu74YZqRa4BnoGjqJn4Ch6Bo6iZ+CoypW9nV1CkRHYnCAl5ZysVufWYDJd/uVmhFrgGugZOIqegaPoGTiKnkFBLBazfH0LD2RnzlxQdnZuKVVUsLz+vR4CWxH5+/vr1KlTdtNOnTqlChUqOHR2TZKsVhnml4qRaoFroGfgKHoGjqJn4Ch6Bo5ypZ5x27tEOqpx48bavn273bRt27apcePGzikIAAAAwC3PbQPbhQsXtH//fu3fv1+SdOzYMe3fv18nTpyQJE2fPl2jRo2yLf/444/r6NGjmjJlig4ePKi4uDht3LhR/fr1c0b5AAAAANyA2w6J3Lt3r/r27Wt7HBMTI0nq2rWrJk+erOTkZCUlJdnm16hRQ/PmzVNMTIyWLVum2267TRMmTOCW/gAAAABKjNsGtmbNmun//u//rjl/8uTJBa6zbt26EqwKAAAAAP7HbYdEAgAAAIDREdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAbl1oEtLi5Obdu2VWhoqHr06KE9e/YUuvySJUv04IMPKiwsTG3atNGkSZN06dKlUqoWAAAAgLtx28CWkJCgmJgYDRkyRGvXrlWDBg0UGRmplJSUApdfv369pk+frqFDhyohIUETJ05UQkKC3n777VKuHAAAAIC7cNvAtnjxYvXs2VPdu3dXvXr1FB0drbJly2r16tUFLr97925FRESoS5cuql69ulq1aqXOnTtf96wcAAAAANwoD2cX4AyZmZnat2+fBg0aZJtmNpvVokUL7d69u8B1wsPD9cknn2jPnj0KCwvT0aNH9fXXX+uRRx5xeP8m0w2XXmzyajBCLXAN9AwcRc/AUfQMHEXP4EaZTM7vm6Lu3y0DW2pqqnJycuTn52c33c/PT4cOHSpwnS5duig1NVVPPPGErFarsrOz9fjjj+uZZ55xeP9+fhVvqO6SYKRa4BroGTiKnoGj6Bk4ip6BoypX9nZ2CUXmloHtRnz33XeaN2+e3njjDYWFhenIkSOaOHGiZs+erSFDhji0rZSUc7JaS6jQIjKZLv9yM0ItcA30DBxFz8BR9AwcRc+gIBaLWb6+hQeyM2cuKDs7t5QqKlhe/16PWwY2X19fWSyWfDcYSUlJkb+/f4HrxMbG6uGHH1aPHj0kSUFBQUpPT9frr7+uwYMHy2wu+uWAVqsM80vFSLXANdAzcBQ9A0fRM3AUPQNHuVLPuOVNR7y8vBQcHKzExETbtNzcXCUmJio8PLzAdS5evJgvlFksFkmS1VWONgAAAACX4pZn2CSpf//+Gj16tEJCQhQWFqalS5cqIyND3bp1kySNGjVKgYGBGjFihCTp/vvv1+LFi9WoUSPbkMjY2Fjdf//9tuAGAAAAAMXJbQNbx44ddfr0ac2cOVPJyclq2LChFixYYBsSmZSUZHdGbfDgwTKZTJoxY4ZOnjypKlWq6P7779eLL77orKcAAAAA4BZnsjKer9SdOuX8C2NNJsnfv6IhaoFroGfgKHoGjqJn4Ch6BgXx8Lh805FOM/+jfSfS7OYFV6ukDc+3VmqqMW464u9//ZuOuOU1bAAAAADgCghsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABuXh7AJuRHp6unbs2KETJ04oKyvLbl7fvn2dVBUAAAAAFC+XC2y//PKLBg4cqIyMDGVkZMjHx0epqakqV66cqlSpQmADAAAAcMtwuSGRMTExuv/++7Vjxw6VKVNG8fHx+vLLLxUcHKzRo0c7tK24uDi1bdtWoaGh6tGjh/bs2VPo8mlpaYqOjlarVq0UEhKiBx98UF9//fXNPB0AAAAAuCaXC2z79+9X//79ZTabZbFYlJmZqdtvv10vvfSS3n777SJvJyEhQTExMRoyZIjWrl2rBg0aKDIyUikpKQUun5mZqf79++v48eOKjY3Vpk2bNH78eAUGBhbXUwMAAAAAOy43JNLDw0Nm8+Wc6efnpxMnTqhu3bqqUKGC/vvf/xZ5O4sXL1bPnj3VvXt3SVJ0dLS++uorrV69WgMHDsy3/OrVq3X27FmtXLlSnp6ekqTq1avf0HMwmW5otWKVV4MRaoFroGfgKHoGjqJn4Ch6BjfKZHJ+3xR1/y4X2Bo1aqSff/5ZtWrVUtOmTTVz5kylpqbq448/1t/+9rcibSMzM1P79u3ToEGDbNPMZrNatGih3bt3F7jO1q1b1bhxY40bN05ffPGFqlSpos6dO2vAgAGyWCwOPQc/v4oOLV+SjFQLXAM9A0fRM3AUPQNH0TNwVOXK3s4uochcLrC9+OKLunDhgu3fo0aN0ptvvqlatWpp0qRJRdpGamqqcnJy5OfnZzfdz89Phw4dKnCdo0ePavv27erSpYvmz5+vI0eOKDo6WtnZ2Ro6dKhDzyEl5ZysVodWKXYm0+VfbkaoBa6BnoGj6Bk4ip6Bo+gZFMRiMcvXt/BAdubMBWVn55ZSRQXL69/rcbnAFhoaavu3n5+fFi5cWCr7tVqt8vPz0/jx42WxWBQSEqKTJ09q4cKFDgc2q1WG+aVipFrgGugZOIqegaPoGTiKnoGjXKlnXO6mI3379lVaWtpNbcPX11cWiyXfDUZSUlLk7+9f4DoBAQGqVauW3fDHOnXqKDk5WZmZmTdVDwAAAAAUxOUC2/fff5/vy7Id5eXlpeDgYCUmJtqm5ebmKjExUeHh4QWuExERoSNHjig393+nTg8fPqyAgAB5eXndVD0AAAAAUBCXC2ySZCqGW7r0799f8fHxWrt2rQ4ePKg333xTGRkZ6tatmyRp1KhRmj59um35Xr166cyZM5o4caL++OMPffXVV5o3b56efPLJm64FAAAAAArictewSdKQIUNst9a/2rJly4q0jY4dO+r06dOaOXOmkpOT1bBhQy1YsMA2JDIpKcn29QGSdPvtt2vhwoWKiYnRww8/rMDAQPXt21cDBgy4+ScEAAAAAAVwycDWuHFjeXvf/K04e/furd69exc4b/ny5fmmhYeHKz4+/qb3CwAAAABF4XKBzWQyKSoqKt8t+QEAAADgVuNy17BZXeX+mwAAAABwk1wusA0dOlTly5d3dhkAAAAAUOJcbkhk3pdUnz59WocOHZJ0+fvQqlSp4syyAAAAAKDYuVxgy8jI0Lhx4/TJJ58oJydHkmSxWPTII49o7NixKleunJMrBAAAAIDi4XJDImNiYrRjxw69++672rlzp3bu3Kl3331XO3bs0OTJk51dHgAAAAAUG5cLbJ999pkmTpyoNm3aqEKFCqpQoYLatGmj8ePH67PPPnN2eQAAAABQbFwusF28eNH25dZX8vPz08WLF51QEQAAAACUDJcLbI0bN9bMmTN16dIl27SLFy/qnXfeUePGjZ1XGAAAAAAUM5e76ciYMWMUFRWl++67Tw0aNJAk/frrrypTpowWLlzo5OoAAAAAoPi4XGALCgrS559/rvXr19tu69+5c2d16dJFZcuWdXJ1AAAAAFB8XC6w7dixQ+Hh4erZs6ezSwEAAACAEuVy17D17dtXZ8+edXYZAAAAAFDiXC6wWa1WZ5cAAAAAAKXC5YZEStLu3bvl4+NT4LymTZuWcjUAAAAAUDJcMrANHTq0wOkmk0n79+8v5WoAAAAAoGS4ZGD79ttv5efn5+wyAAAAAKBEudw1bCaTydklAAAAAECpcLnAxk1HAAAAALgLlxsS+euvvzq7BAAAAAAoFS53hm3evHlatWpVvumrVq3S/PnznVARAAAAAJQMlwtsH374oerUqZNv+t/+9jetXLnSCRUBAAAAQMlwucCWnJysgICAfNOrVKmi5ORkJ1QEAAAAACXD5QLb7bffrl27duWb/sMPP6hq1apOqAgAAAAASobL3XSkR48emjRpkrKzs3XvvfdKkhITEzV16lQ9/fTTTq4OAAAAAIqPywW2qKgonTlzRtHR0crKypIklSlTRlFRURo0aJCTqwMAAACA4uNygc1kMumll17Ss88+q4MHD6ps2bKqVauWvLy8nF0aAAAAABQrlwtseby9vRUWFubsMgAAAACgxLhkYPv555+1ceNGJSUl2YZF5nnnnXecVBUAAAAAFC+Xu0vkhg0b1KtXLx06dEibN29Wdna2fv/9d23fvl0VK1Z0dnkAAAAAUGxcLrDNnTtXr7zyiubOnStPT0+9+uqr2rRpkzp06KDbb7/d2eUBAAAAQLFxucB29OhRtWnTRpLk5eWl9PR0mUwm9evXT/Hx8U6uDgAAAACKj8sFtkqVKunChQuSpKpVq+r333+XJKWlpSkjI8OZpQEAAABAsXK5m440bdpU27ZtU1BQkB566CFNnDhR27dv17Zt29S8eXNnlwcAAAAAxcblAtvYsWN16dIlSdLgwYPl6empXbt2qX379ho8eLCTqwMAAACA4uMyge38+fOSJA8PD3l4eNgeP/HEE3riiSecWRoAAAAAlAiXCWxNmjSRyWS67nL79+8vhWoAAAAAoOS5TGBbtmyZ3WOr1aqBAwdqwoQJCgwMdFJVAAAAAFByXCaw3XPPPfmmmc1mNW7cWDVq1HBCRQAAAABQslzutv4AAAAA4C5cNrAlJSXp4sWLqly5srNLAQAAAIAS4TJDIq+8hi01NVUbNmzQvffeq4oVKzqxKgAAAAAoOS4T2JYsWSJJMplM8vX11f3338/3rgEAAAC4pblMYNu6dauzSwAAAACAUuWy17ABAAAAwK2OwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDcuvAFhcXp7Zt2yo0NFQ9evTQnj17irTehg0bFBQUpGeffbaEKwQAAADgztw2sCUkJCgmJkZDhgzR2rVr1aBBA0VGRiolJaXQ9Y4dO6a33npLTZo0KaVKAQAAALgrtw1sixcvVs+ePdW9e3fVq1dP0dHRKlu2rFavXn3NdXJycjRy5Eg999xzqlGjRilWCwAAAMAdeTi7AGfIzMzUvn37NGjQINs0s9msFi1aaPfu3ddcb/bs2fLz81OPHj30ww8/3PD+TaYbXrXY5NVghFrgGugZOIqegaPoGTiKnsGNMpmc3zdF3b9bBrbU1FTl5OTIz8/Pbrqfn58OHTpU4Do7d+7UqlWrtG7dupvev59fxZveRnExUi1wDfQMHEXPwFH0DBxFz8BRlSt7O7uEInPLwOao8+fPa9SoURo/fryqVKly09tLSTknq7UYCrsJJtPlX25GqAWugZ6Bo+gZOIqegaPoGRTEYjHL17fwQHbmzAVlZ+eWUkUFy+vf63HLwObr6yuLxZLvBiMpKSny9/fPt/zRo0d1/PhxDR482DYtN/fyAW7UqJE2bdqkO++8s8j7t1plmF8qRqoFroGegaPoGTiKnoGj6Bk4ypV6xi0Dm5eXl4KDg5WYmKh27dpJuhzAEhMT1bt373zL16lTR+vXr7ebNmPGDF24cEGvvvqqbrvttlKpGwAAAIB7ccvAJkn9+/fX6NGjFRISorCwMC1dulQZGRnq1q2bJGnUqFEKDAzUiBEjVKZMGdWvX99u/UqVKklSvukAAAAAUFzcNrB17NhRp0+f1syZM5WcnKyGDRtqwYIFtiGRSUlJMpvd9lsPAAAAABiA2wY2Serdu3eBQyAlafny5YWuO3ny5JIoCQAAAABsOIUEAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGBSBDQAAAAAMisAGAAAAAAZFYAMAAAAAgyKwAQAAAIBBEdgAAAAAwKAIbAAAAABgUAQ2AAAAADAoAhsAAAAAGJSHswuAc1ksBWf23FyrcnOtpVwNAAAAgCsR2NyU2WxSTq5Vvr7eBc7PzsnV2TPphDYAAADAiQhsbspkMsliNumFlbt14K/zdvPqVa2g2MfDZTabCGwAAACAExHY3NyBv85r34k0Z5cBAAAAoADcdAQAAAAADIrABgAAAAAGRWADAAAAAIMisAEAAACAQRHYAAAAAMCgCGwAAAAAYFAENgAAAAAwKAIbAAAAABgUgQ0AAAAADIrABgAAAAAGRWADAAAAAIMisAEAAACAQRHYAAAAAMCgCGwAAAAAYFAENgAAAAAwKAIbAAAAABgUgQ0AAAAADIrABgAAAAAGRWADAAAAAIMisAEAAACAQbl1YIuLi1Pbtm0VGhqqHj16aM+ePddcNj4+Xk888YSaNm2qpk2bql+/foUuDwAAAAA3y20DW0JCgmJiYjRkyBCtXbtWDRo0UGRkpFJSUgpc/rvvvlOnTp20bNkyrVy5UrfffruefvppnTx5spQrBwAAAOAu3DawLV68WD179lT37t1Vr149RUdHq2zZslq9enWBy0+fPl1PPvmkGjZsqLp162rChAnKzc1VYmJiKVcOAAAAwF14OLsAZ8jMzNS+ffs0aNAg2zSz2awWLVpo9+7dRdpGRkaGsrOz5ePj4/D+TSaHVyl2Ra3BCLXCGPJ6gZ5AUdEzcBQ9A0fRM7hRJpPz+6ao+3fLwJaamqqcnBz5+fnZTffz89OhQ4eKtI1p06apatWqatGihcP79/Or6PA6zuDr6+3sEmBArtK/MA56Bo6iZ+AoegaOqlzZdd7numVgu1nz589XQkKCli1bpjJlyji8fkrKOVmtJVCYAzw8zNdt1NTUC8rJyS2limB0JtPlP4hG6F+4BnoGjqJn4Ch6BgWxWMzXPfFw5swFZWc7931uXv9ej1sGNl9fX1kslnw3GElJSZG/v3+h6y5cuFDz58/X4sWL1aBBgxvav9Uqp/9SKer+nV0njMcI/QvXQs/AUfQMHEXPwFGu1DNuedMRLy8vBQcH290wJO8GIuHh4ddc77333tO7776rBQsWKDQ0tDRKBQAAAODG3PIMmyT1799fo0ePVkhIiMLCwrR06VJlZGSoW7dukqRRo0YpMDBQI0aMkHR5GOTMmTM1ffp03XHHHUpOTpYklS9fXt7erjMGFgAAAIDrcNvA1rFjR50+fVozZ85UcnKyGjZsqAULFtiGRCYlJcls/t8JyJUrVyorK0vPP/+83XaGDh2q5557rlRrBwAAAOAe3DawSVLv3r3Vu3fvAuctX77c7vHWrVtLoyQAAAAAsHHLa9gAAAAAwBUQ2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQBDYAAAAAMCgCGwAAAAAYFIENAAAAAAyKwAYAAAAABkVgAwAAAACDIrABAAAAgEER2AAAAADAoAhsAAAAAGBQHs4uAAAAwNWYzSaZzaYC5+XmWpWbay3likqXuz9/oDQR2AAAABxgNpvkU7m8PCwFD1TKzsnV2TPpt2xocffnD5Q2AhsAAIADzGaTPCxmvbBytw78dd5uXr2qFRT7eLjMZtMtG1jc/fkDpY3ABgAAcAMO/HVe+06kObsMp3H35w+UFgIbAACljOt/gOJlucbwTP4/4VZAYAMAoBRx/Q9QfMxmk3JyrfL19S5wPv+fcCsgsAEAUIq4/gcoPiaTSRazif9PuKUR2AAAcAKu/wGKD/+fcCvji7MBAAAAwKAIbAAAAABgUAyJBAAAAK7AnVxhJAQ2AAAA4P/jTq4wGgIbAAAA8P9xJ1cYDYENAAAAuAp3noRRcNMRAAAAADAoAhsAAAAAGBSBDQAAAAAMimvYAAAAgGLA1wGgJBDYAAAAgJvE1wGgpBDYAAAAgJvE1wGgpBDYAAAAgGLC1wGguHHTEQAAAAAwKAIbAAAAABgUgQ0AAAAADIpr2AAAAAAn4usAUBgCGwAAAOAkfB0ArofABgAAADgJXweA6yGwAQAAQ7Fc40wDQ8NwK+PrAHAtBDYAMCiuaYC7MZtNysm1ytfXu8D5DA0D4I4IbABgQFzTAHdkMplkMZsYGuYgPty59XHW2b0R2ADAgLimAe6MoWFFx4c7t7aACmWUk2tVpUrlCpxf2PElyN86CGwAYGC8cQVQGD7cubVVKudxQ2edCfK3FgIbAACAi+PDnVubo8eXIH9rIbABAAAAtyCC/K2BwAbAJTE233VwsTyAPPzuBhxHYAPgchib7xq4RfutjzffcAS/u4EbQ2AD4BAjnC1hbL5r4BbttzajvfkmPBofv7uBG+PWgS0uLk4LFy5UcnKyGjRooLFjxyosLOyay2/cuFGxsbE6fvy4atWqpZEjR6pNmzalWDFgrzTfoBjxbAlj813DjRwn3nwbn5HefBstPKJwpf272wgfNBoRr4vrcNvAlpCQoJiYGEVHR+uuu+7S0qVLFRkZqU2bNsnPzy/f8rt27dKIESM0fPhw3X///Vq/fr2GDBmiNWvWqH79+k54BriV3Mib09J+g+IOZ0sICcbAm+/SdzO9b4QPTowUHkvKtY7Rtd50F2WZW/332s18h9n1FPSaFuVYGEFRXpfz5y7Kas3/utzqPWNUbhvYFi9erJ49e6p79+6SpOjoaH311VdavXq1Bg4cmG/5ZcuWqXXr1oqKipIkDRs2TNu2bdP777+vcePGlWrtKHmFvXm5UcUdvJz1BuVWPVtyMyHhev1ilOd4PUY5TjfT2zf6f/dWOEY3ymQyqULFsoX2fkFv3pwVFAp7o1zY76fifDN9M8/9RlzvGF3LzbwxL8yNPv+SeN0KW74o32Hm6WlRTk5ukbd5vdf0Ruq92V5xdJuFvS5Na/lqbOdgVa5cvsB1b7RnSoqr/O6+WW4Z2DIzM7Vv3z4NGjTINs1sNqtFixbavXt3gev8+OOP6tevn920Vq1aacuWLQ7v32yWnN3npv//9z64WiWV87LYzavjf3nI27X+s1ut/1vfVecVNv96fxhzcq2yXOMNU2HzsnNydeH8xXy/WCwWszwsZs396qBOnM2wm1fNp5ye+XvdQv+glPEw5zuGZTzMdstc7UZeN4//v01He6YobwYLel0KqyVvP8XZvzd6HIryRsooz7GweSVxnG60Z260t2/m/+6NPL8bnXejx7YovXajv58kFdj7f6taQU80q3nNN2/Xeh5331n5ukGhoNe7sJ653jZvdL2bec1Ke38FHaPQO3zUo0mNAmsJv7OyLGbTDR3b69UiGec4XWubdQMqSCr4d8ltla4fvBx9TQs7Fjf6uuQ9h5LYZkGvi295rxLpmZvp+xt9byUV/nvWZLr8ntyZCnufarec1SgRuRSdPHlS9913n1auXKnw8HDb9ClTpmjHjh366KOP8q0TEhKiyZMnq3PnzrZpcXFxmj17trZt21YqdQMAAABwL64x2BYAAAAA3JBbBjZfX19ZLBalpKTYTU9JSZG/v3+B6/j7++vUqVNFXh4AAAAAbpZbBjYvLy8FBwcrMTHRNi03N1eJiYl2QySv1LhxY23fvt1u2rZt29S4ceOSLBUAAACAG3PLwCZJ/fv3V3x8vNauXauDBw/qzTffVEZGhrp16yZJGjVqlKZPn25bvm/fvvrPf/6jRYsW6eDBg5o1a5b27t2r3r17O+spAAAAALjFueVdIiWpY8eOOn36tGbOnKnk5GQ1bNhQCxYssA1xTEpKkvmKW8dERERo2rRpmjFjht5++23VqlVLs2fP5jvYAAAAAJQYt7xLJAAAAAC4ArcdEgkAAAAARkdgAwAAAACDIrABAAAAgEER2AAAAADAoAhsbiouLk5t27ZVaGioevTooT179ji7JBjAvHnz1L17d4WHh6t58+Z69tlndejQIbtlLl26pOjoaDVr1kzh4eF67rnn8n2pPNzX/PnzFRQUpIkTJ9qm0TO42smTJzVy5Eg1a9ZMYWFh6tKli37++WfbfKvVqtjYWLVq1UphYWHq16+fDh8+7LyC4VQ5OTmaMWOG2rZtq7CwMLVr106zZ8/WlffNo2fc244dO/TMM8+oVatWCgoK0pYtW+zmF6U/zpw5oxEjRigiIkJNmjTRmDFjdOHChVJ8FtdGYHNDCQkJiomJ0ZAhQ7R27Vo1aNBAkZGRSklJcXZpcLLvv/9eTz75pOLj47V48WJlZ2crMjJS6enptmUmTZqkL7/8UjNmzNDy5cv1119/aejQoU6sGkaxZ88erVy5UkFBQXbT6Rlc6ezZs+rVq5c8PT313nvvacOGDRo9erR8fHxsy7z33ntavny53nzzTcXHx6tcuXKKjIzUpUuXnFg5nOW9997TihUr9PrrryshIUEjR47UggULtHz5crtl6Bn3lZ6erqCgIL3xxhsFzi9Kf4wcOVIHDhzQ4sWLNXfuXO3cuVOvv/56aT2Fwlnhdh599FFrdHS07XFOTo61VatW1nnz5jmxKhhRSkqKtX79+tbvv//earVarWlpadbg4GDrxo0bbcscOHDAWr9+fevu3budVCWM4Pz589b27dtbv/32W2vv3r2tEyZMsFqt9Azymzp1qrVXr17XnJ+bm2tt2bKldcGCBbZpaWlp1pCQEOunn35aGiXCYAYOHGh95ZVX7KYNHTrUOmLECKvVSs/AXv369a2bN2+2PS5Kf+T9XdqzZ49tma+//toaFBRk/e9//1t6xV8DZ9jcTGZmpvbt26cWLVrYppnNZrVo0UK7d+92YmUwonPnzkmS7ZPvvXv3Kisry65/6tatq2rVqunHH390RokwiHHjxqlNmzZ2vSHRM8hv69atCgkJ0fPPP6/mzZvrn//8p+Lj423zjx07puTkZLueqVixou666y7+Trmp8PBwbd++XX/88Yck6ddff9UPP/yg++67TxI9g8IVpT92796tSpUqKTQ01LZMixYtZDabDXHZkIezC0DpSk1NVU5Ojvz8/Oym+/n55btWCe4tNzdXkyZNUkREhOrXry9JOnXqlDw9PVWpUiW7Zf38/JScnOyMMmEAGzZs0C+//KJVq1blm0fP4GpHjx7VihUr1L9/fz3zzDP6+eefNWHCBHl6eqpr1662vijo7xTXPrqngQMH6vz58+rQoYMsFotycnL04osv6uGHH5YkegaFKkp/nDp1SlWqVLGb7+HhIR8fH0P8rSKwAShQdHS0fv/9d33wwQfOLgUGlpSUpIkTJ2rRokUqU6aMs8uBC7BarQoJCdHw4cMlSY0aNdLvv/+ulStXqmvXrk6uDka0ceNGrV+/XtOnT1e9evW0f/9+xcTEqGrVqvQM3AJDIt2Mr6+vLBZLvhuMpKSkyN/f30lVwWjGjRunr776SkuXLtVtt91mm+7v76+srCylpaXZLZ+SkqKAgIDSLhMGsG/fPqWkpKhbt25q1KiRGjVqpO+//17Lly9Xo0aN6BnkExAQoLp169pNq1Onjk6cOGGbL4m/U7CZMmWKBg4cqE6dOikoKEj//Oc/9dRTT2nevHmS6BkUrij94e/vr9OnT9vNz87O1tmzZw3xt4rA5ma8vLwUHBysxMRE27Tc3FwlJiYqPDzciZXBCKxWq8aNG6fNmzdr6dKlqlGjht38kJAQeXp62vXPoUOHdOLECTVu3LiUq4UR3HvvvVq/fr3WrVtn+wkJCVGXLl1s/6ZncKWIiAjbtUh5Dh8+rDvuuEOSVL16dQUEBNj1zPnz5/XTTz/xd8pNXbx4USaTyW6axWKx3dafnkFhitIf4eHhSktL0969e23LbN++Xbm5uQoLCyv1mq/GkEg31L9/f40ePVohISEKCwvT0qVLlZGRoW7dujm7NDhZdHS0Pv30U7377rvy9va2jduuWLGiypYtq4oVK6p79+6aPHmyfHx8VKFCBU2YMEHh4eG8+XZTFSpUsF3jmKd8+fKqXLmybTo9gys99dRT6tWrl+bOnasOHTpoz549io+P17hx4yRJJpNJffv21Zw5c1SzZk1Vr15dsbGxqlq1qtq1a+fk6uEM999/v+bOnatq1arZhkQuXrxY3bt3l0TPQLpw4YKOHDlie3zs2DHt379fPj4+qlat2nX7o27dumrdurXGjh2r6OhoZWVlafz48erUqZMCAwOd9bRsTFbrFd86CLfx/vvva+HChUpOTlbDhg312muv6a677nJ2WXCyq78/K09MTIwt0F+6dEmTJ0/Whg0blJmZqVatWumNN94wxJABGEOfPn3UoEEDvfrqq5LoGeT35Zdf6u2339bhw4dVvXp19e/fXz179rTNt1qtmjlzpuLj45WWlqa7775bb7zxhmrXru3EquEs58+fV2xsrLZs2aKUlBRVrVpVnTp10pAhQ+Tl5SWJnnF33333nfr27ZtveteuXTV58uQi9ceZM2c0fvx4bd26VWazWe3bt9drr70mb2/v0nwqBSKwAQAAAIBBcQ0bAAAAABgUgQ0AAAAADIrABgAAAAAGRWADAAAAAIMisAEAAACAQRHYAAAAAMCgCGwAAAAAYFAENgAAAAAwKAIbAAAAABiUh7MLAIBb0csvv6y1a9dec/6OHTtUqVKlUqwIAAC4IgIbAJSQ1q1bKyYmxm7a7t279dxzzzmpIgAA4GoYEgkAJcTLy0sBAQF2Pz4+PnbLrFmzRk2aNNGWLVvUvn17hYaGKjIyUklJSXbLffDBB2rXrp1CQkL04IMPat26dXbzg4KCtGXLFrtpffr00cSJE+2mffTRR+rQoYNCQ0P10EMPKS4uzjbv2LFjCgoK0v79++3Wadu2rZYsWWL7d1BQUIE/a9asKfB1ePnllxUUFKTFixfbTR8yZEi+9ZKSkvTCCy+oSZMmuueeezR48GAdO3ZMkjRr1qxr7rtPnz62fT377LO27aWmpqpp06Zq0qSJ3b7znuvVP2lpabZ9PfLII7blMzMz9Y9//MNumeI8dnk/ERER6t+/v44cOWKbv27dOnXr1k3h4eFq2bKlRowYoZSUFNv87777zq6uK7eb1xNFObZXr3O1Rx55RLNmzbLtMyQkRDt37rTNf++999S8eXOdOnWqwPXzXq8rFVTX1TV89NFHCgoKsvVynz59rtkHefWdPXtWo0aNUtOmTXXXXXcpKipKhw8ftm0zISFB7dq1U2hoqJo1a6bnn39ep0+fLnINBb12Uv7+y8zM1IQJE9S8eXOFhoaqV69e2rNnj21+3rELCgpSgwYN1Lx5c40ZM0aXLl0q8DUE4J4IbADgZBcvXtScOXP01ltvacWKFUpLS9OLL75om79582ZNmjRJ/fv31/r16/X4449rzJgx2r59u0P7+eSTTxQbG6sXX3xRCQkJGj58uGbOnFno0M2rrVq1St98842++eYb3XbbbRozZoztcceOHa+5XmBgoOLj422PT548qV27dqlcuXK2aVlZWYqMjJS3t7fi4uK0YsUKlS9fXlFRUcrMzNTTTz9t29fTTz+t8PBw2+O8N+pXmz17trKzs/NNt1qtkqQlS5YUun6euLi4AoNIcR27mJgYffPNN4qLi1NKSorefvtt27zs7Gy98MIL+uSTTzR79mwdP35cL7/8cqH1lrRmzZqpb9++GjVqlM6dO6dffvlFsbGxmjBhgvz9/YttP+np6YqNjVX58uVt02bNmmU77uHh4fn6QrocnPbu3as5c+boww8/lNVq1cCBA5WVlSVJqlOnjiZPnqxNmzZpwYIFOnbsmKZOnVrkGopqypQp+uyzzzR58mStXbtWNWvWVFRUlM6cOWO33KZNm/Tvf/9bU6dO1caNG7V69WqH9wXg1sWQSABwsqysLL3++uu66667JEmTJ09Wx44dtWfPHoWFhWnhwoXq2rWrnnzySUlS7dq19eOPP2rRokW69957i7yfWbNm6eWXX1b79u0lSTVq1NCBAwf04YcfqmvXrkXaRpUqVWz/tlgsqlixogICAq67XkhIiI4dO6adO3eqSZMmWr16tTp16mR3tikhIUG5ubmaOHGiTCaTpMtBpmnTpvr+++/VqlUreXt7S5LKly8vT0/PQvf9xx9/aPXq1erXr5+WL19uNy8vxPn7+xd45vNKZ86c0Zw5czRgwADFxsbazSuuY1epUiUFBASobNmy8vb2VsWKFW3zHn30Udu/a9SooVdffVWPPvqoLly4YHs9nGHYsGHatm2bxo4dq99//11du3bVAw88UKz7WLBggerVq6ecnBzbtMqVK9v+7enpqfLly9v1weHDh7V161atWLFCERERkqRp06bp73//u7Zs2aIOHTqoQYMGtuV9fHzk6+ur3NzcItdQFOnp6Vq5cqViYmLUpk0bSdL48eP17bffatWqVYqKirIt6+fnp0qVKikjI0Oenp52xx8ACGwA4GQeHh4KDQ21Pa5bt64qVaqkgwcPKiwsTIcOHdJjjz1mt05ERISWLVtW5H2kp6fryJEjevXVVzV27Fjb9Ozs7HxvDh9//HGZzf8bgJGRkeHoUypQz5499eGHHyoiIkKrVq3SnDlz7ALbr7/+qiNHjtjeZOe5dOmS3RDBopo6daoee+wx1ahRI9+88+fPS5LdGb5rmT17tpo1a6a7774737ziOnbDhw+XxWJRRkaG6tevr+HDh9vm7d27V++8845+/fVXnT171nZ2MCkpSfXq1bMtlxcKClOUY5tXi7e3txo1aqRRo0bZ7SePl5eXpk2bpocffljVqlXTK6+8ct39O+LkyZNavHixVqxYkW9ob2EOHjwoDw8PW4iWJF9fX9WuXVsHDx60Tdu5c6cGDBigjIwM3X333Ro9erTDNUybNs0uxGdmZtqOw5EjR5SVlWXXz56engoLC7OrQ7p87KxWqzIyMtSlSxd17ty5yM8XwK2PwAYAbiA9PV3S5U/4r3wjK8nuDbwk/etf/1LdunVtj/OuD7tZDz/8sGbNmqUNGzYoICBAQUFB+WoMDg7WtGnT8q175Zm9ovj+++/1ww8/KCYmRl988UW++X/99ZfMZvN1zw4ePnxYq1at0rp16/Tf//7XoRoc8corr6hFixZKS0vTjBkz9Morr2ju3LlKT09XZGSkWrVqpWnTpsnX11dJSUmKjIy0De/LExcXZ3fGLe9M6pWKcmyvrOVf//qXhg0bpk8//bTAunfv3i3p8jVjZ8+evaFhg9cyY8YMPfTQQ3Znw4pTSEiI1q5dqxMnTmjixIn66KOPNGjQIIdqiIyMVLdu3WyPp02b5vCZOOl/x+7YsWMaO3asFi9ebBveCQBcwwYATpadna29e/faHh86dEhpaWm2N9Z16tTRrl277NbZtWtXgWc9rsXf319Vq1bV0aNHVbNmTbufq89A3X777XbzPTyK57O9SpUqqW3btnrjjTfUo0ePfPODg4P1559/ys/PL1+NjgwRs1qtmjx5sgYPHnzNoY4///yz6tSpozJlyhS6rWnTpunRRx9VzZo1C5xfXMcuICBANWvWVGhoqHr37q2vvvpKWVlZOnTokM6cOaORI0eqSZMmqlu3rt0NR65UvXp1u9esIEU5tlfW0rdvXx04cCBfOJQun0GaNGmSxo8fr7CwMI0ePfqawwodtX//fm3atEnDhg1zeN26desqOztbP/30k21aamqq/vjjD7vXvWzZsqpVq5ZatGihnj17av369Q7X4Ovra/d6XhmY77zzTnl6etod/6ysLP3888/5jn/esWvZsqUefPBBbd682eHnDeDWxRk2AHAyT09PjR8/Xq+99posFovGjx+vxo0bKywsTJIUFRWlYcOGqWHDhmrRooW+/PJLbd68Od9dF7OysuzuLpebm6vc3FxlZ2fLw8NDzz//vCZMmKCKFSuqdevWyszM1N69e5WWlqb+/fuXynMdMGCAateuXeANSrp06aKFCxdq8ODBeuGFFxQYGKgTJ05o8+bNioqK0m233VakfWzfvl0BAQG268aulJmZqYSEBC1ZsuS6X69w5MgRJSUl6fPPP7/mMsV17NLS0pScnKxz585p1apVqlGjhjw9PVWtWjV5enpq+fLl6tWrl3777Te9++67RXodblReH6WlpemTTz5RrVq15OnpabdMTk6OXnrpJbVu3Vrdu3dX69at1aVLFy1atMju2qyCXNmjmZmZtn1arVbbtYuLFi1S//79FRgY6HD9tWrV0gMPPKCxY8cqOjpaFSpU0LRp0xQYGGi7xm7Dhg2qUaOG/P399eeff2rlypUKCQmx287N1CBdvs6yV69emjJlinx8fFStWjUtWLBAFy9etLsuUZJSUlJ06dIlnThxQlu3bs13N00A7o3ABgBOVrZsWQ0YMEAjRozQyZMn1aRJE7vrZdq1a6cxY8Zo0aJFmjRpku644w5NmjRJzZo1s9tOQWcCdu7cqcqVK+u5555Tjx49VLZsWS1cuFBTpkxR+fLlVb9+fT311FMl/RRt6tSpo4EDBxY4r1y5cnr//fc1bdo0DR06VBcuXFBgYKCaN2+uChUqFHkf6enpGjFiRL6QIUm//fab3nnnHT377LPXDanp6ekaPXq03U0urlZcxy7v+i9vb28FBwdr5syZki4PBZ08ebLefvttLV++XMHBwRo9erQGDx5c1JfDYXl9lFfLjBkz8i0zZ84cHT9+XHPnzpUkVa1aVePHj9fw4cPVqlWraw4hPHfunC3MXqlHjx764osvVL16ddu+rxf8ChMTE6OJEyfqmWeeUVZWlpo0aaL58+fbeuLgwYOaOnWqTp06JV9fX9133335rmG72RokaeTIkbJarRo1apQuXLigkJAQLViwIN+Z34ceekjS5TN2LVu21EsvvXRT+wVwazFZ865eBgCUujVr1mjSpEl232dVnJYsWaJz587xZd0loKSPnTt55JFHNHv2bFtgAwD8D9ewAcAtzMPDQxaLxdllAIXy8vKyDYcEANhjSCQA3MJ69+7t7BKA6/roo4+cXQIAGBZDIgEAAADAoBgSCQAAAAAGRWADAAAAAIMisAEAAACAQRHYAAAAAMCgCGwAAAAAYFAENgAAAAAwKAIbAAAAABgUgQ0AAAAADOr/AQ0Zhbb24TJCAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Визуализируем уровень покрытия и метилирования"
      ],
      "metadata": {
        "id": "AVELA-BETm8C"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! bedtools genomecov -bg -ibam /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bam -g m.chromSizes  >  /content/drive/MyDrive/bioinf/s_8_cell.cover.bedGraph\n",
        "! bedtools genomecov -bg -ibam /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bam -g m.chromSizes  >  /content/drive/MyDrive/bioinf/s_icm.cover.bedGraph\n",
        "! bedtools genomecov -bg -ibam /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bam -g m.chromSizes  >  /content/drive/MyDrive/bioinf/s_epiblast.cover.bedGraph"
      ],
      "metadata": {
        "id": "eOvbw0btcO-Z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! bedGraphToBigWig  /content/drive/MyDrive/bioinf/s_8_cell.cover.bedGraph m.chrom.sizes /content/drive/MyDrive/bioinf/cell8_coverage.bw\n",
        "! bedGraphToBigWig  /content/drive/MyDrive/bioinf/s_epiblast.cover.bedGraph m.chrom.sizes /content/drive/MyDrive/bioinf/epiblast_coverage.bw\n",
        "! bedGraphToBigWig  /content/drive/MyDrive/bioinf/s_icm.cover.bedGraph m.chrom.sizes /content/drive/MyDrive/bioinf/icm_coverage.bw"
      ],
      "metadata": {
        "id": "RLpT4wFCvr29"
      },
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%cd /content/drive/MyDrive/bioinf/\n",
        "! make_tracks_file --trackFiles cell8_coverage.bw icm_coverage.bw epiblast_coverage.bw -o tracks_cov.ini"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hHfBkhpMZ-4K",
        "outputId": "0a0ea726-5a61-4138-a751-34a1074105e7"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/bioinf\n",
            "Adding bigwig file: cell8_coverage.bw\n",
            "Adding bigwig file: icm_coverage.bw\n",
            "Adding bigwig file: epiblast_coverage.bw\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! pyGenomeTracks --tracks tracks_cov.ini --region chr11:3100030-3500030 -o image_cov.png"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bAaOdZCKsVtX",
        "outputId": "510e14c1-c26d-4812-829e-6d6426198a78"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:initialize 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:initialize 3. [cell8_coverage]\n",
            "INFO:pygenometracks.tracksClass:initialize 4. [icm_coverage]\n",
            "INFO:pygenometracks.tracksClass:initialize 5. [epiblast_coverage]\n",
            "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
            "INFO:pygenometracks.tracksClass:0.038223981857299805\n",
            "DEBUG:pygenometracks.tracksClass:Figure size in cm is 40 x 8.909574468085108. Dpi is set to 72\n",
            "\n",
            "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:plotting 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:plotting 3. [cell8_coverage]\n",
            "INFO:pygenometracks.tracksClass:plotting 4. [icm_coverage]\n",
            "INFO:pygenometracks.tracksClass:plotting 5. [epiblast_coverage]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bedGraphToBigWig /content/drive/MyDrive/bioinf/SRR5836473_1_bismark_bt2_pe.deduplicated.bedGraph m.chrom.sizes /content/drive/MyDrive/bioinf/cell8_meth.bw\n",
        "! bedGraphToBigWig /content/drive/MyDrive/bioinf/SRR5836475_1_bismark_bt2_pe.deduplicated.bedGraph m.chrom.sizes /content/drive/MyDrive/bioinf/icm_meth.bw\n",
        "! bedGraphToBigWig /content/drive/MyDrive/bioinf/SRR3824222_1_bismark_bt2_pe.deduplicated.bedGraph m.chrom.sizes /content/drive/MyDrive/bioinf/epiblast_meth.bw"
      ],
      "metadata": {
        "id": "dProVYcFY4pF"
      },
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! make_tracks_file --trackFiles cell8_meth.bw icm_meth.bw epiblast_meth.bw -o tracks_meth.ini"
      ],
      "metadata": {
        "id": "p7CjvnFhwpyL",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c7f8c508-3486-4272-c3c7-b75dd38c9779"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Adding bigwig file: cell8_meth.bw\n",
            "Adding bigwig file: icm_meth.bw\n",
            "Adding bigwig file: epiblast_meth.bw\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! pyGenomeTracks --tracks tracks_meth.ini --region chr11:3100030-3500030 -o image_meth.png"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XLShnovnaYND",
        "outputId": "d0323f27-6560-4eda-ec60-78ac3b1a6b3f"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:initialize 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:initialize 3. [cell8_meth]\n",
            "INFO:pygenometracks.tracksClass:initialize 4. [icm_meth]\n",
            "INFO:pygenometracks.tracksClass:initialize 5. [epiblast_meth]\n",
            "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
            "INFO:pygenometracks.tracksClass:0.028142690658569336\n",
            "DEBUG:pygenometracks.tracksClass:Figure size in cm is 40 x 8.909574468085108. Dpi is set to 72\n",
            "\n",
            "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:plotting 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:plotting 3. [cell8_meth]\n",
            "INFO:pygenometracks.tracksClass:plotting 4. [icm_meth]\n",
            "INFO:pygenometracks.tracksClass:plotting 5. [epiblast_meth]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "bKqR-Gg2au90"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}